{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 553,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "\n",
    "import sbmlcore\n",
    "import pytest\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load WHO data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 554,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>GENBANK_REFERENCE</th>\n",
       "      <th>CATALOGUE_NAME</th>\n",
       "      <th>CATALOGUE_VERSION</th>\n",
       "      <th>CATALOGUE_GRAMMAR</th>\n",
       "      <th>PREDICTION_VALUES</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>OTHER</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1178</th>\n",
       "      <td>645</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA@A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"Present_SOLO_R\": 590, \"Present_SOLO_SR\": 658...</td>\n",
       "      <td>{\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>646</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>MXF</td>\n",
       "      <td>gyrA@A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"Present_SOLO_R\": 253, \"Present_SOLO_SR\": 598...</td>\n",
       "      <td>{\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1180</th>\n",
       "      <td>647</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA@D94A</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"Present_SOLO_R\": 181, \"Present_SOLO_SR\": 207...</td>\n",
       "      <td>{\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181</th>\n",
       "      <td>648</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>MXF</td>\n",
       "      <td>gyrA@D94A</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"Present_SOLO_R\": 89, \"Present_SOLO_SR\": 187,...</td>\n",
       "      <td>{\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182</th>\n",
       "      <td>649</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA@D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"Present_SOLO_R\": 1133, \"Present_SOLO_SR\": 11...</td>\n",
       "      <td>{\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1264</th>\n",
       "      <td>731</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>MXF</td>\n",
       "      <td>gyrB@c-165t</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"Present_SOLO_R\": 2, \"Present_SOLO_SR\": 333, ...</td>\n",
       "      <td>{\"FINAL_CONFIDENCE_GRADING\": \"5) Not assoc w R\"}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2126</th>\n",
       "      <td>1593</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA@*=</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2131</th>\n",
       "      <td>1598</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>MXF</td>\n",
       "      <td>gyrA@*=</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2136</th>\n",
       "      <td>1603</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB@*=</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2141</th>\n",
       "      <td>1608</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>WHO-UCN-GTB-PCI-2021.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GARC1</td>\n",
       "      <td>RFUS</td>\n",
       "      <td>MXF</td>\n",
       "      <td>gyrB@*=</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0.1 GENBANK_REFERENCE          CATALOGUE_NAME  \\\n",
       "1178           645       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "1179           646       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "1180           647       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "1181           648       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "1182           649       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "...            ...               ...                     ...   \n",
       "1264           731       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "2126          1593       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "2131          1598       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "2136          1603       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "2141          1608       NC_000962.3  WHO-UCN-GTB-PCI-2021.7   \n",
       "\n",
       "      CATALOGUE_VERSION CATALOGUE_GRAMMAR PREDICTION_VALUES DRUG     MUTATION  \\\n",
       "1178                1.0             GARC1              RFUS  LEV    gyrA@A90V   \n",
       "1179                1.0             GARC1              RFUS  MXF    gyrA@A90V   \n",
       "1180                1.0             GARC1              RFUS  LEV    gyrA@D94A   \n",
       "1181                1.0             GARC1              RFUS  MXF    gyrA@D94A   \n",
       "1182                1.0             GARC1              RFUS  LEV    gyrA@D94G   \n",
       "...                 ...               ...               ...  ...          ...   \n",
       "1264                1.0             GARC1              RFUS  MXF  gyrB@c-165t   \n",
       "2126                1.0             GARC1              RFUS  LEV      gyrA@*=   \n",
       "2131                1.0             GARC1              RFUS  MXF      gyrA@*=   \n",
       "2136                1.0             GARC1              RFUS  LEV      gyrB@*=   \n",
       "2141                1.0             GARC1              RFUS  MXF      gyrB@*=   \n",
       "\n",
       "     PREDICTION SOURCE                                           EVIDENCE  \\\n",
       "1178          R     {}  {\"Present_SOLO_R\": 590, \"Present_SOLO_SR\": 658...   \n",
       "1179          R     {}  {\"Present_SOLO_R\": 253, \"Present_SOLO_SR\": 598...   \n",
       "1180          R     {}  {\"Present_SOLO_R\": 181, \"Present_SOLO_SR\": 207...   \n",
       "1181          R     {}  {\"Present_SOLO_R\": 89, \"Present_SOLO_SR\": 187,...   \n",
       "1182          R     {}  {\"Present_SOLO_R\": 1133, \"Present_SOLO_SR\": 11...   \n",
       "...         ...    ...                                                ...   \n",
       "1264          S     {}  {\"Present_SOLO_R\": 2, \"Present_SOLO_SR\": 333, ...   \n",
       "2126          S     {}                                                 {}   \n",
       "2131          S     {}                                                 {}   \n",
       "2136          S     {}                                                 {}   \n",
       "2141          S     {}                                                 {}   \n",
       "\n",
       "                                                 OTHER  Unnamed: 0  \n",
       "1178      {\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}         NaN  \n",
       "1179      {\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}         NaN  \n",
       "1180      {\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}         NaN  \n",
       "1181      {\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}         NaN  \n",
       "1182      {\"FINAL_CONFIDENCE_GRADING\": \"1) Assoc w R\"}         NaN  \n",
       "...                                                ...         ...  \n",
       "1264  {\"FINAL_CONFIDENCE_GRADING\": \"5) Not assoc w R\"}         NaN  \n",
       "2126                                                {}         NaN  \n",
       "2131                                                {}         NaN  \n",
       "2136                                                {}         NaN  \n",
       "2141                                                {}         NaN  \n",
       "\n",
       "[89 rows x 13 columns]"
      ]
     },
     "execution_count": 554,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "WHO_data = pd.read_csv('data/NC_000962.3_WHO-UCN-GTB-PCI-2021.7_v1.0_GARC1_RFUS.csv')\n",
    "WHO_data.MUTATION\n",
    "result = ['gyr' in row for row in WHO_data['MUTATION']]\n",
    "WHO_data = WHO_data[result & (WHO_data['PREDICTION'] != 'F') & (WHO_data['PREDICTION'] != 'U')]\n",
    "WHO_data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load MIC (continous, but also binary available) phenotypic data from CRyPTIC table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 555,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>PLATEDESIGN</th>\n",
       "      <th>BELONGS_GPI</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>DILUTION</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>READINGDAY</th>\n",
       "      <th>PRIMARY_DILUTION</th>\n",
       "      <th>PRIMARY_METHOD</th>\n",
       "      <th>AMYGDA_DILUTION</th>\n",
       "      <th>BASHTHEBUG_DILUTION</th>\n",
       "      <th>BASHTHEBUGPRO_DILUTION</th>\n",
       "      <th>PHENOTYPE_DESCRIPTION</th>\n",
       "      <th>BASHTHEBUG_NUMBER_CLASSIFICATIONS</th>\n",
       "      <th>BASHTHEBUGPRO_NUMBER_CLASSIFICATIONS</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_1032.lab.06MIL2037.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0404.lab.06MIL1373.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>21</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.1518.lab.2013102308.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>02</td>\n",
       "      <td>3.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ALL DISAGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;=0.06</td>\n",
       "      <td>-4.06</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_1545.lab.06MIL2824.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>2.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ALL DISAGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0016.lab.2014231005.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>True</td>\n",
       "      <td>02</td>\n",
       "      <td>2.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0547.lab.CO-11838-18.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>05</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.03.subj.GB-90840137.lab.IML-01497.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>03</td>\n",
       "      <td>2.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ALL DISAGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LS-1114.lab.MA-00010-19.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>05</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.08.subj.02TB0334.lab.17491.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>08</td>\n",
       "      <td>2.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,BB AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34092 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG PLATEDESIGN  BELONGS_GPI  \\\n",
       "UNIQUEID                                                                     \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1    MXF      UKMYC6         True   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1    MXF      UKMYC6         True   \n",
       "site.02.subj.1518.lab.2013102308.iso.1        LEV      UKMYC6        False   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1   MXF      UKMYC5         True   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1    MXF      UKMYC6         True   \n",
       "...                                           ...         ...          ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1        LEV      UKMYC5         True   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1    LEV      UKMYC6        False   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1  LEV      UKMYC6        False   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1    LEV      UKMYC6         True   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1         MXF      UKMYC6        False   \n",
       "\n",
       "                                             SITEID  DILUTION  \\\n",
       "UNIQUEID                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1       06       3.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1       06       3.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1           02       3.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1      06       1.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1       06       2.0   \n",
       "...                                             ...       ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1           02       2.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1       05       3.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1     03       2.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1       05       3.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1            08       2.0   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY READINGDAY  \\\n",
       "UNIQUEID                                                                    \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                HIGH         14   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                HIGH         21   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                     LOW         14   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1               HIGH         14   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                 LOW         14   \n",
       "...                                                        ...        ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                    HIGH         14   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                HIGH         14   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1               LOW         14   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                HIGH         14   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                     HIGH         14   \n",
       "\n",
       "                                              PRIMARY_DILUTION PRIMARY_METHOD  \\\n",
       "UNIQUEID                                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                 3.0             VZ   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                 3.0             VZ   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                     3.0             VZ   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                1.0             VZ   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                 2.0             VZ   \n",
       "...                                                        ...            ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                     2.0             VZ   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                 3.0             VZ   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1               2.0             VZ   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                 3.0             VZ   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                      2.0             VZ   \n",
       "\n",
       "                                              AMYGDA_DILUTION  \\\n",
       "UNIQUEID                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                3.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                3.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                    2.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1               1.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                4.0   \n",
       "...                                                       ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                    2.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                3.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1              1.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                3.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                     1.0   \n",
       "\n",
       "                                              BASHTHEBUG_DILUTION  \\\n",
       "UNIQUEID                                                            \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                    8.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                    8.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                        4.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                   8.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                    8.0   \n",
       "...                                                           ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                        2.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                    3.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                  5.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                    4.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                         2.0   \n",
       "\n",
       "                                              BASHTHEBUGPRO_DILUTION  \\\n",
       "UNIQUEID                                                               \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                       NaN   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                       NaN   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                           3.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                      NaN   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                       4.0   \n",
       "...                                                              ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                           NaN   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                       NaN   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                     2.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                       NaN   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                            NaN   \n",
       "\n",
       "                                             PHENOTYPE_DESCRIPTION  \\\n",
       "UNIQUEID                                                             \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1             VZ,IM AGREE   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1             VZ,IM AGREE   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                ALL DISAGREE   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1            VZ,IM AGREE   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1            ALL DISAGREE   \n",
       "...                                                            ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                 VZ,IM AGREE   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1             VZ,IM AGREE   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1          ALL DISAGREE   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1             VZ,IM AGREE   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                  VZ,BB AGREE   \n",
       "\n",
       "                                              BASHTHEBUG_NUMBER_CLASSIFICATIONS  \\\n",
       "UNIQUEID                                                                          \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                                 11.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                                 11.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                                     11.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                                11.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                                 11.0   \n",
       "...                                                                         ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                                     11.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                                 11.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                               11.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                                 11.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                                      11.0   \n",
       "\n",
       "                                              BASHTHEBUGPRO_NUMBER_CLASSIFICATIONS  \\\n",
       "UNIQUEID                                                                             \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                                     NaN   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                                     NaN   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                                         3.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                                    NaN   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                                     3.0   \n",
       "...                                                                            ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                                         NaN   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                                     NaN   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                                   3.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                                     NaN   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                                          NaN   \n",
       "\n",
       "                                                 MIC  LOG2MIC BINARY_PHENOTYPE  \n",
       "UNIQUEID                                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1      0.25    -2.00                S  \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1      0.25    -2.00                S  \n",
       "site.02.subj.1518.lab.2013102308.iso.1           0.5    -1.00                S  \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1   <=0.06    -4.06                S  \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1      0.12    -3.06                S  \n",
       "...                                              ...      ...              ...  \n",
       "site.02.subj.0016.lab.2014231005.iso.1          0.25    -2.00                S  \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1       0.5    -1.00                S  \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1    0.25    -2.00                S  \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1       0.5    -1.00                S  \n",
       "site.08.subj.02TB0334.lab.17491.iso.1           0.12    -3.06                S  \n",
       "\n",
       "[34092 rows x 18 columns]"
      ]
     },
     "execution_count": 555,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PHENOTYPES = pd.read_pickle('data/UKMYC_PHENOTYPES.pkl.gz')\n",
    "PHENOTYPES.reset_index(inplace = True)\n",
    "PHENOTYPES.set_index('UNIQUEID', inplace = True)\n",
    "PHENOTYPES = PHENOTYPES[(PHENOTYPES.DRUG == 'LEV') | (PHENOTYPES.DRUG == 'MXF')]\n",
    "PHENOTYPES"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load metadata from genomes table (lineage and site)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 556,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UNIQUEID\n",
       "site.02.subj.0958.lab.22A197.iso.1                        Lineage 2\n",
       "site.02.subj.0823.lab.2013241494.iso.1                    Lineage 4\n",
       "site.02.subj.0359.lab.222018-14.iso.1                     Lineage 2\n",
       "site.02.subj.0224.lab.2013221088.iso.1                    Lineage 2\n",
       "site.02.subj.0918.lab.22A153.iso.1                        Lineage 2\n",
       "                                                            ...    \n",
       "site.10.subj.YA00023171.lab.YA00023171.iso.1              Lineage 4\n",
       "site.10.subj.SADH00634109_S16.lab.DH00634109_S16.iso.1             \n",
       "site.10.subj.H37RV_SA2437.lab.H37RV_SA2437.iso.1                   \n",
       "site.10.subj.YA00044774.lab.YA00044774.iso.1              Lineage 4\n",
       "site.10.subj.YA00166043.lab.YA00166043.iso.1              Lineage 2\n",
       "Name: LINEAGE_NAME, Length: 77860, dtype: category\n",
       "Categories (7, object): ['', 'Lineage 1', 'Lineage 2', 'Lineage 3', 'Lineage 4', 'Lineage 5', 'Lineage 6']"
      ]
     },
     "execution_count": 556,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "METADATA = pd.read_pickle('data/GENOMES.pkl.gz')\n",
    "METADATA = METADATA['LINEAGE_NAME']\n",
    "METADATA"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load mutation data for gyrA/B genes from CRyPTIC data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 557,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>...</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0122.lab.22A076.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>P232A</td>\n",
       "      <td>232.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ccg</td>\n",
       "      <td>gcg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LR-3059.lab.CR-00865-16.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LR-2342.lab.FN-01704-18.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LI2163509.lab.14722_6_94.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LI2149109.lab.14893_2_65.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00127822.lab.YA00127822.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.PL00350106.lab.PL00350106.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00024313.lab.YA00024313.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00025292.lab.YA00025292.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315G</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>ggc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00038932.lab.YA00038932.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>aca</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>155 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              GENE MUTATION  POSITION  \\\n",
       "UNIQUEID                                                                \n",
       "site.02.subj.0122.lab.22A076.iso.1            katG    P232A     232.0   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1    katG    S315T     315.0   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1    katG    S315T     315.0   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1   katG    S315T     315.0   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1   katG    S315T     315.0   \n",
       "...                                            ...      ...       ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1  katG    S315T     315.0   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1  katG    S315T     315.0   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1  katG    S315T     315.0   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1  katG    S315G     315.0   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1  katG    S315T     315.0   \n",
       "\n",
       "                                              AMINO_ACID_NUMBER  GENOME_INDEX  \\\n",
       "UNIQUEID                                                                        \n",
       "site.02.subj.0122.lab.22A076.iso.1                        232.0           NaN   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1                315.0           NaN   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1                315.0           NaN   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1               315.0           NaN   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1               315.0           NaN   \n",
       "...                                                         ...           ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1              315.0           NaN   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1              315.0           NaN   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1              315.0           NaN   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1              315.0           NaN   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1              315.0           NaN   \n",
       "\n",
       "                                              NUCLEOTIDE_NUMBER  REF  ALT  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.0122.lab.22A076.iso.1                          NaN  ccg  gcg   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1                  NaN  agc  acc   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1                  NaN  agc  acc   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1                 NaN  agc  acc   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1                 NaN  agc  acc   \n",
       "...                                                         ...  ...  ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1                NaN  agc  acc   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1                NaN  agc  acc   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1                NaN  agc  acc   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1                NaN  agc  ggc   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1                NaN  agc  aca   \n",
       "\n",
       "                                              IS_SNP  IS_INDEL  ...  IS_HET  \\\n",
       "UNIQUEID                                                        ...           \n",
       "site.02.subj.0122.lab.22A076.iso.1              True     False  ...   False   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1      True     False  ...   False   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1      True     False  ...   False   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1     True     False  ...   False   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1     True     False  ...   False   \n",
       "...                                              ...       ...  ...     ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1    True     False  ...   False   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1    True     False  ...   False   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1    True     False  ...   False   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1    True     False  ...   False   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1    True     False  ...   False   \n",
       "\n",
       "                                              IS_NULL  IS_FILTER_PASS  \\\n",
       "UNIQUEID                                                                \n",
       "site.02.subj.0122.lab.22A076.iso.1              False            True   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1      False            True   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1      False            True   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1     False            True   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1     False            True   \n",
       "...                                               ...             ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1    False            True   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1    False            True   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1    False            True   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1    False            True   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1    False            True   \n",
       "\n",
       "                                              ELEMENT_TYPE  MUTATION_TYPE  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.0122.lab.22A076.iso.1                    GENE            AAM   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1            GENE            AAM   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1            GENE            AAM   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1           GENE            AAM   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1           GENE            AAM   \n",
       "...                                                    ...            ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1          GENE            AAM   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1          GENE            AAM   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1          GENE            AAM   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1          GENE            AAM   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1          GENE            AAM   \n",
       "\n",
       "                                              INDEL_LENGTH  INDEL_1 INDEL_2  \\\n",
       "UNIQUEID                                                                      \n",
       "site.02.subj.0122.lab.22A076.iso.1                     NaN                    \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1             NaN                    \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1             NaN                    \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1            NaN                    \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1            NaN                    \n",
       "...                                                    ...      ...     ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1           NaN                    \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1           NaN                    \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1           NaN                    \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1           NaN                    \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1           NaN                    \n",
       "\n",
       "                                             SITEID  NUMBER_NUCLEOTIDE_CHANGES  \n",
       "UNIQUEID                                                                        \n",
       "site.02.subj.0122.lab.22A076.iso.1               02                          1  \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1       05                          1  \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1       05                          1  \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1      05                          1  \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1      05                          1  \n",
       "...                                             ...                        ...  \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1     10                          1  \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1     10                          1  \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1     10                          1  \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1     10                          1  \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1     10                          2  \n",
       "\n",
       "[155 rows x 24 columns]"
      ]
     },
     "execution_count": 557,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS = pd.read_pickle('data/MUTATIONS_SAMPLE.pkl.gz')\n",
    "MUTATIONS = MUTATIONS.reset_index()\n",
    "MUTATIONS.set_index('UNIQUEID', inplace = True)\n",
    "\n",
    "#R463L is a known phylogenetic marker for Beijing lineage\n",
    "MUTATIONS = MUTATIONS[(~MUTATIONS.IS_NULL) & (MUTATIONS.IS_FILTER_PASS) & (~MUTATIONS.IS_HET) & (MUTATIONS.IS_SNP) \\\n",
    "                    & (MUTATIONS.IN_CDS) & (MUTATIONS.GENE == 'katG') & (MUTATIONS.IS_NONSYNONYMOUS) & (~(MUTATIONS.MUTATION == 'R463L'))]\n",
    "MUTATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 558,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Out of the 5063 samples with non-lineage mutations in gyrA/B 3974 show only one such mutation per sample\n"
     ]
    }
   ],
   "source": [
    "#read in mutations (pre-filtered for gyrA/B mutations due to size of mutations data file)\n",
    "MUTATIONS = pd.read_csv('data/MUTATIONS-gyrA_B.csv.gz', compression='gzip', header = None)\n",
    "\n",
    "#restore header through shortened mutations file\n",
    "HEADER = pd.read_pickle('data/MUTATIONS_SAMPLE.pkl.gz', compression='gzip')\n",
    "\n",
    "HEADER.reset_index(inplace = True)\n",
    "HEADER = HEADER.columns.to_list()\n",
    "MUTATIONS.columns = HEADER\n",
    "\n",
    "#remove non-synonymous, promoter and mutations with inconclusive sequencing, as well as indels\n",
    "MUTATIONS = MUTATIONS[(~MUTATIONS.IS_SYNONYMOUS) & (~MUTATIONS.IN_PROMOTER) & (MUTATIONS.IS_SNP) & (~MUTATIONS.IS_NULL) & (MUTATIONS.IS_FILTER_PASS)]\n",
    "MUTATIONS['GENE_MUTATION'] = MUTATIONS['GENE'] + '_' + MUTATIONS['MUTATION']\n",
    "\n",
    "#label lineage mutations\n",
    "#ref for lineage mutations: Miotto ERJ 2017\n",
    "LINEAGE_MUTATIONS=['gyrA_S95T','gyrA_E21Q','gyrA_T80A','gyrA_A90G','gyrA_G247S','gyrA_A384V',\n",
    "             'gyrA_G668D','gyrA_L712V', 'gyrA_S250A','gyrA_R252L','gyrA_L398F','gyrA_A463S','gyrA_D639A','gyrA_V742L',\n",
    "              'gyrB_V301L','gyrB_M291I','gyrB_A403S']\n",
    "MUTATIONS['LINEAGE_MUTATION'] = MUTATIONS.GENE_MUTATION.isin(LINEAGE_MUTATIONS)\n",
    "\n",
    "#record how many mutations we see for each sample per UNIQUEID -> we can only perform mutation based approach for samples with < 2 non-lineage mutations in pncA\n",
    "MUTATIONS.set_index('UNIQUEID', inplace = True)\n",
    "\n",
    "number_mutations = MUTATIONS[MUTATIONS.LINEAGE_MUTATION == False].index.value_counts()\n",
    "number_mutations.name = 'MUTATION_COUNT'\n",
    "MUTATIONS = MUTATIONS.join(number_mutations.to_frame())\n",
    "\n",
    "print(\"Out of the\", len(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False)]), \n",
    "      \"samples with non-lineage mutations in gyrA/B\", \n",
    "      len(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) & (MUTATIONS.MUTATION_COUNT < 2)]), \n",
    "          \"show only one such mutation per sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 559,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Good-quality Phenotypic data is available for 2441 of samples that show only one non-lineage mutation per sample in gyrA/B\n"
     ]
    }
   ],
   "source": [
    "#amount of eligible samples where phenotypic data is available\n",
    "print(\"Good-quality Phenotypic data is available for\", len(PHENOTYPES[PHENOTYPES.index.isin(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) \n",
    "                                                                                         & (MUTATIONS.MUTATION_COUNT < 2) ].index.to_list()) \n",
    "                                                                      & (PHENOTYPES.PHENOTYPE_QUALITY!='LOW')].index.unique()),\n",
    "      \"of samples that show only one non-lineage mutation per sample in gyrA/B\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove low quality phenotypes and divide data by drug (LEV, MXF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 560,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BINARY_PHENOTYPE\n",
       "R    1236\n",
       "S     903\n",
       "I       0\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 560,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LEV_PHENOTYPES = PHENOTYPES[(PHENOTYPES.index.isin(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) \n",
    "                                           & (MUTATIONS.MUTATION_COUNT < 2)].index.to_list())) \n",
    "           & (PHENOTYPES.DRUG == 'LEV') & (PHENOTYPES.PHENOTYPE_QUALITY!='LOW')]\n",
    "LEV_PHENOTYPES.BINARY_PHENOTYPE.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 561,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BINARY_PHENOTYPE\n",
       "R    1031\n",
       "S     981\n",
       "I       0\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 561,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_PHENOTYPES = PHENOTYPES[(PHENOTYPES.index.isin(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) \n",
    "                                           & (MUTATIONS.MUTATION_COUNT < 2)].index.to_list())) \n",
    "           & (PHENOTYPES.DRUG == 'MXF') & (PHENOTYPES.PHENOTYPE_QUALITY!='LOW')]\n",
    "MXF_PHENOTYPES.BINARY_PHENOTYPE.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Join the mutations data with the phenotypic data frame for both drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 562,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0027.lab.06MIL0488.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0843.lab.06MIL1790.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94Y</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00624.lab.709302.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>N499D</td>\n",
       "      <td>499.0</td>\n",
       "      <td>gyrB_N499D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00021495.lab.YA00021495.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>N193S</td>\n",
       "      <td>193.0</td>\n",
       "      <td>gyrA_N193S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.XD02435891.lab.XD02435891.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0482.lab.235081-14.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>G512R</td>\n",
       "      <td>512.0</td>\n",
       "      <td>gyrB_G512R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LR-2321.lab.FN-01077-18.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G88C</td>\n",
       "      <td>88.0</td>\n",
       "      <td>gyrA_G88C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0910.lab.22A144.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2171 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG  MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                          \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1    LEV  2.0      1.0   \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1    LEV   >8      4.0   \n",
       "site.04.subj.00624.lab.709302.iso.1           LEV   >8      4.0   \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1  LEV  2.0      1.0   \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1  LEV  0.5     -1.0   \n",
       "...                                           ...  ...      ...   \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1  LEV  8.0      3.0   \n",
       "site.02.subj.0482.lab.235081-14.iso.1         LEV  1.0      0.0   \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1    LEV   >8      4.0   \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1  LEV  8.0      3.0   \n",
       "site.02.subj.0910.lab.22A144.iso.1            LEV  4.0      2.0   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1                  R   \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1                  R   \n",
       "site.04.subj.00624.lab.709302.iso.1                         R   \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1                R   \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1                S   \n",
       "...                                                       ...   \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1                R   \n",
       "site.02.subj.0482.lab.235081-14.iso.1                       S   \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1                  R   \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1                R   \n",
       "site.02.subj.0910.lab.22A144.iso.1                          R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE MUTATION  \\\n",
       "UNIQUEID                                                                        \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1              MEDIUM  gyrA     A90V   \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1                HIGH  gyrA     D94Y   \n",
       "site.04.subj.00624.lab.709302.iso.1                       HIGH  gyrA     D94N   \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1            MEDIUM  gyrB    N499D   \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1              HIGH  gyrA    N193S   \n",
       "...                                                        ...   ...      ...   \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1              HIGH  gyrA     D94N   \n",
       "site.02.subj.0482.lab.235081-14.iso.1                     HIGH  gyrB    G512R   \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1                HIGH  gyrA     G88C   \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1            MEDIUM  gyrA     D94N   \n",
       "site.02.subj.0910.lab.22A144.iso.1                        HIGH  gyrA     A90V   \n",
       "\n",
       "                                              POSITION GENE_MUTATION  \n",
       "UNIQUEID                                                              \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1        90.0     gyrA_A90V  \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1        94.0     gyrA_D94Y  \n",
       "site.04.subj.00624.lab.709302.iso.1               94.0     gyrA_D94N  \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1     499.0    gyrB_N499D  \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1     193.0    gyrA_N193S  \n",
       "...                                                ...           ...  \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1      94.0     gyrA_D94N  \n",
       "site.02.subj.0482.lab.235081-14.iso.1            512.0    gyrB_G512R  \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1        88.0     gyrA_G88C  \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1      94.0     gyrA_D94N  \n",
       "site.02.subj.0910.lab.22A144.iso.1                90.0     gyrA_A90V  \n",
       "\n",
       "[2171 rows x 9 columns]"
      ]
     },
     "execution_count": 562,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS = MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) & (MUTATIONS.MUTATION_COUNT < 2)]\n",
    "\n",
    "LEV_DATASET = pd.merge(LEV_PHENOTYPES, MUTATIONS, how = 'left', on = 'UNIQUEID')\n",
    "#LEV_DATASET = pd.merge(LEV_DATASET, METADATA, how = 'left', on = 'UNIQUEID')\n",
    "LEV_DATASET = LEV_DATASET[['DRUG', 'MIC', 'LOG2MIC', 'BINARY_PHENOTYPE', 'PHENOTYPE_QUALITY', 'GENE', 'MUTATION', 'POSITION', 'GENE_MUTATION']]\n",
    "LEV_DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 563,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00097.lab.633433.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH00935117.lab.UH00935117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00121.lab.633619.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.14.subj.2591.lab.2591.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.01310.lab.724468.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S91P</td>\n",
       "      <td>91.0</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH01302806.lab.UH01302806.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0020.lab.CO-05421-18.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>G594R</td>\n",
       "      <td>594.0</td>\n",
       "      <td>gyrB_G594R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0968.lab.06MIL1963.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2043 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG   MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                           \n",
       "site.04.subj.00097.lab.633433.iso.1           MXF   2.0     1.00   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1  MXF   4.0     2.00   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1  MXF    >4     3.00   \n",
       "site.04.subj.00121.lab.633619.iso.1           MXF   2.0     1.00   \n",
       "site.14.subj.2591.lab.2591.iso.1              MXF    >4     3.00   \n",
       "...                                           ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1           MXF   1.0     0.00   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1  MXF   2.0     1.00   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1  MXF   2.0     1.00   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1    MXF  0.12    -3.06   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1    MXF    >4     3.00   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                         R   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1                R   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1                R   \n",
       "site.04.subj.00121.lab.633619.iso.1                         R   \n",
       "site.14.subj.2591.lab.2591.iso.1                            R   \n",
       "...                                                       ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                         S   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1                R   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1                R   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                  S   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                  R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE MUTATION  \\\n",
       "UNIQUEID                                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                       HIGH  gyrA     D94G   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1              HIGH  gyrA     A90V   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1            MEDIUM  gyrA     D94N   \n",
       "site.04.subj.00121.lab.633619.iso.1                       HIGH  gyrA     D94G   \n",
       "site.14.subj.2591.lab.2591.iso.1                          HIGH  gyrA     D94N   \n",
       "...                                                        ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                       HIGH  gyrA     S91P   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1            MEDIUM  gyrA     A90V   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1              HIGH  gyrA     A90V   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                HIGH  gyrB    G594R   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                HIGH  gyrA     A90V   \n",
       "\n",
       "                                              POSITION GENE_MUTATION  \n",
       "UNIQUEID                                                              \n",
       "site.04.subj.00097.lab.633433.iso.1               94.0     gyrA_D94G  \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1      90.0     gyrA_A90V  \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1      94.0     gyrA_D94N  \n",
       "site.04.subj.00121.lab.633619.iso.1               94.0     gyrA_D94G  \n",
       "site.14.subj.2591.lab.2591.iso.1                  94.0     gyrA_D94N  \n",
       "...                                                ...           ...  \n",
       "site.04.subj.01310.lab.724468.iso.1               91.0     gyrA_S91P  \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1      90.0     gyrA_A90V  \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1      90.0     gyrA_A90V  \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1       594.0    gyrB_G594R  \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1        90.0     gyrA_A90V  \n",
       "\n",
       "[2043 rows x 9 columns]"
      ]
     },
     "execution_count": 563,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET = pd.merge(MXF_PHENOTYPES, MUTATIONS, how = 'left', on = 'UNIQUEID')\n",
    "#MXF_DATASET = pd.merge(MXF_DATASET, METADATA, how = 'left', on = 'UNIQUEID')\n",
    "MXF_DATASET = MXF_DATASET[['DRUG', 'MIC', 'LOG2MIC', 'BINARY_PHENOTYPE', 'PHENOTYPE_QUALITY', 'GENE', 'MUTATION', 'POSITION', 'GENE_MUTATION']]\n",
    "MXF_DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 564,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00097.lab.633433.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH00935117.lab.UH00935117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00121.lab.633619.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.14.subj.2591.lab.2591.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.01310.lab.724468.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S91P</td>\n",
       "      <td>91.0</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH01302806.lab.UH01302806.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0020.lab.CO-05421-18.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>G594R</td>\n",
       "      <td>594.0</td>\n",
       "      <td>gyrB_G594R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0968.lab.06MIL1963.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2043 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG   MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                           \n",
       "site.04.subj.00097.lab.633433.iso.1           MXF   2.0     1.00   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1  MXF   4.0     2.00   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1  MXF    >4     3.00   \n",
       "site.04.subj.00121.lab.633619.iso.1           MXF   2.0     1.00   \n",
       "site.14.subj.2591.lab.2591.iso.1              MXF    >4     3.00   \n",
       "...                                           ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1           MXF   1.0     0.00   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1  MXF   2.0     1.00   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1  MXF   2.0     1.00   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1    MXF  0.12    -3.06   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1    MXF    >4     3.00   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                         R   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1                R   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1                R   \n",
       "site.04.subj.00121.lab.633619.iso.1                         R   \n",
       "site.14.subj.2591.lab.2591.iso.1                            R   \n",
       "...                                                       ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                         S   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1                R   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1                R   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                  S   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                  R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE MUTATION  \\\n",
       "UNIQUEID                                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                       HIGH  gyrA     D94G   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1              HIGH  gyrA     A90V   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1            MEDIUM  gyrA     D94N   \n",
       "site.04.subj.00121.lab.633619.iso.1                       HIGH  gyrA     D94G   \n",
       "site.14.subj.2591.lab.2591.iso.1                          HIGH  gyrA     D94N   \n",
       "...                                                        ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                       HIGH  gyrA     S91P   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1            MEDIUM  gyrA     A90V   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1              HIGH  gyrA     A90V   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                HIGH  gyrB    G594R   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                HIGH  gyrA     A90V   \n",
       "\n",
       "                                              POSITION GENE_MUTATION  \n",
       "UNIQUEID                                                              \n",
       "site.04.subj.00097.lab.633433.iso.1               94.0     gyrA_D94G  \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1      90.0     gyrA_A90V  \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1      94.0     gyrA_D94N  \n",
       "site.04.subj.00121.lab.633619.iso.1               94.0     gyrA_D94G  \n",
       "site.14.subj.2591.lab.2591.iso.1                  94.0     gyrA_D94N  \n",
       "...                                                ...           ...  \n",
       "site.04.subj.01310.lab.724468.iso.1               91.0     gyrA_S91P  \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1      90.0     gyrA_A90V  \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1      90.0     gyrA_A90V  \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1       594.0    gyrB_G594R  \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1        90.0     gyrA_A90V  \n",
       "\n",
       "[2043 rows x 9 columns]"
      ]
     },
     "execution_count": 564,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 565,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_binary_phenotype(row, dataset):\n",
    "        \n",
    "    target_string = row['GENE_MUTATION']\n",
    "    \n",
    "    filtered_df = dataset[dataset['GENE_MUTATION'] == target_string]\n",
    "    \n",
    "    if len(filtered_df) < 2:\n",
    "        \n",
    "        return pd.Series([row['BINARY_PHENOTYPE'], 1.0])\n",
    "    \n",
    "    else:\n",
    "    \n",
    "        most_common_value = filtered_df['BINARY_PHENOTYPE'].mode().iloc[0]\n",
    "        label_support = filtered_df.BINARY_PHENOTYPE.value_counts()[most_common_value]/sum(filtered_df.BINARY_PHENOTYPE.value_counts())\n",
    "\n",
    "        return pd.Series([most_common_value, label_support])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 566,
   "metadata": {},
   "outputs": [],
   "source": [
    "MXF_DATASET[['CLEAN_BINARY_PHENOTYPE', 'LABEL_SUPPORT']] = MXF_DATASET.apply(lambda row: update_binary_phenotype(row, MXF_DATASET), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "metadata": {},
   "outputs": [],
   "source": [
    "LEV_DATASET[['CLEAN_BINARY_PHENOTYPE', 'LABEL_SUPPORT']] = LEV_DATASET.apply(lambda row: update_binary_phenotype(row, LEV_DATASET), axis=1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check which WHO mutations are not in CRyPTIC dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "metadata": {},
   "outputs": [],
   "source": [
    "WHO_list_MXF = WHO_data[WHO_data.DRUG == 'MXF'].MUTATION.unique()\n",
    "WHO_list_LEV = WHO_data[WHO_data.DRUG == 'LEV'].MUTATION.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 569,
   "metadata": {},
   "outputs": [],
   "source": [
    "found_strings_MXF = [string for string in WHO_list_MXF if string not in MXF_DATASET.GENE_MUTATION.unique()]\n",
    "found_strings_LEV = [string for string in WHO_list_LEV if string not in LEV_DATASET.GENE_MUTATION.unique()]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter WHO data based on unseen mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 571,
   "metadata": {},
   "outputs": [],
   "source": [
    "MXF_WHO = WHO_data[(WHO_data.MUTATION.isin(found_strings_MXF)) & (WHO_data.DRUG == 'MXF')]\n",
    "LEV_WHO = WHO_data[(WHO_data.MUTATION.isin(found_strings_LEV)) & (WHO_data.DRUG == 'LEV')]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert WHO data to CRyPTIC table format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 572,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "      <th>CLEAN_BINARY_PHENOTYPE</th>\n",
       "      <th>LABEL_SUPPORT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1178</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1180</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94A</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1184</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94H</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1186</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1188</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94Y</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1190</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_G88A</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G88A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1192</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_G88C</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G88C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1194</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S91P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1196</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_A504V</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>A504V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1198</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_D461N</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>D461N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_E501D</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>E501D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_E501V</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>E501V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_N499D</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>N499D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1206</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_A384V</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A384V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_A463S</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A463S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1208</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_A90G</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1209</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_E21Q</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>E21Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1210</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_G247S</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G247S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1211</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_G668D</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G668D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1212</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_Q613E</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>Q613E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1213</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_R252L</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>R252L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1214</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_S95T</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S95T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1215</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_T80A</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>T80A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1217</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_A403S</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>A403S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1218</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_M291I</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>M291I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1219</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_P94L</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>P94L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1220</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_V301L</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>V301L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1225</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1227</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94A</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1231</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94H</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1233</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1235</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_D94Y</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1237</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_G88A</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G88A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1239</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_G88C</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G88C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1241</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S91P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1243</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_A504V</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>A504V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1245</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_D461N</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>D461N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_E501D</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>E501D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_E501V</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>E501V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1251</th>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrB_N499D</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>N499D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     DRUG GENE_MUTATION CLEAN_BINARY_PHENOTYPE  LABEL_SUPPORT  GENE MUTATION\n",
       "1178  LEV     gyrA_A90V                      R              1  gyrA     A90V\n",
       "1180  LEV     gyrA_D94A                      R              1  gyrA     D94A\n",
       "1182  LEV     gyrA_D94G                      R              1  gyrA     D94G\n",
       "1184  LEV     gyrA_D94H                      R              1  gyrA     D94H\n",
       "1186  LEV     gyrA_D94N                      R              1  gyrA     D94N\n",
       "1188  LEV     gyrA_D94Y                      R              1  gyrA     D94Y\n",
       "1190  LEV     gyrA_G88A                      R              1  gyrA     G88A\n",
       "1192  LEV     gyrA_G88C                      R              1  gyrA     G88C\n",
       "1194  LEV     gyrA_S91P                      R              1  gyrA     S91P\n",
       "1196  LEV    gyrB_A504V                      R              1  gyrB    A504V\n",
       "1198  LEV    gyrB_D461N                      R              1  gyrB    D461N\n",
       "1200  LEV    gyrB_E501D                      R              1  gyrB    E501D\n",
       "1202  LEV    gyrB_E501V                      R              1  gyrB    E501V\n",
       "1204  LEV    gyrB_N499D                      R              1  gyrB    N499D\n",
       "1206  LEV    gyrA_A384V                      S              1  gyrA    A384V\n",
       "1207  LEV    gyrA_A463S                      S              1  gyrA    A463S\n",
       "1208  LEV     gyrA_A90G                      S              1  gyrA     A90G\n",
       "1209  LEV     gyrA_E21Q                      S              1  gyrA     E21Q\n",
       "1210  LEV    gyrA_G247S                      S              1  gyrA    G247S\n",
       "1211  LEV    gyrA_G668D                      S              1  gyrA    G668D\n",
       "1212  LEV    gyrA_Q613E                      S              1  gyrA    Q613E\n",
       "1213  LEV    gyrA_R252L                      S              1  gyrA    R252L\n",
       "1214  LEV     gyrA_S95T                      S              1  gyrA     S95T\n",
       "1215  LEV     gyrA_T80A                      S              1  gyrA     T80A\n",
       "1217  LEV    gyrB_A403S                      S              1  gyrB    A403S\n",
       "1218  LEV    gyrB_M291I                      S              1  gyrB    M291I\n",
       "1219  LEV     gyrB_P94L                      S              1  gyrB     P94L\n",
       "1220  LEV    gyrB_V301L                      S              1  gyrB    V301L\n",
       "1225  LEV     gyrA_A90V                      R              1  gyrA     A90V\n",
       "1227  LEV     gyrA_D94A                      R              1  gyrA     D94A\n",
       "1229  LEV     gyrA_D94G                      R              1  gyrA     D94G\n",
       "1231  LEV     gyrA_D94H                      R              1  gyrA     D94H\n",
       "1233  LEV     gyrA_D94N                      R              1  gyrA     D94N\n",
       "1235  LEV     gyrA_D94Y                      R              1  gyrA     D94Y\n",
       "1237  LEV     gyrA_G88A                      R              1  gyrA     G88A\n",
       "1239  LEV     gyrA_G88C                      R              1  gyrA     G88C\n",
       "1241  LEV     gyrA_S91P                      R              1  gyrA     S91P\n",
       "1243  LEV    gyrB_A504V                      R              1  gyrB    A504V\n",
       "1245  LEV    gyrB_D461N                      R              1  gyrB    D461N\n",
       "1247  LEV    gyrB_E501D                      R              1  gyrB    E501D\n",
       "1249  LEV    gyrB_E501V                      R              1  gyrB    E501V\n",
       "1251  LEV    gyrB_N499D                      R              1  gyrB    N499D"
      ]
     },
     "execution_count": 572,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def replace_symbol(input_string, target_symbol, replacement_symbol):\n",
    "   \n",
    "    modified_string = input_string.replace(target_symbol, replacement_symbol)\n",
    "    \n",
    "    return modified_string\n",
    "\n",
    "def add_gene_mutation(row):\n",
    "\n",
    "    gene = row.GENE_MUTATION[:4]\n",
    "    mutation = row.GENE_MUTATION[5:]\n",
    "\n",
    "    return pd.Series([gene, mutation])\n",
    "\n",
    "MXF_WHO['MUTATION'] = [replace_symbol(s, '@', '_') for s in MXF_WHO.MUTATION]\n",
    "MXF_WHO[['MUTATION', 'PREDICTION']]\n",
    "MXF_WHO.rename(columns = {'MUTATION' : 'GENE_MUTATION', 'PREDICTION' : 'CLEAN_BINARY_PHENOTYPE'}, inplace = True)\n",
    "MXF_WHO = MXF_WHO[['DRUG', 'GENE_MUTATION', 'CLEAN_BINARY_PHENOTYPE']]\n",
    "MXF_WHO['LABEL_SUPPORT'] = 1\n",
    "MXF_WHO[['GENE', 'MUTATION']] = MXF_WHO.apply(lambda row: add_gene_mutation(row), axis = 1)\n",
    "mask = MXF_WHO['MUTATION'].str.contains('c-|=', case=False, regex=True)\n",
    "MXF_WHO = MXF_WHO[~mask]\n",
    "MXF_WHO\n",
    "\n",
    "LEV_WHO['MUTATION'] = [replace_symbol(s, '@', '_') for s in LEV_WHO.MUTATION]\n",
    "LEV_WHO[['MUTATION', 'PREDICTION']]\n",
    "LEV_WHO.rename(columns = {'MUTATION' : 'GENE_MUTATION', 'PREDICTION' : 'CLEAN_BINARY_PHENOTYPE'}, inplace = True)\n",
    "LEV_WHO = LEV_WHO[['DRUG', 'GENE_MUTATION', 'CLEAN_BINARY_PHENOTYPE']]\n",
    "LEV_WHO['LABEL_SUPPORT'] = 1\n",
    "LEV_WHO[['GENE', 'MUTATION']] = LEV_WHO.apply(lambda row: add_gene_mutation(row), axis = 1)\n",
    "mask = LEV_WHO['MUTATION'].str.contains('c-|=', case=False, regex=True)\n",
    "LEV_WHO = LEV_WHO[~mask]\n",
    "LEV_WHO"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Optionally drop rows with low label support"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 573,
   "metadata": {},
   "outputs": [],
   "source": [
    "MXF_DATASET = MXF_DATASET[MXF_DATASET.LABEL_SUPPORT > 0.50]\n",
    "LEV_DATASET = LEV_DATASET[LEV_DATASET.LABEL_SUPPORT > 0.50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 574,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UNIQUEID\n",
       "site.04.subj.00097.lab.633433.iso.1              D94G\n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1     A90V\n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1     D94N\n",
       "site.04.subj.00121.lab.633619.iso.1              D94G\n",
       "site.14.subj.2591.lab.2591.iso.1                 D94N\n",
       "                                                ...  \n",
       "site.04.subj.01310.lab.724468.iso.1              S91P\n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1     A90V\n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1     A90V\n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1      G594R\n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1       A90V\n",
       "Name: MUTATION, Length: 2035, dtype: object"
      ]
     },
     "execution_count": 574,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET['MUTATION']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add segid column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 575,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_segid(row):\n",
    "    return row['GENE'][3:]\n",
    "\n",
    "MXF_DATASET['segid'] = MXF_DATASET.apply(make_segid, axis = 1)\n",
    "MXF_DATASET['mutation'] = MXF_DATASET['MUTATION']\n",
    "MXF_DATASET.drop('MUTATION', axis = 1, inplace = True)\n",
    "\n",
    "LEV_DATASET['segid'] = LEV_DATASET.apply(make_segid, axis = 1)\n",
    "LEV_DATASET['mutation'] = LEV_DATASET['MUTATION']\n",
    "LEV_DATASET.drop('MUTATION', axis = 1, inplace = True)\n",
    "\n",
    "MXF_WHO['segid'] = MXF_WHO.apply(make_segid, axis = 1)\n",
    "MXF_WHO['mutation'] = MXF_WHO['MUTATION']\n",
    "MXF_WHO.drop('MUTATION', axis = 1, inplace = True)\n",
    "\n",
    "LEV_WHO['segid'] = LEV_WHO.apply(make_segid, axis = 1)\n",
    "LEV_WHO['mutation'] = LEV_WHO['MUTATION']\n",
    "LEV_WHO.drop('MUTATION', axis = 1, inplace = True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For now, drop rows with mutations outside of crystallized regions, might impute values later on (segid A: 15-501, segid B: 425-675)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 576,
   "metadata": {},
   "outputs": [],
   "source": [
    "MXF_DATASET = MXF_DATASET[(MXF_DATASET.segid == 'A') & MXF_DATASET.POSITION.astype(int).isin(list(range(15,502))) \n",
    "            | (MXF_DATASET.segid == 'B') & MXF_DATASET.POSITION.astype(int).isin(list(range(425,676)))]\n",
    "\n",
    "LEV_DATASET = LEV_DATASET[(LEV_DATASET.segid == 'A') & LEV_DATASET.POSITION.astype(int).isin(list(range(15,502))) \n",
    "            | (LEV_DATASET.segid == 'B') & LEV_DATASET.POSITION.astype(int).isin(list(range(425,676)))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 577,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "      <th>CLEAN_BINARY_PHENOTYPE</th>\n",
       "      <th>LABEL_SUPPORT</th>\n",
       "      <th>segid</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00097.lab.633433.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>0.818923</td>\n",
       "      <td>A</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH00935117.lab.UH00935117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>R</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>A</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00121.lab.633619.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>0.818923</td>\n",
       "      <td>A</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.14.subj.2591.lab.2591.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>R</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>A</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.01310.lab.724468.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>91.0</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "      <td>R</td>\n",
       "      <td>0.676923</td>\n",
       "      <td>A</td>\n",
       "      <td>S91P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH01302806.lab.UH01302806.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0020.lab.CO-05421-18.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>594.0</td>\n",
       "      <td>gyrB_G594R</td>\n",
       "      <td>S</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B</td>\n",
       "      <td>G594R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0968.lab.06MIL1963.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1665 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG   MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                           \n",
       "site.04.subj.00097.lab.633433.iso.1           MXF   2.0     1.00   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1  MXF   4.0     2.00   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1  MXF    >4     3.00   \n",
       "site.04.subj.00121.lab.633619.iso.1           MXF   2.0     1.00   \n",
       "site.14.subj.2591.lab.2591.iso.1              MXF    >4     3.00   \n",
       "...                                           ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1           MXF   1.0     0.00   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1  MXF   2.0     1.00   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1  MXF   2.0     1.00   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1    MXF  0.12    -3.06   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1    MXF    >4     3.00   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                         R   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1                R   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1                R   \n",
       "site.04.subj.00121.lab.633619.iso.1                         R   \n",
       "site.14.subj.2591.lab.2591.iso.1                            R   \n",
       "...                                                       ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                         S   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1                R   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1                R   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                  S   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                  R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE  \\\n",
       "UNIQUEID                                                               \n",
       "site.04.subj.00097.lab.633433.iso.1                       HIGH  gyrA   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1              HIGH  gyrA   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1            MEDIUM  gyrA   \n",
       "site.04.subj.00121.lab.633619.iso.1                       HIGH  gyrA   \n",
       "site.14.subj.2591.lab.2591.iso.1                          HIGH  gyrA   \n",
       "...                                                        ...   ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                       HIGH  gyrA   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1            MEDIUM  gyrA   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1              HIGH  gyrA   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                HIGH  gyrB   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                HIGH  gyrA   \n",
       "\n",
       "                                              POSITION GENE_MUTATION  \\\n",
       "UNIQUEID                                                               \n",
       "site.04.subj.00097.lab.633433.iso.1               94.0     gyrA_D94G   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1      90.0     gyrA_A90V   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1      94.0     gyrA_D94N   \n",
       "site.04.subj.00121.lab.633619.iso.1               94.0     gyrA_D94G   \n",
       "site.14.subj.2591.lab.2591.iso.1                  94.0     gyrA_D94N   \n",
       "...                                                ...           ...   \n",
       "site.04.subj.01310.lab.724468.iso.1               91.0     gyrA_S91P   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1      90.0     gyrA_A90V   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1      90.0     gyrA_A90V   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1       594.0    gyrB_G594R   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1        90.0     gyrA_A90V   \n",
       "\n",
       "                                             CLEAN_BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                              \n",
       "site.04.subj.00097.lab.633433.iso.1                               R   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1                      R   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1                      R   \n",
       "site.04.subj.00121.lab.633619.iso.1                               R   \n",
       "site.14.subj.2591.lab.2591.iso.1                                  R   \n",
       "...                                                             ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                               R   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1                      R   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1                      R   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                        S   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                        R   \n",
       "\n",
       "                                              LABEL_SUPPORT segid mutation  \n",
       "UNIQUEID                                                                    \n",
       "site.04.subj.00097.lab.633433.iso.1                0.818923     A     D94G  \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1       0.555556     A     A90V  \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1       0.823529     A     D94N  \n",
       "site.04.subj.00121.lab.633619.iso.1                0.818923     A     D94G  \n",
       "site.14.subj.2591.lab.2591.iso.1                   0.823529     A     D94N  \n",
       "...                                                     ...   ...      ...  \n",
       "site.04.subj.01310.lab.724468.iso.1                0.676923     A     S91P  \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1       0.555556     A     A90V  \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1       0.555556     A     A90V  \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1         1.000000     B    G594R  \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1         0.555556     A     A90V  \n",
       "\n",
       "[1665 rows x 12 columns]"
      ]
     },
     "execution_count": 577,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Join CRyPTIC and WHO datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 578,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "      <th>CLEAN_BINARY_PHENOTYPE</th>\n",
       "      <th>LABEL_SUPPORT</th>\n",
       "      <th>segid</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LEV</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>0.869048</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94Y</td>\n",
       "      <td>R</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>A</td>\n",
       "      <td>D94Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>R</td>\n",
       "      <td>0.859259</td>\n",
       "      <td>A</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LEV</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>499.0</td>\n",
       "      <td>gyrB_N499D</td>\n",
       "      <td>R</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B</td>\n",
       "      <td>N499D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>193.0</td>\n",
       "      <td>gyrA_N193S</td>\n",
       "      <td>S</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>A</td>\n",
       "      <td>N193S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1760</th>\n",
       "      <td>LEV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB_A504V</td>\n",
       "      <td>R</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B</td>\n",
       "      <td>A504V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1761</th>\n",
       "      <td>LEV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB_D461N</td>\n",
       "      <td>R</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B</td>\n",
       "      <td>D461N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1762</th>\n",
       "      <td>LEV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB_E501D</td>\n",
       "      <td>R</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B</td>\n",
       "      <td>E501D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1763</th>\n",
       "      <td>LEV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB_E501V</td>\n",
       "      <td>R</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B</td>\n",
       "      <td>E501V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1764</th>\n",
       "      <td>LEV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gyrB_N499D</td>\n",
       "      <td>R</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B</td>\n",
       "      <td>N499D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1765 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     DRUG  MIC  LOG2MIC BINARY_PHENOTYPE PHENOTYPE_QUALITY  GENE  POSITION  \\\n",
       "0     LEV  2.0      1.0                R            MEDIUM  gyrA      90.0   \n",
       "1     LEV   >8      4.0                R              HIGH  gyrA      94.0   \n",
       "2     LEV   >8      4.0                R              HIGH  gyrA      94.0   \n",
       "3     LEV  2.0      1.0                R            MEDIUM  gyrB     499.0   \n",
       "4     LEV  0.5     -1.0                S              HIGH  gyrA     193.0   \n",
       "...   ...  ...      ...              ...               ...   ...       ...   \n",
       "1760  LEV  NaN      NaN              NaN               NaN  gyrB       NaN   \n",
       "1761  LEV  NaN      NaN              NaN               NaN  gyrB       NaN   \n",
       "1762  LEV  NaN      NaN              NaN               NaN  gyrB       NaN   \n",
       "1763  LEV  NaN      NaN              NaN               NaN  gyrB       NaN   \n",
       "1764  LEV  NaN      NaN              NaN               NaN  gyrB       NaN   \n",
       "\n",
       "     GENE_MUTATION CLEAN_BINARY_PHENOTYPE  LABEL_SUPPORT segid mutation  \n",
       "0        gyrA_A90V                      R       0.869048     A     A90V  \n",
       "1        gyrA_D94Y                      R       0.928571     A     D94Y  \n",
       "2        gyrA_D94N                      R       0.859259     A     D94N  \n",
       "3       gyrB_N499D                      R       1.000000     B    N499D  \n",
       "4       gyrA_N193S                      S       1.000000     A    N193S  \n",
       "...            ...                    ...            ...   ...      ...  \n",
       "1760    gyrB_A504V                      R       1.000000     B    A504V  \n",
       "1761    gyrB_D461N                      R       1.000000     B    D461N  \n",
       "1762    gyrB_E501D                      R       1.000000     B    E501D  \n",
       "1763    gyrB_E501V                      R       1.000000     B    E501V  \n",
       "1764    gyrB_N499D                      R       1.000000     B    N499D  \n",
       "\n",
       "[1765 rows x 12 columns]"
      ]
     },
     "execution_count": 578,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET = pd.concat([MXF_DATASET, MXF_WHO], ignore_index=True)\n",
    "LEV_DATASET = pd.concat([LEV_DATASET, LEV_WHO], ignore_index=True)\n",
    "MXF_DATASET\n",
    "LEV_DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 579,
   "metadata": {},
   "outputs": [],
   "source": [
    "import MDAnalysis as mda\n",
    "\n",
    "pdb = mda.Universe('data/pdb/MXF/5bs8.pdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 580,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AtomGroup with 23125 atoms>"
      ]
     },
     "execution_count": 580,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein = pdb.select_atoms(\"protein\")\n",
    "protein\n",
    "#protein.write(\"data/5bs8_atoms.pdb\") #can change this to "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 581,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant</th>\n",
       "      <th>Predicted Effect</th>\n",
       "      <th>Score</th>\n",
       "      <th>Expected Accuracy</th>\n",
       "      <th>segid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T1A</td>\n",
       "      <td>effect</td>\n",
       "      <td>25</td>\n",
       "      <td>63%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T1R</td>\n",
       "      <td>effect</td>\n",
       "      <td>57</td>\n",
       "      <td>75%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T1N</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-80</td>\n",
       "      <td>87%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T1D</td>\n",
       "      <td>effect</td>\n",
       "      <td>21</td>\n",
       "      <td>63%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T1C</td>\n",
       "      <td>effect</td>\n",
       "      <td>13</td>\n",
       "      <td>59%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15115</th>\n",
       "      <td>V253S</td>\n",
       "      <td>effect</td>\n",
       "      <td>52</td>\n",
       "      <td>75%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15116</th>\n",
       "      <td>V253T</td>\n",
       "      <td>effect</td>\n",
       "      <td>30</td>\n",
       "      <td>66%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15117</th>\n",
       "      <td>V253W</td>\n",
       "      <td>effect</td>\n",
       "      <td>66</td>\n",
       "      <td>80%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15118</th>\n",
       "      <td>V253Y</td>\n",
       "      <td>effect</td>\n",
       "      <td>56</td>\n",
       "      <td>75%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15119</th>\n",
       "      <td>V253V</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-99</td>\n",
       "      <td>97%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15120 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Variant Predicted Effect  Score Expected Accuracy segid\n",
       "0         T1A           effect     25               63%     A\n",
       "1         T1R           effect     57               75%     A\n",
       "2         T1N          neutral    -80               87%     A\n",
       "3         T1D           effect     21               63%     A\n",
       "4         T1C           effect     13               59%     A\n",
       "...       ...              ...    ...               ...   ...\n",
       "15115   V253S           effect     52               75%     B\n",
       "15116   V253T           effect     30               66%     B\n",
       "15117   V253W           effect     66               80%     B\n",
       "15118   V253Y           effect     56               75%     B\n",
       "15119   V253V          neutral    -99               97%     B\n",
       "\n",
       "[15120 rows x 5 columns]"
      ]
     },
     "execution_count": 581,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snap2 = pd.read_csv('data/SNAP2_complete.csv')\n",
    "snap2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 582,
   "metadata": {},
   "outputs": [],
   "source": [
    "def correct_position(row):\n",
    "    \n",
    "    if row.segid == 'A':\n",
    "        increase_by = 1\n",
    "    else:\n",
    "        increase_by = 422\n",
    "\n",
    "    letter1 = row.Variant[0]\n",
    "    letter2 = row.Variant[-1]  # Extract the letters (all characters except the last one)\n",
    "    number = int(row.Variant[1:-1])  # Extract the number and convert it to an integer\n",
    "    updated_number = number + increase_by  # Increase the number by the specified amount\n",
    "    updated_string = f'{letter1}{updated_number}{letter2}'  # Create the updated string\n",
    "\n",
    "    return updated_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 583,
   "metadata": {},
   "outputs": [],
   "source": [
    "snap2['Variant'] = snap2.apply(correct_position, axis = 1)\n",
    "snap2.to_csv('data/SNAP2_complete_clean.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create features using sbml-core"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 584,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MXF_df\n",
      "LEV_df\n"
     ]
    }
   ],
   "source": [
    "data_dict = {}\n",
    "names = ['MXF_df', 'LEV_df']\n",
    "i=0\n",
    "\n",
    "for dataset in [MXF_DATASET, LEV_DATASET]:\n",
    "\n",
    "    sbml_dataset = dataset[['segid', 'mutation']]\n",
    "    sbml_dataset.drop_duplicates(subset='mutation', keep='first', inplace=True)\n",
    "\n",
    "    features = sbmlcore.FeatureDataset(sbml_dataset, species='M. tuberculosis', protein = 'DNA Gyrase')\n",
    "\n",
    "    a = sbmlcore.AminoAcidVolumeChange()\n",
    "    b1 = sbmlcore.AminoAcidHydropathyChangeKyteDoolittle()\n",
    "    b2 = sbmlcore.AminoAcidHydropathyChangeWimleyWhite()\n",
    "    c = sbmlcore.AminoAcidMWChange()\n",
    "    d = sbmlcore.AminoAcidPiChange()\n",
    "    e = sbmlcore.Stride('data/pdb/5bs8_atoms.pdb')\n",
    "    f = sbmlcore.FreeSASA('data/pdb/5bs8_atoms.pdb')\n",
    "    g = sbmlcore.SNAP2(\"data/SNAP2_complete_clean.csv\", offsets = {'A': 0, 'B': 0})\n",
    "\n",
    "    print(names[i])\n",
    "    if names[i] == 'MXF_df':\n",
    "\n",
    "        h1 = sbmlcore.StructuralDistances('data/pdb/MXF/5bs8_no_PTR_su_AB_MGB.pdb', distance_selection=\"resname MG\", distance_name='dist_MGB')\n",
    "        h2 = sbmlcore.StructuralDistances('data/pdb/MXF/5bs8_no_PTR_su_AB_MGE.pdb', distance_selection=\"resname MG\", distance_name='dist_MGE')\n",
    "        # h3 = sbmlcore.StructuralDistances('data/pdb/5bs8_no_PTR_su_AB_MGE.pdb', distance_selection=\"resname MG\", distance_name='dist_MGE')\n",
    "        # h4 = sbmlcore.StructuralDistances('data/pdb/5bs8_no_PTR_su_AB_MGF.pdb', distance_selection=\"resname MG\", distance_name='dist_MGF')\n",
    "        k = sbmlcore.StructuralDistances('data/pdb/MXF/5bs8_no_PTR_su_AB_MXF_H.pdb', distance_selection=\"resname MFX\", distance_name='dist_drug')\n",
    "\n",
    "    else:\n",
    "\n",
    "        h1 = sbmlcore.StructuralDistances('data/pdb/LEV/5btg_no_PTR_su_AB_MGB.pdb', distance_selection=\"resname MG\", distance_name='dist_MGB')\n",
    "        h2 = sbmlcore.StructuralDistances('data/pdb/LEV/5btg_no_PTR_su_AB_MGE.pdb', distance_selection=\"resname MG\", distance_name='dist_MGE')\n",
    "        # h3 = sbmlcore.StructuralDistances('data/pdb/5btg_no_PTR_su_AB_MGE.pdb', distance_selection=\"resname MG\", distance_name='dist_MGE')\n",
    "        # h4 = sbmlcore.StructuralDistances('data/pdb/5btg_no_PTR_su_AB_MGF.pdb', distance_selection=\"resname MG\", distance_name='dist_MGF')\n",
    "        k = sbmlcore.StructuralDistances('data/pdb/LEV/5btg_no_PTR_su_AB_LEV_E.pdb', distance_selection=\"resname LFX\", distance_name='dist_drug')\n",
    "\n",
    "    features.add_feature([a, b1, b2, c, d, e, f, g, h1, h2, k])\n",
    "    features = features.df\n",
    "\n",
    "    dataset.reset_index(inplace = True)\n",
    "\n",
    "    variable_name = names[i]\n",
    "    i = i+1\n",
    "    \n",
    "    data_dict[variable_name] = pd.merge(dataset, features, how='left', on=['segid','mutation'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 585,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['index', 'DRUG', 'MIC', 'LOG2MIC', 'BINARY_PHENOTYPE',\n",
       "       'PHENOTYPE_QUALITY', 'GENE', 'POSITION', 'GENE_MUTATION',\n",
       "       'CLEAN_BINARY_PHENOTYPE', 'LABEL_SUPPORT', 'segid', 'mutation',\n",
       "       'd_volume', 'd_hydropathy_KD', 'd_hydropathy_WW', 'd_MW', 'd_Pi',\n",
       "       'secondary_structure', 'secondary_structure_long', 'phi', 'psi',\n",
       "       'residue_sasa', 'n_hbond_acceptors', 'n_hbond_donors', 'B', 'C', 'E',\n",
       "       'G', 'H', 'T', 'SASA', 'snap2_score', 'snap2_accuracy', 'dist_MGB',\n",
       "       'dist_MGE', 'dist_drug'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 585,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = data_dict['MXF_df']\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 586,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.drop_duplicates(subset=['GENE_MUTATION','LINEAGE_NAME'], keep='first', inplace=True)\n",
    "# df.GENE_MUTATION.value_counts()[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 587,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "162"
      ]
     },
     "execution_count": 587,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.GENE_MUTATION.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 588,
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "File does not exist!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[588], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m g \u001b[39m=\u001b[39m sbmlcore\u001b[39m.\u001b[39;49mTempFactors(\u001b[39m'\u001b[39;49m\u001b[39mdata/5bs8.pdb\u001b[39;49m\u001b[39m'\u001b[39;49m)\n\u001b[1;32m      2\u001b[0m g\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/sbmlcore/TempFactors.py:29\u001b[0m, in \u001b[0;36mTempFactors.__init__\u001b[0;34m(self, pdb_file, offsets)\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\u001b[39mself\u001b[39m, pdb_file, offsets\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[1;32m     27\u001b[0m \n\u001b[1;32m     28\u001b[0m     \u001b[39m# check file exists\u001b[39;00m\n\u001b[0;32m---> 29\u001b[0m     \u001b[39massert\u001b[39;00m pathlib\u001b[39m.\u001b[39mPath(pdb_file)\u001b[39m.\u001b[39mis_file(), \u001b[39m\"\u001b[39m\u001b[39mFile does not exist!\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     31\u001b[0m     u \u001b[39m=\u001b[39m MDAnalysis\u001b[39m.\u001b[39mUniverse(pdb_file)\n\u001b[1;32m     33\u001b[0m     \u001b[39m# apply any offsets to the residue numbering\u001b[39;00m\n\u001b[1;32m     34\u001b[0m     \u001b[39m# as specified in the supplied offsets dict e.g. {'A': 3, 'B': -4}\u001b[39;00m\n\u001b[1;32m     35\u001b[0m     \u001b[39m# Chain is segid i.e. A, B, C etc.\u001b[39;00m\n",
      "\u001b[0;31mAssertionError\u001b[0m: File does not exist!"
     ]
    }
   ],
   "source": [
    "g = sbmlcore.TempFactors('data/5bs8.pdb')\n",
    "g"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Further features to be added soon (DeepDG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1765 entries, 0 to 1764\n",
      "Data columns (total 37 columns):\n",
      " #   Column                    Non-Null Count  Dtype   \n",
      "---  ------                    --------------  -----   \n",
      " 0   index                     1765 non-null   int64   \n",
      " 1   DRUG                      1765 non-null   object  \n",
      " 2   MIC                       1710 non-null   category\n",
      " 3   LOG2MIC                   1710 non-null   float64 \n",
      " 4   BINARY_PHENOTYPE          1710 non-null   category\n",
      " 5   PHENOTYPE_QUALITY         1723 non-null   category\n",
      " 6   GENE                      1765 non-null   object  \n",
      " 7   POSITION                  1723 non-null   float64 \n",
      " 8   GENE_MUTATION             1765 non-null   object  \n",
      " 9   CLEAN_BINARY_PHENOTYPE    1764 non-null   object  \n",
      " 10  LABEL_SUPPORT             1765 non-null   float64 \n",
      " 11  segid                     1765 non-null   object  \n",
      " 12  mutation                  1765 non-null   object  \n",
      " 13  d_volume                  1765 non-null   float64 \n",
      " 14  d_hydropathy_KD           1765 non-null   float64 \n",
      " 15  d_hydropathy_WW           1765 non-null   float64 \n",
      " 16  d_MW                      1765 non-null   float64 \n",
      " 17  d_Pi                      1765 non-null   float64 \n",
      " 18  secondary_structure       1750 non-null   object  \n",
      " 19  secondary_structure_long  1750 non-null   object  \n",
      " 20  phi                       1750 non-null   float64 \n",
      " 21  psi                       1750 non-null   float64 \n",
      " 22  residue_sasa              1750 non-null   float64 \n",
      " 23  n_hbond_acceptors         1750 non-null   float64 \n",
      " 24  n_hbond_donors            1750 non-null   float64 \n",
      " 25  B                         1750 non-null   object  \n",
      " 26  C                         1750 non-null   object  \n",
      " 27  E                         1750 non-null   object  \n",
      " 28  G                         1750 non-null   object  \n",
      " 29  H                         1750 non-null   object  \n",
      " 30  T                         1750 non-null   object  \n",
      " 31  SASA                      1765 non-null   float64 \n",
      " 32  snap2_score               1759 non-null   float64 \n",
      " 33  snap2_accuracy            1759 non-null   float64 \n",
      " 34  dist_MGB                  1750 non-null   float64 \n",
      " 35  dist_MGE                  1750 non-null   float64 \n",
      " 36  dist_drug                 1750 non-null   float64 \n",
      "dtypes: category(3), float64(19), int64(1), object(14)\n",
      "memory usage: 475.7+ KB\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1704 entries, 0 to 1703\n",
      "Data columns (total 37 columns):\n",
      " #   Column                    Non-Null Count  Dtype   \n",
      "---  ------                    --------------  -----   \n",
      " 0   index                     1704 non-null   int64   \n",
      " 1   DRUG                      1704 non-null   object  \n",
      " 2   MIC                       1647 non-null   category\n",
      " 3   LOG2MIC                   1647 non-null   float64 \n",
      " 4   BINARY_PHENOTYPE          1647 non-null   category\n",
      " 5   PHENOTYPE_QUALITY         1665 non-null   category\n",
      " 6   GENE                      1704 non-null   object  \n",
      " 7   POSITION                  1665 non-null   float64 \n",
      " 8   GENE_MUTATION             1704 non-null   object  \n",
      " 9   CLEAN_BINARY_PHENOTYPE    1702 non-null   object  \n",
      " 10  LABEL_SUPPORT             1704 non-null   float64 \n",
      " 11  segid                     1704 non-null   object  \n",
      " 12  mutation                  1704 non-null   object  \n",
      " 13  d_volume                  1704 non-null   float64 \n",
      " 14  d_hydropathy_KD           1704 non-null   float64 \n",
      " 15  d_hydropathy_WW           1704 non-null   float64 \n",
      " 16  d_MW                      1704 non-null   float64 \n",
      " 17  d_Pi                      1704 non-null   float64 \n",
      " 18  secondary_structure       1697 non-null   object  \n",
      " 19  secondary_structure_long  1697 non-null   object  \n",
      " 20  phi                       1697 non-null   float64 \n",
      " 21  psi                       1697 non-null   float64 \n",
      " 22  residue_sasa              1697 non-null   float64 \n",
      " 23  n_hbond_acceptors         1697 non-null   float64 \n",
      " 24  n_hbond_donors            1697 non-null   float64 \n",
      " 25  B                         1697 non-null   object  \n",
      " 26  C                         1697 non-null   object  \n",
      " 27  E                         1697 non-null   object  \n",
      " 28  G                         1697 non-null   object  \n",
      " 29  H                         1697 non-null   object  \n",
      " 30  T                         1697 non-null   object  \n",
      " 31  SASA                      1704 non-null   float64 \n",
      " 32  snap2_score               1700 non-null   float64 \n",
      " 33  snap2_accuracy            1700 non-null   float64 \n",
      " 34  dist_MGB                  1697 non-null   float64 \n",
      " 35  dist_MGE                  1697 non-null   float64 \n",
      " 36  dist_drug                 1697 non-null   float64 \n",
      "dtypes: category(3), float64(19), int64(1), object(14)\n",
      "memory usage: 459.3+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "for dataset in ['LEV_df', 'MXF_df']:\n",
    "\n",
    "    df = data_dict[dataset]\n",
    "    print(df.info())\n",
    "    \n",
    "    name = f'data/ML_DATA_{dataset[:3]}.csv'\n",
    "    df.to_csv(name, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 590,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE_MUTATION\n",
       "gyrA_D94G     621\n",
       "gyrA_A90V     338\n",
       "gyrA_D94N     138\n",
       "gyrA_D94A      88\n",
       "gyrA_S91P      67\n",
       "             ... \n",
       "gyrA_T230S      1\n",
       "gyrA_V235A      1\n",
       "gyrB_L638V      1\n",
       "gyrA_S411A      1\n",
       "gyrB_V301L      1\n",
       "Name: count, Length: 162, dtype: int64"
      ]
     },
     "execution_count": 590,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.GENE_MUTATION.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
