{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "\n",
    "import sbmlcore\n",
    "import pytest\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load MIC (continous, but also binary available) phenotypic data from CRyPTIC table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>PLATEDESIGN</th>\n",
       "      <th>BELONGS_GPI</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>DILUTION</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>READINGDAY</th>\n",
       "      <th>PRIMARY_DILUTION</th>\n",
       "      <th>PRIMARY_METHOD</th>\n",
       "      <th>AMYGDA_DILUTION</th>\n",
       "      <th>BASHTHEBUG_DILUTION</th>\n",
       "      <th>BASHTHEBUGPRO_DILUTION</th>\n",
       "      <th>PHENOTYPE_DESCRIPTION</th>\n",
       "      <th>BASHTHEBUG_NUMBER_CLASSIFICATIONS</th>\n",
       "      <th>BASHTHEBUGPRO_NUMBER_CLASSIFICATIONS</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_1032.lab.06MIL2037.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0404.lab.06MIL1373.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>21</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.1518.lab.2013102308.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>02</td>\n",
       "      <td>3.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ALL DISAGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;=0.06</td>\n",
       "      <td>-4.06</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_1545.lab.06MIL2824.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>2.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ALL DISAGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0016.lab.2014231005.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>True</td>\n",
       "      <td>02</td>\n",
       "      <td>2.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0547.lab.CO-11838-18.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>05</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.03.subj.GB-90840137.lab.IML-01497.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>03</td>\n",
       "      <td>2.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ALL DISAGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LS-1114.lab.MA-00010-19.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>05</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.08.subj.02TB0334.lab.17491.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>08</td>\n",
       "      <td>2.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,BB AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34092 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG PLATEDESIGN  BELONGS_GPI  \\\n",
       "UNIQUEID                                                                     \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1    MXF      UKMYC6         True   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1    MXF      UKMYC6         True   \n",
       "site.02.subj.1518.lab.2013102308.iso.1        LEV      UKMYC6        False   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1   MXF      UKMYC5         True   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1    MXF      UKMYC6         True   \n",
       "...                                           ...         ...          ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1        LEV      UKMYC5         True   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1    LEV      UKMYC6        False   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1  LEV      UKMYC6        False   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1    LEV      UKMYC6         True   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1         MXF      UKMYC6        False   \n",
       "\n",
       "                                             SITEID  DILUTION  \\\n",
       "UNIQUEID                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1       06       3.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1       06       3.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1           02       3.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1      06       1.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1       06       2.0   \n",
       "...                                             ...       ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1           02       2.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1       05       3.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1     03       2.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1       05       3.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1            08       2.0   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY READINGDAY  \\\n",
       "UNIQUEID                                                                    \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                HIGH         14   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                HIGH         21   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                     LOW         14   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1               HIGH         14   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                 LOW         14   \n",
       "...                                                        ...        ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                    HIGH         14   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                HIGH         14   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1               LOW         14   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                HIGH         14   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                     HIGH         14   \n",
       "\n",
       "                                              PRIMARY_DILUTION PRIMARY_METHOD  \\\n",
       "UNIQUEID                                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                 3.0             VZ   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                 3.0             VZ   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                     3.0             VZ   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                1.0             VZ   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                 2.0             VZ   \n",
       "...                                                        ...            ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                     2.0             VZ   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                 3.0             VZ   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1               2.0             VZ   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                 3.0             VZ   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                      2.0             VZ   \n",
       "\n",
       "                                              AMYGDA_DILUTION  \\\n",
       "UNIQUEID                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                3.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                3.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                    2.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1               1.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                4.0   \n",
       "...                                                       ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                    2.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                3.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1              1.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                3.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                     1.0   \n",
       "\n",
       "                                              BASHTHEBUG_DILUTION  \\\n",
       "UNIQUEID                                                            \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                    8.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                    8.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                        4.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                   8.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                    8.0   \n",
       "...                                                           ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                        2.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                    3.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                  5.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                    4.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                         2.0   \n",
       "\n",
       "                                              BASHTHEBUGPRO_DILUTION  \\\n",
       "UNIQUEID                                                               \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                       NaN   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                       NaN   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                           3.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                      NaN   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                       4.0   \n",
       "...                                                              ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                           NaN   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                       NaN   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                     2.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                       NaN   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                            NaN   \n",
       "\n",
       "                                             PHENOTYPE_DESCRIPTION  \\\n",
       "UNIQUEID                                                             \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1             VZ,IM AGREE   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1             VZ,IM AGREE   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                ALL DISAGREE   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1            VZ,IM AGREE   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1            ALL DISAGREE   \n",
       "...                                                            ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                 VZ,IM AGREE   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1             VZ,IM AGREE   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1          ALL DISAGREE   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1             VZ,IM AGREE   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                  VZ,BB AGREE   \n",
       "\n",
       "                                              BASHTHEBUG_NUMBER_CLASSIFICATIONS  \\\n",
       "UNIQUEID                                                                          \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                                 11.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                                 11.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                                     11.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                                11.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                                 11.0   \n",
       "...                                                                         ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                                     11.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                                 11.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                               11.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                                 11.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                                      11.0   \n",
       "\n",
       "                                              BASHTHEBUGPRO_NUMBER_CLASSIFICATIONS  \\\n",
       "UNIQUEID                                                                             \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                                     NaN   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                                     NaN   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                                         3.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                                    NaN   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                                     3.0   \n",
       "...                                                                            ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                                         NaN   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                                     NaN   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                                   3.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                                     NaN   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                                          NaN   \n",
       "\n",
       "                                                 MIC  LOG2MIC BINARY_PHENOTYPE  \n",
       "UNIQUEID                                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1      0.25    -2.00                S  \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1      0.25    -2.00                S  \n",
       "site.02.subj.1518.lab.2013102308.iso.1           0.5    -1.00                S  \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1   <=0.06    -4.06                S  \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1      0.12    -3.06                S  \n",
       "...                                              ...      ...              ...  \n",
       "site.02.subj.0016.lab.2014231005.iso.1          0.25    -2.00                S  \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1       0.5    -1.00                S  \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1    0.25    -2.00                S  \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1       0.5    -1.00                S  \n",
       "site.08.subj.02TB0334.lab.17491.iso.1           0.12    -3.06                S  \n",
       "\n",
       "[34092 rows x 18 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PHENOTYPES = pd.read_pickle('data/UKMYC_PHENOTYPES.pkl.gz')\n",
    "PHENOTYPES.reset_index(inplace = True)\n",
    "PHENOTYPES.set_index('UNIQUEID', inplace = True)\n",
    "PHENOTYPES = PHENOTYPES[(PHENOTYPES.DRUG == 'LEV') | (PHENOTYPES.DRUG == 'MXF')]\n",
    "PHENOTYPES"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load mutation data for gyrA/B genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>...</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0122.lab.22A076.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>P232A</td>\n",
       "      <td>232.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ccg</td>\n",
       "      <td>gcg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LR-3059.lab.CR-00865-16.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LR-2342.lab.FN-01704-18.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LI2163509.lab.14722_6_94.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LI2149109.lab.14893_2_65.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00127822.lab.YA00127822.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.PL00350106.lab.PL00350106.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00024313.lab.YA00024313.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00025292.lab.YA00025292.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315G</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>ggc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00038932.lab.YA00038932.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>aca</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>155 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              GENE MUTATION  POSITION  \\\n",
       "UNIQUEID                                                                \n",
       "site.02.subj.0122.lab.22A076.iso.1            katG    P232A     232.0   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1    katG    S315T     315.0   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1    katG    S315T     315.0   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1   katG    S315T     315.0   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1   katG    S315T     315.0   \n",
       "...                                            ...      ...       ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1  katG    S315T     315.0   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1  katG    S315T     315.0   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1  katG    S315T     315.0   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1  katG    S315G     315.0   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1  katG    S315T     315.0   \n",
       "\n",
       "                                              AMINO_ACID_NUMBER  GENOME_INDEX  \\\n",
       "UNIQUEID                                                                        \n",
       "site.02.subj.0122.lab.22A076.iso.1                        232.0           NaN   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1                315.0           NaN   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1                315.0           NaN   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1               315.0           NaN   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1               315.0           NaN   \n",
       "...                                                         ...           ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1              315.0           NaN   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1              315.0           NaN   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1              315.0           NaN   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1              315.0           NaN   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1              315.0           NaN   \n",
       "\n",
       "                                              NUCLEOTIDE_NUMBER  REF  ALT  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.0122.lab.22A076.iso.1                          NaN  ccg  gcg   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1                  NaN  agc  acc   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1                  NaN  agc  acc   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1                 NaN  agc  acc   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1                 NaN  agc  acc   \n",
       "...                                                         ...  ...  ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1                NaN  agc  acc   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1                NaN  agc  acc   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1                NaN  agc  acc   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1                NaN  agc  ggc   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1                NaN  agc  aca   \n",
       "\n",
       "                                              IS_SNP  IS_INDEL  ...  IS_HET  \\\n",
       "UNIQUEID                                                        ...           \n",
       "site.02.subj.0122.lab.22A076.iso.1              True     False  ...   False   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1      True     False  ...   False   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1      True     False  ...   False   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1     True     False  ...   False   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1     True     False  ...   False   \n",
       "...                                              ...       ...  ...     ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1    True     False  ...   False   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1    True     False  ...   False   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1    True     False  ...   False   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1    True     False  ...   False   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1    True     False  ...   False   \n",
       "\n",
       "                                              IS_NULL  IS_FILTER_PASS  \\\n",
       "UNIQUEID                                                                \n",
       "site.02.subj.0122.lab.22A076.iso.1              False            True   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1      False            True   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1      False            True   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1     False            True   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1     False            True   \n",
       "...                                               ...             ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1    False            True   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1    False            True   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1    False            True   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1    False            True   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1    False            True   \n",
       "\n",
       "                                              ELEMENT_TYPE  MUTATION_TYPE  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.0122.lab.22A076.iso.1                    GENE            AAM   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1            GENE            AAM   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1            GENE            AAM   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1           GENE            AAM   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1           GENE            AAM   \n",
       "...                                                    ...            ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1          GENE            AAM   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1          GENE            AAM   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1          GENE            AAM   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1          GENE            AAM   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1          GENE            AAM   \n",
       "\n",
       "                                              INDEL_LENGTH  INDEL_1 INDEL_2  \\\n",
       "UNIQUEID                                                                      \n",
       "site.02.subj.0122.lab.22A076.iso.1                     NaN                    \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1             NaN                    \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1             NaN                    \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1            NaN                    \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1            NaN                    \n",
       "...                                                    ...      ...     ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1           NaN                    \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1           NaN                    \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1           NaN                    \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1           NaN                    \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1           NaN                    \n",
       "\n",
       "                                             SITEID  NUMBER_NUCLEOTIDE_CHANGES  \n",
       "UNIQUEID                                                                        \n",
       "site.02.subj.0122.lab.22A076.iso.1               02                          1  \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1       05                          1  \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1       05                          1  \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1      05                          1  \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1      05                          1  \n",
       "...                                             ...                        ...  \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1     10                          1  \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1     10                          1  \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1     10                          1  \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1     10                          1  \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1     10                          2  \n",
       "\n",
       "[155 rows x 24 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS = pd.read_pickle('data/MUTATIONS_SAMPLE.pkl.gz')\n",
    "MUTATIONS = MUTATIONS.reset_index()\n",
    "MUTATIONS.set_index('UNIQUEID', inplace = True)\n",
    "\n",
    "#R463L is a known phylogenetic marker for Beijing lineage\n",
    "MUTATIONS = MUTATIONS[(~MUTATIONS.IS_NULL) & (MUTATIONS.IS_FILTER_PASS) & (~MUTATIONS.IS_HET) & (MUTATIONS.IS_SNP) \\\n",
    "                    & (MUTATIONS.IN_CDS) & (MUTATIONS.GENE == 'katG') & (MUTATIONS.IS_NONSYNONYMOUS) & (~(MUTATIONS.MUTATION == 'R463L'))]\n",
    "MUTATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Out of the 5063 samples with non-lineage mutations in gyrA/B 3974 show only one such mutation per sample\n"
     ]
    }
   ],
   "source": [
    "#read in mutations (pre-filtered for gyrA/B mutations due to size of mutations data file)\n",
    "MUTATIONS = pd.read_csv('data/MUTATIONS-gyrA_B.csv.gz', compression='gzip', header = None)\n",
    "\n",
    "#restore header through shortened mutations file\n",
    "HEADER = pd.read_pickle('data/MUTATIONS_SAMPLE.pkl.gz', compression='gzip')\n",
    "\n",
    "HEADER.reset_index(inplace = True)\n",
    "HEADER = HEADER.columns.to_list()\n",
    "MUTATIONS.columns = HEADER\n",
    "\n",
    "#remove non-synonymous, promoter and mutations with inconclusive sequencing, as well as indels\n",
    "MUTATIONS = MUTATIONS[(~MUTATIONS.IS_SYNONYMOUS) & (~MUTATIONS.IN_PROMOTER) & (MUTATIONS.IS_SNP) & (~MUTATIONS.IS_NULL) & (MUTATIONS.IS_FILTER_PASS)]\n",
    "MUTATIONS['GENE_MUTATION'] = MUTATIONS['GENE'] + '_' + MUTATIONS['MUTATION']\n",
    "\n",
    "#label lineage mutations\n",
    "#ref for lineage mutations: Miotto ERJ 2017\n",
    "LINEAGE_MUTATIONS=['gyrA_S95T','gyrA_E21Q','gyrA_T80A','gyrA_A90G','gyrA_G247S','gyrA_A384V',\n",
    "             'gyrA_G668D','gyrA_L712V', 'gyrA_S250A','gyrA_R252L','gyrA_L398F','gyrA_A463S','gyrA_D639A','gyrA_V742L',\n",
    "              'gyrB_V301L','gyrB_M291I','gyrB_A403S']\n",
    "MUTATIONS['LINEAGE_MUTATION'] = MUTATIONS.GENE_MUTATION.isin(LINEAGE_MUTATIONS)\n",
    "\n",
    "#record how many mutations we see for each sample per UNIQUEID -> we can only perform mutation based approach for samples with < 2 non-lineage mutations in pncA\n",
    "MUTATIONS.set_index('UNIQUEID', inplace = True)\n",
    "\n",
    "number_mutations = MUTATIONS[MUTATIONS.LINEAGE_MUTATION == False].index.value_counts()\n",
    "number_mutations.name = 'MUTATION_COUNT'\n",
    "MUTATIONS = MUTATIONS.join(number_mutations.to_frame())\n",
    "\n",
    "print(\"Out of the\", len(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False)]), \n",
    "      \"samples with non-lineage mutations in gyrA/B\", \n",
    "      len(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) & (MUTATIONS.MUTATION_COUNT < 2)]), \n",
    "          \"show only one such mutation per sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Good-quality Phenotypic data is available for 2441 of samples that show only one non-lineage mutation per sample in gyrA/B\n"
     ]
    }
   ],
   "source": [
    "#amount of eligible samples where phenotypic data is available\n",
    "print(\"Good-quality Phenotypic data is available for\", len(PHENOTYPES[PHENOTYPES.index.isin(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) \n",
    "                                                                                         & (MUTATIONS.MUTATION_COUNT < 2) ].index.to_list()) \n",
    "                                                                      & (PHENOTYPES.PHENOTYPE_QUALITY!='LOW')].index.unique()),\n",
    "      \"of samples that show only one non-lineage mutation per sample in gyrA/B\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove low quality phenotypes and divide data by drug (LEV, MXF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BINARY_PHENOTYPE\n",
       "R    1236\n",
       "S     903\n",
       "I       0\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LEV_PHENOTYPES = PHENOTYPES[(PHENOTYPES.index.isin(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) \n",
    "                                           & (MUTATIONS.MUTATION_COUNT < 2)].index.to_list())) \n",
    "           & (PHENOTYPES.DRUG == 'LEV') & (PHENOTYPES.PHENOTYPE_QUALITY!='LOW')]\n",
    "LEV_PHENOTYPES.BINARY_PHENOTYPE.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BINARY_PHENOTYPE\n",
       "R    1031\n",
       "S     981\n",
       "I       0\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_PHENOTYPES = PHENOTYPES[(PHENOTYPES.index.isin(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) \n",
    "                                           & (MUTATIONS.MUTATION_COUNT < 2)].index.to_list())) \n",
    "           & (PHENOTYPES.DRUG == 'MXF') & (PHENOTYPES.PHENOTYPE_QUALITY!='LOW')]\n",
    "MXF_PHENOTYPES.BINARY_PHENOTYPE.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Join the mutations data with the phenotypic data frame for both drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0027.lab.06MIL0488.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0843.lab.06MIL1790.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94Y</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00624.lab.709302.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>N499D</td>\n",
       "      <td>499.0</td>\n",
       "      <td>gyrB_N499D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00021495.lab.YA00021495.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>N193S</td>\n",
       "      <td>193.0</td>\n",
       "      <td>gyrA_N193S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.XD02435891.lab.XD02435891.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0482.lab.235081-14.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>G512R</td>\n",
       "      <td>512.0</td>\n",
       "      <td>gyrB_G512R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LR-2321.lab.FN-01077-18.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G88C</td>\n",
       "      <td>88.0</td>\n",
       "      <td>gyrA_G88C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0910.lab.22A144.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2171 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG  MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                          \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1    LEV  2.0      1.0   \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1    LEV   >8      4.0   \n",
       "site.04.subj.00624.lab.709302.iso.1           LEV   >8      4.0   \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1  LEV  2.0      1.0   \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1  LEV  0.5     -1.0   \n",
       "...                                           ...  ...      ...   \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1  LEV  8.0      3.0   \n",
       "site.02.subj.0482.lab.235081-14.iso.1         LEV  1.0      0.0   \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1    LEV   >8      4.0   \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1  LEV  8.0      3.0   \n",
       "site.02.subj.0910.lab.22A144.iso.1            LEV  4.0      2.0   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1                  R   \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1                  R   \n",
       "site.04.subj.00624.lab.709302.iso.1                         R   \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1                R   \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1                S   \n",
       "...                                                       ...   \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1                R   \n",
       "site.02.subj.0482.lab.235081-14.iso.1                       S   \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1                  R   \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1                R   \n",
       "site.02.subj.0910.lab.22A144.iso.1                          R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE MUTATION  \\\n",
       "UNIQUEID                                                                        \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1              MEDIUM  gyrA     A90V   \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1                HIGH  gyrA     D94Y   \n",
       "site.04.subj.00624.lab.709302.iso.1                       HIGH  gyrA     D94N   \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1            MEDIUM  gyrB    N499D   \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1              HIGH  gyrA    N193S   \n",
       "...                                                        ...   ...      ...   \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1              HIGH  gyrA     D94N   \n",
       "site.02.subj.0482.lab.235081-14.iso.1                     HIGH  gyrB    G512R   \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1                HIGH  gyrA     G88C   \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1            MEDIUM  gyrA     D94N   \n",
       "site.02.subj.0910.lab.22A144.iso.1                        HIGH  gyrA     A90V   \n",
       "\n",
       "                                              POSITION GENE_MUTATION  \n",
       "UNIQUEID                                                              \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1        90.0     gyrA_A90V  \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1        94.0     gyrA_D94Y  \n",
       "site.04.subj.00624.lab.709302.iso.1               94.0     gyrA_D94N  \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1     499.0    gyrB_N499D  \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1     193.0    gyrA_N193S  \n",
       "...                                                ...           ...  \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1      94.0     gyrA_D94N  \n",
       "site.02.subj.0482.lab.235081-14.iso.1            512.0    gyrB_G512R  \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1        88.0     gyrA_G88C  \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1      94.0     gyrA_D94N  \n",
       "site.02.subj.0910.lab.22A144.iso.1                90.0     gyrA_A90V  \n",
       "\n",
       "[2171 rows x 9 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS = MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) & (MUTATIONS.MUTATION_COUNT < 2)]\n",
    "\n",
    "LEV_DATASET = pd.merge(LEV_PHENOTYPES, MUTATIONS, how = 'left', on = 'UNIQUEID')\n",
    "LEV_DATASET = LEV_DATASET[['DRUG', 'MIC', 'LOG2MIC', 'BINARY_PHENOTYPE', 'PHENOTYPE_QUALITY', 'GENE', 'MUTATION', 'POSITION', 'GENE_MUTATION']]\n",
    "LEV_DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00097.lab.633433.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH00935117.lab.UH00935117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00121.lab.633619.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.14.subj.2591.lab.2591.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.01310.lab.724468.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S91P</td>\n",
       "      <td>91.0</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH01302806.lab.UH01302806.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0020.lab.CO-05421-18.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>G594R</td>\n",
       "      <td>594.0</td>\n",
       "      <td>gyrB_G594R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0968.lab.06MIL1963.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2043 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG   MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                           \n",
       "site.04.subj.00097.lab.633433.iso.1           MXF   2.0     1.00   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1  MXF   4.0     2.00   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1  MXF    >4     3.00   \n",
       "site.04.subj.00121.lab.633619.iso.1           MXF   2.0     1.00   \n",
       "site.14.subj.2591.lab.2591.iso.1              MXF    >4     3.00   \n",
       "...                                           ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1           MXF   1.0     0.00   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1  MXF   2.0     1.00   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1  MXF   2.0     1.00   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1    MXF  0.12    -3.06   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1    MXF    >4     3.00   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                         R   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1                R   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1                R   \n",
       "site.04.subj.00121.lab.633619.iso.1                         R   \n",
       "site.14.subj.2591.lab.2591.iso.1                            R   \n",
       "...                                                       ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                         S   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1                R   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1                R   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                  S   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                  R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE MUTATION  \\\n",
       "UNIQUEID                                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                       HIGH  gyrA     D94G   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1              HIGH  gyrA     A90V   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1            MEDIUM  gyrA     D94N   \n",
       "site.04.subj.00121.lab.633619.iso.1                       HIGH  gyrA     D94G   \n",
       "site.14.subj.2591.lab.2591.iso.1                          HIGH  gyrA     D94N   \n",
       "...                                                        ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                       HIGH  gyrA     S91P   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1            MEDIUM  gyrA     A90V   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1              HIGH  gyrA     A90V   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                HIGH  gyrB    G594R   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                HIGH  gyrA     A90V   \n",
       "\n",
       "                                              POSITION GENE_MUTATION  \n",
       "UNIQUEID                                                              \n",
       "site.04.subj.00097.lab.633433.iso.1               94.0     gyrA_D94G  \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1      90.0     gyrA_A90V  \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1      94.0     gyrA_D94N  \n",
       "site.04.subj.00121.lab.633619.iso.1               94.0     gyrA_D94G  \n",
       "site.14.subj.2591.lab.2591.iso.1                  94.0     gyrA_D94N  \n",
       "...                                                ...           ...  \n",
       "site.04.subj.01310.lab.724468.iso.1               91.0     gyrA_S91P  \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1      90.0     gyrA_A90V  \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1      90.0     gyrA_A90V  \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1       594.0    gyrB_G594R  \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1        90.0     gyrA_A90V  \n",
       "\n",
       "[2043 rows x 9 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET = pd.merge(MXF_PHENOTYPES, MUTATIONS, how = 'left', on = 'UNIQUEID')\n",
    "MXF_DATASET = MXF_DATASET[['DRUG', 'MIC', 'LOG2MIC', 'BINARY_PHENOTYPE', 'PHENOTYPE_QUALITY', 'GENE', 'MUTATION', 'POSITION', 'GENE_MUTATION']]\n",
    "MXF_DATASET"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create features using sbml core"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_segid(row):\n",
    "    return row['GENE'][3:]\n",
    "\n",
    "MXF_DATASET['segid'] = MXF_DATASET.apply(make_segid, axis = 1)\n",
    "MXF_DATASET['mutation'] = MXF_DATASET['MUTATION']\n",
    "MXF_DATASET.drop('MUTATION', axis = 1, inplace = True)\n",
    "\n",
    "LEV_DATASET['segid'] = LEV_DATASET.apply(make_segid, axis = 1)\n",
    "LEV_DATASET['mutation'] = LEV_DATASET['MUTATION']\n",
    "LEV_DATASET.drop('MUTATION', axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "      <th>segid</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00097.lab.633433.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>A</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH00935117.lab.UH00935117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>A</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00121.lab.633619.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>A</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.14.subj.2591.lab.2591.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>A</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.01310.lab.724468.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>91.0</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "      <td>A</td>\n",
       "      <td>S91P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH01302806.lab.UH01302806.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0020.lab.CO-05421-18.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>594.0</td>\n",
       "      <td>gyrB_G594R</td>\n",
       "      <td>B</td>\n",
       "      <td>G594R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0968.lab.06MIL1963.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2043 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG   MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                           \n",
       "site.04.subj.00097.lab.633433.iso.1           MXF   2.0     1.00   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1  MXF   4.0     2.00   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1  MXF    >4     3.00   \n",
       "site.04.subj.00121.lab.633619.iso.1           MXF   2.0     1.00   \n",
       "site.14.subj.2591.lab.2591.iso.1              MXF    >4     3.00   \n",
       "...                                           ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1           MXF   1.0     0.00   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1  MXF   2.0     1.00   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1  MXF   2.0     1.00   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1    MXF  0.12    -3.06   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1    MXF    >4     3.00   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                         R   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1                R   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1                R   \n",
       "site.04.subj.00121.lab.633619.iso.1                         R   \n",
       "site.14.subj.2591.lab.2591.iso.1                            R   \n",
       "...                                                       ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                         S   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1                R   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1                R   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                  S   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                  R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE  \\\n",
       "UNIQUEID                                                               \n",
       "site.04.subj.00097.lab.633433.iso.1                       HIGH  gyrA   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1              HIGH  gyrA   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1            MEDIUM  gyrA   \n",
       "site.04.subj.00121.lab.633619.iso.1                       HIGH  gyrA   \n",
       "site.14.subj.2591.lab.2591.iso.1                          HIGH  gyrA   \n",
       "...                                                        ...   ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                       HIGH  gyrA   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1            MEDIUM  gyrA   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1              HIGH  gyrA   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                HIGH  gyrB   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                HIGH  gyrA   \n",
       "\n",
       "                                              POSITION GENE_MUTATION segid  \\\n",
       "UNIQUEID                                                                     \n",
       "site.04.subj.00097.lab.633433.iso.1               94.0     gyrA_D94G     A   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1      90.0     gyrA_A90V     A   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1      94.0     gyrA_D94N     A   \n",
       "site.04.subj.00121.lab.633619.iso.1               94.0     gyrA_D94G     A   \n",
       "site.14.subj.2591.lab.2591.iso.1                  94.0     gyrA_D94N     A   \n",
       "...                                                ...           ...   ...   \n",
       "site.04.subj.01310.lab.724468.iso.1               91.0     gyrA_S91P     A   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1      90.0     gyrA_A90V     A   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1      90.0     gyrA_A90V     A   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1       594.0    gyrB_G594R     B   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1        90.0     gyrA_A90V     A   \n",
       "\n",
       "                                             mutation  \n",
       "UNIQUEID                                               \n",
       "site.04.subj.00097.lab.633433.iso.1              D94G  \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1     A90V  \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1     D94N  \n",
       "site.04.subj.00121.lab.633619.iso.1              D94G  \n",
       "site.14.subj.2591.lab.2591.iso.1                 D94N  \n",
       "...                                               ...  \n",
       "site.04.subj.01310.lab.724468.iso.1              S91P  \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1     A90V  \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1     A90V  \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1      G594R  \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1       A90V  \n",
       "\n",
       "[2043 rows x 10 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For now, drop rows with mutations outside of crystallized regions, might impute values later on (segid A: 15-501, segid B: 425-675)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "MXF_DATASET = MXF_DATASET[(MXF_DATASET.segid == 'A') & MXF_DATASET.POSITION.astype(int).isin(list(range(15,502))) \n",
    "            | (MXF_DATASET.segid == 'B') & MXF_DATASET.POSITION.astype(int).isin(list(range(425,676)))]\n",
    "\n",
    "LEV_DATASET = LEV_DATASET[(LEV_DATASET.segid == 'A') & LEV_DATASET.POSITION.astype(int).isin(list(range(15,502))) \n",
    "            | (LEV_DATASET.segid == 'B') & LEV_DATASET.POSITION.astype(int).isin(list(range(425,676)))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "import MDAnalysis as mda\n",
    "\n",
    "pdb = mda.Universe('data/5bs8.pdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AtomGroup with 23125 atoms>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein = pdb.select_atoms(\"protein\")\n",
    "protein\n",
    "#protein.write(\"data/5bs8_atoms.pdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant</th>\n",
       "      <th>Predicted Effect</th>\n",
       "      <th>Score</th>\n",
       "      <th>Expected Accuracy</th>\n",
       "      <th>segid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T1A</td>\n",
       "      <td>effect</td>\n",
       "      <td>25</td>\n",
       "      <td>63%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T1R</td>\n",
       "      <td>effect</td>\n",
       "      <td>57</td>\n",
       "      <td>75%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T1N</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-80</td>\n",
       "      <td>87%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T1D</td>\n",
       "      <td>effect</td>\n",
       "      <td>21</td>\n",
       "      <td>63%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T1C</td>\n",
       "      <td>effect</td>\n",
       "      <td>13</td>\n",
       "      <td>59%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15115</th>\n",
       "      <td>V253S</td>\n",
       "      <td>effect</td>\n",
       "      <td>52</td>\n",
       "      <td>75%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15116</th>\n",
       "      <td>V253T</td>\n",
       "      <td>effect</td>\n",
       "      <td>30</td>\n",
       "      <td>66%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15117</th>\n",
       "      <td>V253W</td>\n",
       "      <td>effect</td>\n",
       "      <td>66</td>\n",
       "      <td>80%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15118</th>\n",
       "      <td>V253Y</td>\n",
       "      <td>effect</td>\n",
       "      <td>56</td>\n",
       "      <td>75%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15119</th>\n",
       "      <td>V253V</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-99</td>\n",
       "      <td>97%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15120 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Variant Predicted Effect  Score Expected Accuracy segid\n",
       "0         T1A           effect     25               63%     A\n",
       "1         T1R           effect     57               75%     A\n",
       "2         T1N          neutral    -80               87%     A\n",
       "3         T1D           effect     21               63%     A\n",
       "4         T1C           effect     13               59%     A\n",
       "...       ...              ...    ...               ...   ...\n",
       "15115   V253S           effect     52               75%     B\n",
       "15116   V253T           effect     30               66%     B\n",
       "15117   V253W           effect     66               80%     B\n",
       "15118   V253Y           effect     56               75%     B\n",
       "15119   V253V          neutral    -99               97%     B\n",
       "\n",
       "[15120 rows x 5 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snap2 = pd.read_csv('data/5bs8_SNAP2_complete.csv')\n",
    "snap2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def correct_position(row):\n",
    "    \n",
    "    if row.segid == 'A':\n",
    "        increase_by = 1\n",
    "    else:\n",
    "        increase_by = 422\n",
    "\n",
    "    letter1 = row.Variant[0]\n",
    "    letter2 = row.Variant[-1]  # Extract the letters (all characters except the last one)\n",
    "    number = int(row.Variant[1:-1])  # Extract the number and convert it to an integer\n",
    "    updated_number = number + increase_by  # Increase the number by the specified amount\n",
    "    updated_string = f'{letter1}{updated_number}{letter2}'  # Create the updated string\n",
    "\n",
    "    return updated_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "snap2['Variant'] = snap2.apply(correct_position, axis = 1)\n",
    "snap2.to_csv('data/5bs8_SNAP2_complete_clean.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dict = {}\n",
    "names = ['MXF_df', 'LEV_df']\n",
    "i=0\n",
    "\n",
    "for dataset in [MXF_DATASET, LEV_DATASET]:\n",
    "\n",
    "    sbml_dataset = dataset[['segid', 'mutation']]\n",
    "    sbml_dataset.drop_duplicates(subset='mutation', keep='first', inplace=True)\n",
    "\n",
    "    features = sbmlcore.FeatureDataset(sbml_dataset, species='M. tuberculosis', protein = 'DNA Gyrase')\n",
    "\n",
    "    a = sbmlcore.AminoAcidVolumeChange()\n",
    "    b1 = sbmlcore.AminoAcidHydropathyChangeKyteDoolittle()\n",
    "    b2 = sbmlcore.AminoAcidHydropathyChangeWimleyWhite()\n",
    "    c = sbmlcore.AminoAcidMWChange()\n",
    "    d = sbmlcore.AminoAcidPiChange()\n",
    "    e = sbmlcore.Stride('data/5bs8_atoms.pdb')\n",
    "    f = sbmlcore.FreeSASA('data/5bs8_atoms.pdb')\n",
    "    g = sbmlcore.SNAP2(\"data/5bs8_SNAP2_complete_clean.csv\", offsets = {'A': 0, 'B': 0})\n",
    "\n",
    "    features.add_feature([a, b1, b2, c, d, e, f, g])\n",
    "    features = features.df\n",
    "\n",
    "    dataset.reset_index(inplace = True)\n",
    "\n",
    "    variable_name = names[i]\n",
    "    i = i+1\n",
    "    \n",
    "    data_dict[variable_name] = pd.merge(dataset, features, how='left', on=['segid','mutation'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['UNIQUEID', 'DRUG', 'MIC', 'LOG2MIC', 'BINARY_PHENOTYPE',\n",
       "       'PHENOTYPE_QUALITY', 'GENE', 'POSITION', 'GENE_MUTATION', 'segid',\n",
       "       'mutation', 'd_volume', 'd_hydropathy_KD', 'd_hydropathy_WW', 'd_MW',\n",
       "       'd_Pi', 'secondary_structure', 'secondary_structure_long', 'phi', 'psi',\n",
       "       'residue_sasa', 'n_hbond_acceptors', 'n_hbond_donors', 'B', 'C', 'E',\n",
       "       'G', 'H', 'T', 'SASA', 'snap2_score', 'snap2_accuracy'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = data_dict['LEV_df']\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'PTR'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[62], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m g \u001b[39m=\u001b[39m sbmlcore\u001b[39m.\u001b[39;49mTempFactors(\u001b[39m'\u001b[39;49m\u001b[39mdata/5bs8.pdb\u001b[39;49m\u001b[39m'\u001b[39;49m)\n\u001b[1;32m      2\u001b[0m g\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/sbmlcore/TempFactors.py:53\u001b[0m, in \u001b[0;36mTempFactors.__init__\u001b[0;34m(self, pdb_file, offsets)\u001b[0m\n\u001b[1;32m     50\u001b[0m     \u001b[39mreturn\u001b[39;00m(sbmlcore\u001b[39m.\u001b[39mamino_acid_3to1letter[row\u001b[39m.\u001b[39mresname])\n\u001b[1;32m     52\u001b[0m results \u001b[39m=\u001b[39m pandas\u001b[39m.\u001b[39mDataFrame(Ca_data)\n\u001b[0;32m---> 53\u001b[0m results[\u001b[39m'\u001b[39m\u001b[39mamino_acid\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m results\u001b[39m.\u001b[39;49mapply(one_letter, axis\u001b[39m=\u001b[39;49m\u001b[39m1\u001b[39;49m)\n\u001b[1;32m     55\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mresults \u001b[39m=\u001b[39m results\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/pandas/core/frame.py:9423\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[0;34m(self, func, axis, raw, result_type, args, **kwargs)\u001b[0m\n\u001b[1;32m   9412\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpandas\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcore\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mapply\u001b[39;00m \u001b[39mimport\u001b[39;00m frame_apply\n\u001b[1;32m   9414\u001b[0m op \u001b[39m=\u001b[39m frame_apply(\n\u001b[1;32m   9415\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m   9416\u001b[0m     func\u001b[39m=\u001b[39mfunc,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   9421\u001b[0m     kwargs\u001b[39m=\u001b[39mkwargs,\n\u001b[1;32m   9422\u001b[0m )\n\u001b[0;32m-> 9423\u001b[0m \u001b[39mreturn\u001b[39;00m op\u001b[39m.\u001b[39;49mapply()\u001b[39m.\u001b[39m__finalize__(\u001b[39mself\u001b[39m, method\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mapply\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/pandas/core/apply.py:678\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    675\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraw:\n\u001b[1;32m    676\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapply_raw()\n\u001b[0;32m--> 678\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_standard()\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/pandas/core/apply.py:798\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    797\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mapply_standard\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 798\u001b[0m     results, res_index \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_series_generator()\n\u001b[1;32m    800\u001b[0m     \u001b[39m# wrap results\u001b[39;00m\n\u001b[1;32m    801\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwrap_results(results, res_index)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/pandas/core/apply.py:814\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    811\u001b[0m \u001b[39mwith\u001b[39;00m option_context(\u001b[39m\"\u001b[39m\u001b[39mmode.chained_assignment\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m):\n\u001b[1;32m    812\u001b[0m     \u001b[39mfor\u001b[39;00m i, v \u001b[39min\u001b[39;00m \u001b[39menumerate\u001b[39m(series_gen):\n\u001b[1;32m    813\u001b[0m         \u001b[39m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[0;32m--> 814\u001b[0m         results[i] \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mf(v)\n\u001b[1;32m    815\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[1;32m    816\u001b[0m             \u001b[39m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[1;32m    817\u001b[0m             \u001b[39m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[1;32m    818\u001b[0m             results[i] \u001b[39m=\u001b[39m results[i]\u001b[39m.\u001b[39mcopy(deep\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/sbmlcore/TempFactors.py:50\u001b[0m, in \u001b[0;36mTempFactors.__init__.<locals>.one_letter\u001b[0;34m(row)\u001b[0m\n\u001b[1;32m     49\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mone_letter\u001b[39m(row):\n\u001b[0;32m---> 50\u001b[0m     \u001b[39mreturn\u001b[39;00m(sbmlcore\u001b[39m.\u001b[39;49mamino_acid_3to1letter[row\u001b[39m.\u001b[39;49mresname])\n",
      "\u001b[0;31mKeyError\u001b[0m: 'PTR'"
     ]
    }
   ],
   "source": [
    "g = sbmlcore.TempFactors('data/5bs8.pdb')\n",
    "g"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Further features to be added soon (SNAP2, DeepDG, StructuralDistances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "AtomGroup is empty.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[63], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m i \u001b[39m=\u001b[39m sbmlcore\u001b[39m.\u001b[39;49mStructuralDistances(\u001b[39m'\u001b[39;49m\u001b[39mdata/5bs8.pdb\u001b[39;49m\u001b[39m'\u001b[39;49m, distance_selection\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mresname FE2\u001b[39;49m\u001b[39m\"\u001b[39;49m, distance_name\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mdist_FE\u001b[39;49m\u001b[39m'\u001b[39;49m)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/sbmlcore/StructuralDistances.py:42\u001b[0m, in \u001b[0;36mStructuralDistances.__init__\u001b[0;34m(self, pdb_file, distance_selection, distance_name, infer_masses, offsets)\u001b[0m\n\u001b[1;32m     40\u001b[0m \u001b[39m# prefer to calculate the centre of mass, but the masses cannot always be inferred from the PDB file\u001b[39;00m\n\u001b[1;32m     41\u001b[0m \u001b[39mif\u001b[39;00m infer_masses:\n\u001b[0;32m---> 42\u001b[0m     reference_com \u001b[39m=\u001b[39m u\u001b[39m.\u001b[39;49mselect_atoms(distance_selection)\u001b[39m.\u001b[39;49mcenter_of_mass()\n\u001b[1;32m     43\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m     44\u001b[0m     reference_com \u001b[39m=\u001b[39m u\u001b[39m.\u001b[39mselect_atoms(distance_selection)\u001b[39m.\u001b[39mcenter_of_geometry()\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:1914\u001b[0m, in \u001b[0;36mwarn_if_not_unique.<locals>.wrapper\u001b[0;34m(group, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1909\u001b[0m \u001b[39m@wraps\u001b[39m(groupmethod)\n\u001b[1;32m   1910\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mwrapper\u001b[39m(group, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m   1911\u001b[0m     \u001b[39m# Proceed as usual if the calling group is unique or a DuplicateWarning\u001b[39;00m\n\u001b[1;32m   1912\u001b[0m     \u001b[39m# has already been thrown:\u001b[39;00m\n\u001b[1;32m   1913\u001b[0m     \u001b[39mif\u001b[39;00m group\u001b[39m.\u001b[39misunique \u001b[39mor\u001b[39;00m warn_if_not_unique\u001b[39m.\u001b[39mwarned:\n\u001b[0;32m-> 1914\u001b[0m         \u001b[39mreturn\u001b[39;00m groupmethod(group, \u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m   1915\u001b[0m     \u001b[39m# Otherwise, throw a DuplicateWarning and execute the method.\u001b[39;00m\n\u001b[1;32m   1916\u001b[0m     method_name \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m.\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mjoin(\n\u001b[1;32m   1917\u001b[0m         (group\u001b[39m.\u001b[39m\u001b[39m__class__\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__name__\u001b[39m, groupmethod\u001b[39m.\u001b[39m\u001b[39m__name__\u001b[39m))\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/groups.py:451\u001b[0m, in \u001b[0;36m_pbc_to_wrap.<locals>.wrapped\u001b[0;34m(group, *args, **kwargs)\u001b[0m\n\u001b[1;32m    445\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\u001b[39m\"\u001b[39m\u001b[39mThe \u001b[39m\u001b[39m'\u001b[39m\u001b[39mpbc\u001b[39m\u001b[39m'\u001b[39m\u001b[39m kwarg has been deprecated and will be \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    446\u001b[0m                   \u001b[39m\"\u001b[39m\u001b[39mremoved in version 3.0., \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    447\u001b[0m                   \u001b[39m\"\u001b[39m\u001b[39mplease use \u001b[39m\u001b[39m'\u001b[39m\u001b[39mwrap\u001b[39m\u001b[39m'\u001b[39m\u001b[39m instead\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    448\u001b[0m                   \u001b[39mDeprecationWarning\u001b[39;00m)\n\u001b[1;32m    449\u001b[0m     kwargs[\u001b[39m'\u001b[39m\u001b[39mwrap\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m kwargs\u001b[39m.\u001b[39mpop(\u001b[39m'\u001b[39m\u001b[39mpbc\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m--> 451\u001b[0m \u001b[39mreturn\u001b[39;00m function(group, \u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/groups.py:461\u001b[0m, in \u001b[0;36mcheck_wrap_and_unwrap.<locals>.wrapped\u001b[0;34m(group, *args, **kwargs)\u001b[0m\n\u001b[1;32m    459\u001b[0m \u001b[39mif\u001b[39;00m kwargs\u001b[39m.\u001b[39mget(\u001b[39m'\u001b[39m\u001b[39mwrap\u001b[39m\u001b[39m'\u001b[39m) \u001b[39mand\u001b[39;00m kwargs\u001b[39m.\u001b[39mget(\u001b[39m'\u001b[39m\u001b[39munwrap\u001b[39m\u001b[39m'\u001b[39m):\n\u001b[1;32m    460\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mboth \u001b[39m\u001b[39m'\u001b[39m\u001b[39mwrap\u001b[39m\u001b[39m'\u001b[39m\u001b[39m and \u001b[39m\u001b[39m'\u001b[39m\u001b[39munwrap\u001b[39m\u001b[39m'\u001b[39m\u001b[39m can not be set to true\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 461\u001b[0m \u001b[39mreturn\u001b[39;00m function(group, \u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:2209\u001b[0m, in \u001b[0;36mcheck_atomgroup_not_empty.<locals>.wrapper\u001b[0;34m(group, *args, **kwargs)\u001b[0m\n\u001b[1;32m   2205\u001b[0m \u001b[39m@wraps\u001b[39m(groupmethod)\n\u001b[1;32m   2206\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mwrapper\u001b[39m(group, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m   2207\u001b[0m     \u001b[39m# Throw error if the group is empty.\u001b[39;00m\n\u001b[1;32m   2208\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m group\u001b[39m.\u001b[39matoms:\n\u001b[0;32m-> 2209\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mAtomGroup is empty.\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m   2210\u001b[0m     \u001b[39m# Proceed as usual if the calling group is not empty.\u001b[39;00m\n\u001b[1;32m   2211\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   2212\u001b[0m         result \u001b[39m=\u001b[39m groupmethod(group, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n",
      "\u001b[0;31mValueError\u001b[0m: AtomGroup is empty."
     ]
    }
   ],
   "source": [
    "i = sbmlcore.StructuralDistances('data/5bs8.pdb', distance_selection=\"resname FE2\", distance_name='dist_FE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1735 entries, 0 to 1734\n",
      "Data columns (total 32 columns):\n",
      " #   Column                    Non-Null Count  Dtype   \n",
      "---  ------                    --------------  -----   \n",
      " 0   UNIQUEID                  1735 non-null   object  \n",
      " 1   DRUG                      1735 non-null   category\n",
      " 2   MIC                       1722 non-null   category\n",
      " 3   LOG2MIC                   1722 non-null   float64 \n",
      " 4   BINARY_PHENOTYPE          1722 non-null   category\n",
      " 5   PHENOTYPE_QUALITY         1735 non-null   category\n",
      " 6   GENE                      1735 non-null   object  \n",
      " 7   POSITION                  1735 non-null   float64 \n",
      " 8   GENE_MUTATION             1735 non-null   object  \n",
      " 9   segid                     1735 non-null   object  \n",
      " 10  mutation                  1735 non-null   object  \n",
      " 11  d_volume                  1735 non-null   float64 \n",
      " 12  d_hydropathy_KD           1735 non-null   float64 \n",
      " 13  d_hydropathy_WW           1735 non-null   float64 \n",
      " 14  d_MW                      1735 non-null   float64 \n",
      " 15  d_Pi                      1735 non-null   float64 \n",
      " 16  secondary_structure       1726 non-null   object  \n",
      " 17  secondary_structure_long  1726 non-null   object  \n",
      " 18  phi                       1726 non-null   float64 \n",
      " 19  psi                       1726 non-null   float64 \n",
      " 20  residue_sasa              1726 non-null   float64 \n",
      " 21  n_hbond_acceptors         1726 non-null   float64 \n",
      " 22  n_hbond_donors            1726 non-null   float64 \n",
      " 23  B                         1726 non-null   object  \n",
      " 24  C                         1726 non-null   object  \n",
      " 25  E                         1726 non-null   object  \n",
      " 26  G                         1726 non-null   object  \n",
      " 27  H                         1726 non-null   object  \n",
      " 28  T                         1726 non-null   object  \n",
      " 29  SASA                      1735 non-null   float64 \n",
      " 30  snap2_score               1735 non-null   int64   \n",
      " 31  snap2_accuracy            1735 non-null   int64   \n",
      "dtypes: category(4), float64(13), int64(2), object(13)\n",
      "memory usage: 388.7+ KB\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1669 entries, 0 to 1668\n",
      "Data columns (total 32 columns):\n",
      " #   Column                    Non-Null Count  Dtype   \n",
      "---  ------                    --------------  -----   \n",
      " 0   UNIQUEID                  1669 non-null   object  \n",
      " 1   DRUG                      1669 non-null   category\n",
      " 2   MIC                       1651 non-null   category\n",
      " 3   LOG2MIC                   1651 non-null   float64 \n",
      " 4   BINARY_PHENOTYPE          1651 non-null   category\n",
      " 5   PHENOTYPE_QUALITY         1669 non-null   category\n",
      " 6   GENE                      1669 non-null   object  \n",
      " 7   POSITION                  1669 non-null   float64 \n",
      " 8   GENE_MUTATION             1669 non-null   object  \n",
      " 9   segid                     1669 non-null   object  \n",
      " 10  mutation                  1669 non-null   object  \n",
      " 11  d_volume                  1669 non-null   float64 \n",
      " 12  d_hydropathy_KD           1669 non-null   float64 \n",
      " 13  d_hydropathy_WW           1669 non-null   float64 \n",
      " 14  d_MW                      1669 non-null   float64 \n",
      " 15  d_Pi                      1669 non-null   float64 \n",
      " 16  secondary_structure       1666 non-null   object  \n",
      " 17  secondary_structure_long  1666 non-null   object  \n",
      " 18  phi                       1666 non-null   float64 \n",
      " 19  psi                       1666 non-null   float64 \n",
      " 20  residue_sasa              1666 non-null   float64 \n",
      " 21  n_hbond_acceptors         1666 non-null   float64 \n",
      " 22  n_hbond_donors            1666 non-null   float64 \n",
      " 23  B                         1666 non-null   object  \n",
      " 24  C                         1666 non-null   object  \n",
      " 25  E                         1666 non-null   object  \n",
      " 26  G                         1666 non-null   object  \n",
      " 27  H                         1666 non-null   object  \n",
      " 28  T                         1666 non-null   object  \n",
      " 29  SASA                      1669 non-null   float64 \n",
      " 30  snap2_score               1669 non-null   int64   \n",
      " 31  snap2_accuracy            1669 non-null   int64   \n",
      "dtypes: category(4), float64(13), int64(2), object(13)\n",
      "memory usage: 374.0+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "for dataset in ['LEV_df', 'MXF_df']:\n",
    "\n",
    "    df = data_dict[dataset]\n",
    "    print(df.info())\n",
    "    \n",
    "    name = f'data/ML_DATA_{dataset[:3]}.csv'\n",
    "    df.to_csv(name, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "      <th>segid</th>\n",
       "      <th>...</th>\n",
       "      <th>n_hbond_donors</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>E</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>T</th>\n",
       "      <th>SASA</th>\n",
       "      <th>snap2_score</th>\n",
       "      <th>snap2_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.04.subj.00097.lab.633433.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>88.155089</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.10.subj.UH00935117.lab.UH00935117.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55.555816</td>\n",
       "      <td>49.0</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>88.155089</td>\n",
       "      <td>-8.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.04.subj.00121.lab.633619.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>88.155089</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>site.14.subj.2591.lab.2591.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>88.155089</td>\n",
       "      <td>-8.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664</th>\n",
       "      <td>site.04.subj.01310.lab.724468.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>91.0</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>65.228155</td>\n",
       "      <td>68.0</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665</th>\n",
       "      <td>site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55.555816</td>\n",
       "      <td>49.0</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1666</th>\n",
       "      <td>site.10.subj.UH01302806.lab.UH01302806.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55.555816</td>\n",
       "      <td>49.0</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1667</th>\n",
       "      <td>site.05.subj.CA-0020.lab.CO-05421-18.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>594.0</td>\n",
       "      <td>gyrB_G594R</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>2.551348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1668</th>\n",
       "      <td>site.06.subj.06TB_0968.lab.06MIL1963.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55.555816</td>\n",
       "      <td>49.0</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1669 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          UNIQUEID DRUG   MIC  LOG2MIC  \\\n",
       "0              site.04.subj.00097.lab.633433.iso.1  MXF   2.0     1.00   \n",
       "1     site.10.subj.UH00935117.lab.UH00935117.iso.1  MXF   4.0     2.00   \n",
       "2     site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1  MXF    >4     3.00   \n",
       "3              site.04.subj.00121.lab.633619.iso.1  MXF   2.0     1.00   \n",
       "4                 site.14.subj.2591.lab.2591.iso.1  MXF    >4     3.00   \n",
       "...                                            ...  ...   ...      ...   \n",
       "1664           site.04.subj.01310.lab.724468.iso.1  MXF   1.0     0.00   \n",
       "1665  site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1  MXF   2.0     1.00   \n",
       "1666  site.10.subj.UH01302806.lab.UH01302806.iso.1  MXF   2.0     1.00   \n",
       "1667    site.05.subj.CA-0020.lab.CO-05421-18.iso.1  MXF  0.12    -3.06   \n",
       "1668    site.06.subj.06TB_0968.lab.06MIL1963.iso.1  MXF    >4     3.00   \n",
       "\n",
       "     BINARY_PHENOTYPE PHENOTYPE_QUALITY  GENE  POSITION GENE_MUTATION segid  \\\n",
       "0                   R              HIGH  gyrA      94.0     gyrA_D94G     A   \n",
       "1                   R              HIGH  gyrA      90.0     gyrA_A90V     A   \n",
       "2                   R            MEDIUM  gyrA      94.0     gyrA_D94N     A   \n",
       "3                   R              HIGH  gyrA      94.0     gyrA_D94G     A   \n",
       "4                   R              HIGH  gyrA      94.0     gyrA_D94N     A   \n",
       "...               ...               ...   ...       ...           ...   ...   \n",
       "1664                S              HIGH  gyrA      91.0     gyrA_S91P     A   \n",
       "1665                R            MEDIUM  gyrA      90.0     gyrA_A90V     A   \n",
       "1666                R              HIGH  gyrA      90.0     gyrA_A90V     A   \n",
       "1667                S              HIGH  gyrB     594.0    gyrB_G594R     B   \n",
       "1668                R              HIGH  gyrA      90.0     gyrA_A90V     A   \n",
       "\n",
       "      ... n_hbond_donors      B      C      E      G     H      T       SASA  \\\n",
       "0     ...            1.0  False  False  False  False  True  False  88.155089   \n",
       "1     ...            0.0  False  False  False  False  True  False  55.555816   \n",
       "2     ...            1.0  False  False  False  False  True  False  88.155089   \n",
       "3     ...            1.0  False  False  False  False  True  False  88.155089   \n",
       "4     ...            1.0  False  False  False  False  True  False  88.155089   \n",
       "...   ...            ...    ...    ...    ...    ...   ...    ...        ...   \n",
       "1664  ...            0.0  False  False  False  False  True  False  65.228155   \n",
       "1665  ...            0.0  False  False  False  False  True  False  55.555816   \n",
       "1666  ...            0.0  False  False  False  False  True  False  55.555816   \n",
       "1667  ...            1.0  False  False  False  False  True  False   2.551348   \n",
       "1668  ...            0.0  False  False  False  False  True  False  55.555816   \n",
       "\n",
       "      snap2_score  snap2_accuracy  \n",
       "0            -7.0            53.0  \n",
       "1            49.0            71.0  \n",
       "2            -8.0            53.0  \n",
       "3            -7.0            53.0  \n",
       "4            -8.0            53.0  \n",
       "...           ...             ...  \n",
       "1664         68.0            80.0  \n",
       "1665         49.0            71.0  \n",
       "1666         49.0            71.0  \n",
       "1667          NaN             NaN  \n",
       "1668         49.0            71.0  \n",
       "\n",
       "[1669 rows x 32 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
