{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "\n",
    "import sbmlcore\n",
    "import pytest\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load MIC (continous, but also binary available) phenotypic data from CRyPTIC table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>PLATEDESIGN</th>\n",
       "      <th>BELONGS_GPI</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>DILUTION</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>READINGDAY</th>\n",
       "      <th>PRIMARY_DILUTION</th>\n",
       "      <th>PRIMARY_METHOD</th>\n",
       "      <th>AMYGDA_DILUTION</th>\n",
       "      <th>BASHTHEBUG_DILUTION</th>\n",
       "      <th>BASHTHEBUGPRO_DILUTION</th>\n",
       "      <th>PHENOTYPE_DESCRIPTION</th>\n",
       "      <th>BASHTHEBUG_NUMBER_CLASSIFICATIONS</th>\n",
       "      <th>BASHTHEBUGPRO_NUMBER_CLASSIFICATIONS</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_1032.lab.06MIL2037.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0404.lab.06MIL1373.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>21</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.1518.lab.2013102308.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>02</td>\n",
       "      <td>3.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ALL DISAGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;=0.06</td>\n",
       "      <td>-4.06</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_1545.lab.06MIL2824.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>06</td>\n",
       "      <td>2.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ALL DISAGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0016.lab.2014231005.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC5</td>\n",
       "      <td>True</td>\n",
       "      <td>02</td>\n",
       "      <td>2.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0547.lab.CO-11838-18.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>05</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.03.subj.GB-90840137.lab.IML-01497.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>03</td>\n",
       "      <td>2.0</td>\n",
       "      <td>LOW</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ALL DISAGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LS-1114.lab.MA-00010-19.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>True</td>\n",
       "      <td>05</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,IM AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.08.subj.02TB0334.lab.17491.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC6</td>\n",
       "      <td>False</td>\n",
       "      <td>08</td>\n",
       "      <td>2.0</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>VZ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VZ,BB AGREE</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34092 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG PLATEDESIGN  BELONGS_GPI  \\\n",
       "UNIQUEID                                                                     \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1    MXF      UKMYC6         True   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1    MXF      UKMYC6         True   \n",
       "site.02.subj.1518.lab.2013102308.iso.1        LEV      UKMYC6        False   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1   MXF      UKMYC5         True   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1    MXF      UKMYC6         True   \n",
       "...                                           ...         ...          ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1        LEV      UKMYC5         True   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1    LEV      UKMYC6        False   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1  LEV      UKMYC6        False   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1    LEV      UKMYC6         True   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1         MXF      UKMYC6        False   \n",
       "\n",
       "                                             SITEID  DILUTION  \\\n",
       "UNIQUEID                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1       06       3.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1       06       3.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1           02       3.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1      06       1.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1       06       2.0   \n",
       "...                                             ...       ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1           02       2.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1       05       3.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1     03       2.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1       05       3.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1            08       2.0   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY READINGDAY  \\\n",
       "UNIQUEID                                                                    \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                HIGH         14   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                HIGH         21   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                     LOW         14   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1               HIGH         14   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                 LOW         14   \n",
       "...                                                        ...        ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                    HIGH         14   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                HIGH         14   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1               LOW         14   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                HIGH         14   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                     HIGH         14   \n",
       "\n",
       "                                              PRIMARY_DILUTION PRIMARY_METHOD  \\\n",
       "UNIQUEID                                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                 3.0             VZ   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                 3.0             VZ   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                     3.0             VZ   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                1.0             VZ   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                 2.0             VZ   \n",
       "...                                                        ...            ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                     2.0             VZ   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                 3.0             VZ   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1               2.0             VZ   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                 3.0             VZ   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                      2.0             VZ   \n",
       "\n",
       "                                              AMYGDA_DILUTION  \\\n",
       "UNIQUEID                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                3.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                3.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                    2.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1               1.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                4.0   \n",
       "...                                                       ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                    2.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                3.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1              1.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                3.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                     1.0   \n",
       "\n",
       "                                              BASHTHEBUG_DILUTION  \\\n",
       "UNIQUEID                                                            \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                    8.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                    8.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                        4.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                   8.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                    8.0   \n",
       "...                                                           ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                        2.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                    3.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                  5.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                    4.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                         2.0   \n",
       "\n",
       "                                              BASHTHEBUGPRO_DILUTION  \\\n",
       "UNIQUEID                                                               \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                       NaN   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                       NaN   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                           3.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                      NaN   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                       4.0   \n",
       "...                                                              ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                           NaN   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                       NaN   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                     2.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                       NaN   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                            NaN   \n",
       "\n",
       "                                             PHENOTYPE_DESCRIPTION  \\\n",
       "UNIQUEID                                                             \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1             VZ,IM AGREE   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1             VZ,IM AGREE   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                ALL DISAGREE   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1            VZ,IM AGREE   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1            ALL DISAGREE   \n",
       "...                                                            ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                 VZ,IM AGREE   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1             VZ,IM AGREE   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1          ALL DISAGREE   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1             VZ,IM AGREE   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                  VZ,BB AGREE   \n",
       "\n",
       "                                              BASHTHEBUG_NUMBER_CLASSIFICATIONS  \\\n",
       "UNIQUEID                                                                          \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                                 11.0   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                                 11.0   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                                     11.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                                11.0   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                                 11.0   \n",
       "...                                                                         ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                                     11.0   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                                 11.0   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                               11.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                                 11.0   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                                      11.0   \n",
       "\n",
       "                                              BASHTHEBUGPRO_NUMBER_CLASSIFICATIONS  \\\n",
       "UNIQUEID                                                                             \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1                                     NaN   \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1                                     NaN   \n",
       "site.02.subj.1518.lab.2013102308.iso.1                                         3.0   \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1                                    NaN   \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1                                     3.0   \n",
       "...                                                                            ...   \n",
       "site.02.subj.0016.lab.2014231005.iso.1                                         NaN   \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1                                     NaN   \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1                                   3.0   \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1                                     NaN   \n",
       "site.08.subj.02TB0334.lab.17491.iso.1                                          NaN   \n",
       "\n",
       "                                                 MIC  LOG2MIC BINARY_PHENOTYPE  \n",
       "UNIQUEID                                                                        \n",
       "site.06.subj.06TB_1032.lab.06MIL2037.iso.1      0.25    -2.00                S  \n",
       "site.06.subj.06TB_0404.lab.06MIL1373.iso.1      0.25    -2.00                S  \n",
       "site.02.subj.1518.lab.2013102308.iso.1           0.5    -1.00                S  \n",
       "site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1   <=0.06    -4.06                S  \n",
       "site.06.subj.06TB_1545.lab.06MIL2824.iso.1      0.12    -3.06                S  \n",
       "...                                              ...      ...              ...  \n",
       "site.02.subj.0016.lab.2014231005.iso.1          0.25    -2.00                S  \n",
       "site.05.subj.CA-0547.lab.CO-11838-18.iso.1       0.5    -1.00                S  \n",
       "site.03.subj.GB-90840137.lab.IML-01497.iso.1    0.25    -2.00                S  \n",
       "site.05.subj.LS-1114.lab.MA-00010-19.iso.1       0.5    -1.00                S  \n",
       "site.08.subj.02TB0334.lab.17491.iso.1           0.12    -3.06                S  \n",
       "\n",
       "[34092 rows x 18 columns]"
      ]
     },
     "execution_count": 339,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PHENOTYPES = pd.read_pickle('data/UKMYC_PHENOTYPES.pkl.gz')\n",
    "PHENOTYPES.reset_index(inplace = True)\n",
    "PHENOTYPES.set_index('UNIQUEID', inplace = True)\n",
    "PHENOTYPES = PHENOTYPES[(PHENOTYPES.DRUG == 'LEV') | (PHENOTYPES.DRUG == 'MXF')]\n",
    "PHENOTYPES"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load mutation data for gyrA/B genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>GENOME_INDEX</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IS_SNP</th>\n",
       "      <th>IS_INDEL</th>\n",
       "      <th>...</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>IS_NULL</th>\n",
       "      <th>IS_FILTER_PASS</th>\n",
       "      <th>ELEMENT_TYPE</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_1</th>\n",
       "      <th>INDEL_2</th>\n",
       "      <th>SITEID</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0122.lab.22A076.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>P232A</td>\n",
       "      <td>232.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ccg</td>\n",
       "      <td>gcg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>02</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LR-3059.lab.CR-00865-16.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LR-2342.lab.FN-01704-18.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LI2163509.lab.14722_6_94.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LI2149109.lab.14893_2_65.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00127822.lab.YA00127822.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.PL00350106.lab.PL00350106.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00024313.lab.YA00024313.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>acc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00025292.lab.YA00025292.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315G</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>ggc</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00038932.lab.YA00038932.iso.1</th>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>315.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agc</td>\n",
       "      <td>aca</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>GENE</td>\n",
       "      <td>AAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>155 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              GENE MUTATION  POSITION  \\\n",
       "UNIQUEID                                                                \n",
       "site.02.subj.0122.lab.22A076.iso.1            katG    P232A     232.0   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1    katG    S315T     315.0   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1    katG    S315T     315.0   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1   katG    S315T     315.0   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1   katG    S315T     315.0   \n",
       "...                                            ...      ...       ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1  katG    S315T     315.0   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1  katG    S315T     315.0   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1  katG    S315T     315.0   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1  katG    S315G     315.0   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1  katG    S315T     315.0   \n",
       "\n",
       "                                              AMINO_ACID_NUMBER  GENOME_INDEX  \\\n",
       "UNIQUEID                                                                        \n",
       "site.02.subj.0122.lab.22A076.iso.1                        232.0           NaN   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1                315.0           NaN   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1                315.0           NaN   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1               315.0           NaN   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1               315.0           NaN   \n",
       "...                                                         ...           ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1              315.0           NaN   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1              315.0           NaN   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1              315.0           NaN   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1              315.0           NaN   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1              315.0           NaN   \n",
       "\n",
       "                                              NUCLEOTIDE_NUMBER  REF  ALT  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.0122.lab.22A076.iso.1                          NaN  ccg  gcg   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1                  NaN  agc  acc   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1                  NaN  agc  acc   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1                 NaN  agc  acc   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1                 NaN  agc  acc   \n",
       "...                                                         ...  ...  ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1                NaN  agc  acc   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1                NaN  agc  acc   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1                NaN  agc  acc   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1                NaN  agc  ggc   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1                NaN  agc  aca   \n",
       "\n",
       "                                              IS_SNP  IS_INDEL  ...  IS_HET  \\\n",
       "UNIQUEID                                                        ...           \n",
       "site.02.subj.0122.lab.22A076.iso.1              True     False  ...   False   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1      True     False  ...   False   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1      True     False  ...   False   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1     True     False  ...   False   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1     True     False  ...   False   \n",
       "...                                              ...       ...  ...     ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1    True     False  ...   False   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1    True     False  ...   False   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1    True     False  ...   False   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1    True     False  ...   False   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1    True     False  ...   False   \n",
       "\n",
       "                                              IS_NULL  IS_FILTER_PASS  \\\n",
       "UNIQUEID                                                                \n",
       "site.02.subj.0122.lab.22A076.iso.1              False            True   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1      False            True   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1      False            True   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1     False            True   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1     False            True   \n",
       "...                                               ...             ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1    False            True   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1    False            True   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1    False            True   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1    False            True   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1    False            True   \n",
       "\n",
       "                                              ELEMENT_TYPE  MUTATION_TYPE  \\\n",
       "UNIQUEID                                                                    \n",
       "site.02.subj.0122.lab.22A076.iso.1                    GENE            AAM   \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1            GENE            AAM   \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1            GENE            AAM   \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1           GENE            AAM   \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1           GENE            AAM   \n",
       "...                                                    ...            ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1          GENE            AAM   \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1          GENE            AAM   \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1          GENE            AAM   \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1          GENE            AAM   \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1          GENE            AAM   \n",
       "\n",
       "                                              INDEL_LENGTH  INDEL_1 INDEL_2  \\\n",
       "UNIQUEID                                                                      \n",
       "site.02.subj.0122.lab.22A076.iso.1                     NaN                    \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1             NaN                    \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1             NaN                    \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1            NaN                    \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1            NaN                    \n",
       "...                                                    ...      ...     ...   \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1           NaN                    \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1           NaN                    \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1           NaN                    \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1           NaN                    \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1           NaN                    \n",
       "\n",
       "                                             SITEID  NUMBER_NUCLEOTIDE_CHANGES  \n",
       "UNIQUEID                                                                        \n",
       "site.02.subj.0122.lab.22A076.iso.1               02                          1  \n",
       "site.05.subj.LR-3059.lab.CR-00865-16.iso.1       05                          1  \n",
       "site.05.subj.LR-2342.lab.FN-01704-18.iso.1       05                          1  \n",
       "site.05.subj.LI2163509.lab.14722_6_94.iso.1      05                          1  \n",
       "site.05.subj.LI2149109.lab.14893_2_65.iso.1      05                          1  \n",
       "...                                             ...                        ...  \n",
       "site.10.subj.YA00127822.lab.YA00127822.iso.1     10                          1  \n",
       "site.10.subj.PL00350106.lab.PL00350106.iso.1     10                          1  \n",
       "site.10.subj.YA00024313.lab.YA00024313.iso.1     10                          1  \n",
       "site.10.subj.YA00025292.lab.YA00025292.iso.1     10                          1  \n",
       "site.10.subj.YA00038932.lab.YA00038932.iso.1     10                          2  \n",
       "\n",
       "[155 rows x 24 columns]"
      ]
     },
     "execution_count": 340,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS = pd.read_pickle('data/MUTATIONS_SAMPLE.pkl.gz')\n",
    "MUTATIONS = MUTATIONS.reset_index()\n",
    "MUTATIONS.set_index('UNIQUEID', inplace = True)\n",
    "\n",
    "#R463L is a known phylogenetic marker for Beijing lineage\n",
    "MUTATIONS = MUTATIONS[(~MUTATIONS.IS_NULL) & (MUTATIONS.IS_FILTER_PASS) & (~MUTATIONS.IS_HET) & (MUTATIONS.IS_SNP) \\\n",
    "                    & (MUTATIONS.IN_CDS) & (MUTATIONS.GENE == 'katG') & (MUTATIONS.IS_NONSYNONYMOUS) & (~(MUTATIONS.MUTATION == 'R463L'))]\n",
    "MUTATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Out of the 5063 samples with non-lineage mutations in gyrA/B 3974 show only one such mutation per sample\n"
     ]
    }
   ],
   "source": [
    "#read in mutations (pre-filtered for gyrA/B mutations due to size of mutations data file)\n",
    "MUTATIONS = pd.read_csv('data/MUTATIONS-gyrA_B.csv.gz', compression='gzip', header = None)\n",
    "\n",
    "#restore header through shortened mutations file\n",
    "HEADER = pd.read_pickle('data/MUTATIONS_SAMPLE.pkl.gz', compression='gzip')\n",
    "\n",
    "HEADER.reset_index(inplace = True)\n",
    "HEADER = HEADER.columns.to_list()\n",
    "MUTATIONS.columns = HEADER\n",
    "\n",
    "#remove non-synonymous, promoter and mutations with inconclusive sequencing, as well as indels\n",
    "MUTATIONS = MUTATIONS[(~MUTATIONS.IS_SYNONYMOUS) & (~MUTATIONS.IN_PROMOTER) & (MUTATIONS.IS_SNP) & (~MUTATIONS.IS_NULL) & (MUTATIONS.IS_FILTER_PASS)]\n",
    "MUTATIONS['GENE_MUTATION'] = MUTATIONS['GENE'] + '_' + MUTATIONS['MUTATION']\n",
    "\n",
    "#label lineage mutations\n",
    "#ref for lineage mutations: Miotto ERJ 2017\n",
    "LINEAGE_MUTATIONS=['gyrA_S95T','gyrA_E21Q','gyrA_T80A','gyrA_A90G','gyrA_G247S','gyrA_A384V',\n",
    "             'gyrA_G668D','gyrA_L712V', 'gyrA_S250A','gyrA_R252L','gyrA_L398F','gyrA_A463S','gyrA_D639A','gyrA_V742L',\n",
    "              'gyrB_V301L','gyrB_M291I','gyrB_A403S']\n",
    "MUTATIONS['LINEAGE_MUTATION'] = MUTATIONS.GENE_MUTATION.isin(LINEAGE_MUTATIONS)\n",
    "\n",
    "#record how many mutations we see for each sample per UNIQUEID -> we can only perform mutation based approach for samples with < 2 non-lineage mutations in pncA\n",
    "MUTATIONS.set_index('UNIQUEID', inplace = True)\n",
    "\n",
    "number_mutations = MUTATIONS[MUTATIONS.LINEAGE_MUTATION == False].index.value_counts()\n",
    "number_mutations.name = 'MUTATION_COUNT'\n",
    "MUTATIONS = MUTATIONS.join(number_mutations.to_frame())\n",
    "\n",
    "print(\"Out of the\", len(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False)]), \n",
    "      \"samples with non-lineage mutations in gyrA/B\", \n",
    "      len(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) & (MUTATIONS.MUTATION_COUNT < 2)]), \n",
    "          \"show only one such mutation per sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Good-quality Phenotypic data is available for 2441 of samples that show only one non-lineage mutation per sample in gyrA/B\n"
     ]
    }
   ],
   "source": [
    "#amount of eligible samples where phenotypic data is available\n",
    "print(\"Good-quality Phenotypic data is available for\", len(PHENOTYPES[PHENOTYPES.index.isin(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) \n",
    "                                                                                         & (MUTATIONS.MUTATION_COUNT < 2) ].index.to_list()) \n",
    "                                                                      & (PHENOTYPES.PHENOTYPE_QUALITY!='LOW')].index.unique()),\n",
    "      \"of samples that show only one non-lineage mutation per sample in gyrA/B\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove low quality phenotypes and divide data by drug (LEV, MXF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BINARY_PHENOTYPE\n",
       "R    1236\n",
       "S     903\n",
       "I       0\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 343,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LEV_PHENOTYPES = PHENOTYPES[(PHENOTYPES.index.isin(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) \n",
    "                                           & (MUTATIONS.MUTATION_COUNT < 2)].index.to_list())) \n",
    "           & (PHENOTYPES.DRUG == 'LEV') & (PHENOTYPES.PHENOTYPE_QUALITY!='LOW')]\n",
    "LEV_PHENOTYPES.BINARY_PHENOTYPE.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BINARY_PHENOTYPE\n",
       "R    1031\n",
       "S     981\n",
       "I       0\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 344,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_PHENOTYPES = PHENOTYPES[(PHENOTYPES.index.isin(MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) \n",
    "                                           & (MUTATIONS.MUTATION_COUNT < 2)].index.to_list())) \n",
    "           & (PHENOTYPES.DRUG == 'MXF') & (PHENOTYPES.PHENOTYPE_QUALITY!='LOW')]\n",
    "MXF_PHENOTYPES.BINARY_PHENOTYPE.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Join the mutations data with the phenotypic data frame for both drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0027.lab.06MIL0488.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0843.lab.06MIL1790.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94Y</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00624.lab.709302.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>N499D</td>\n",
       "      <td>499.0</td>\n",
       "      <td>gyrB_N499D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.YA00021495.lab.YA00021495.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>N193S</td>\n",
       "      <td>193.0</td>\n",
       "      <td>gyrA_N193S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.XD02435891.lab.XD02435891.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0482.lab.235081-14.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>G512R</td>\n",
       "      <td>512.0</td>\n",
       "      <td>gyrB_G512R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.LR-2321.lab.FN-01077-18.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>&gt;8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G88C</td>\n",
       "      <td>88.0</td>\n",
       "      <td>gyrA_G88C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.02.subj.0910.lab.22A144.iso.1</th>\n",
       "      <td>LEV</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2171 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG  MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                          \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1    LEV  2.0      1.0   \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1    LEV   >8      4.0   \n",
       "site.04.subj.00624.lab.709302.iso.1           LEV   >8      4.0   \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1  LEV  2.0      1.0   \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1  LEV  0.5     -1.0   \n",
       "...                                           ...  ...      ...   \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1  LEV  8.0      3.0   \n",
       "site.02.subj.0482.lab.235081-14.iso.1         LEV  1.0      0.0   \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1    LEV   >8      4.0   \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1  LEV  8.0      3.0   \n",
       "site.02.subj.0910.lab.22A144.iso.1            LEV  4.0      2.0   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1                  R   \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1                  R   \n",
       "site.04.subj.00624.lab.709302.iso.1                         R   \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1                R   \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1                S   \n",
       "...                                                       ...   \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1                R   \n",
       "site.02.subj.0482.lab.235081-14.iso.1                       S   \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1                  R   \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1                R   \n",
       "site.02.subj.0910.lab.22A144.iso.1                          R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE MUTATION  \\\n",
       "UNIQUEID                                                                        \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1              MEDIUM  gyrA     A90V   \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1                HIGH  gyrA     D94Y   \n",
       "site.04.subj.00624.lab.709302.iso.1                       HIGH  gyrA     D94N   \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1            MEDIUM  gyrB    N499D   \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1              HIGH  gyrA    N193S   \n",
       "...                                                        ...   ...      ...   \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1              HIGH  gyrA     D94N   \n",
       "site.02.subj.0482.lab.235081-14.iso.1                     HIGH  gyrB    G512R   \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1                HIGH  gyrA     G88C   \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1            MEDIUM  gyrA     D94N   \n",
       "site.02.subj.0910.lab.22A144.iso.1                        HIGH  gyrA     A90V   \n",
       "\n",
       "                                              POSITION GENE_MUTATION  \n",
       "UNIQUEID                                                              \n",
       "site.06.subj.06TB_0027.lab.06MIL0488.iso.1        90.0     gyrA_A90V  \n",
       "site.06.subj.06TB_0843.lab.06MIL1790.iso.1        94.0     gyrA_D94Y  \n",
       "site.04.subj.00624.lab.709302.iso.1               94.0     gyrA_D94N  \n",
       "site.11.subj.XTB-18-196.lab.XTB-18-196.iso.1     499.0    gyrB_N499D  \n",
       "site.10.subj.YA00021495.lab.YA00021495.iso.1     193.0    gyrA_N193S  \n",
       "...                                                ...           ...  \n",
       "site.10.subj.XD02435891.lab.XD02435891.iso.1      94.0     gyrA_D94N  \n",
       "site.02.subj.0482.lab.235081-14.iso.1            512.0    gyrB_G512R  \n",
       "site.05.subj.LR-2321.lab.FN-01077-18.iso.1        88.0     gyrA_G88C  \n",
       "site.11.subj.XTB_18-072.lab.XTB_18-072.iso.1      94.0     gyrA_D94N  \n",
       "site.02.subj.0910.lab.22A144.iso.1                90.0     gyrA_A90V  \n",
       "\n",
       "[2171 rows x 9 columns]"
      ]
     },
     "execution_count": 345,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS = MUTATIONS[(MUTATIONS.LINEAGE_MUTATION == False) & (MUTATIONS.MUTATION_COUNT < 2)]\n",
    "\n",
    "LEV_DATASET = pd.merge(LEV_PHENOTYPES, MUTATIONS, how = 'left', on = 'UNIQUEID')\n",
    "LEV_DATASET = LEV_DATASET[['DRUG', 'MIC', 'LOG2MIC', 'BINARY_PHENOTYPE', 'PHENOTYPE_QUALITY', 'GENE', 'MUTATION', 'POSITION', 'GENE_MUTATION']]\n",
    "LEV_DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00097.lab.633433.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH00935117.lab.UH00935117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00121.lab.633619.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.14.subj.2591.lab.2591.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94N</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.01310.lab.724468.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S91P</td>\n",
       "      <td>91.0</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH01302806.lab.UH01302806.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0020.lab.CO-05421-18.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>G594R</td>\n",
       "      <td>594.0</td>\n",
       "      <td>gyrB_G594R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0968.lab.06MIL1963.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2043 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG   MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                           \n",
       "site.04.subj.00097.lab.633433.iso.1           MXF   2.0     1.00   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1  MXF   4.0     2.00   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1  MXF    >4     3.00   \n",
       "site.04.subj.00121.lab.633619.iso.1           MXF   2.0     1.00   \n",
       "site.14.subj.2591.lab.2591.iso.1              MXF    >4     3.00   \n",
       "...                                           ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1           MXF   1.0     0.00   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1  MXF   2.0     1.00   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1  MXF   2.0     1.00   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1    MXF  0.12    -3.06   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1    MXF    >4     3.00   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                         R   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1                R   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1                R   \n",
       "site.04.subj.00121.lab.633619.iso.1                         R   \n",
       "site.14.subj.2591.lab.2591.iso.1                            R   \n",
       "...                                                       ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                         S   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1                R   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1                R   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                  S   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                  R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE MUTATION  \\\n",
       "UNIQUEID                                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                       HIGH  gyrA     D94G   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1              HIGH  gyrA     A90V   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1            MEDIUM  gyrA     D94N   \n",
       "site.04.subj.00121.lab.633619.iso.1                       HIGH  gyrA     D94G   \n",
       "site.14.subj.2591.lab.2591.iso.1                          HIGH  gyrA     D94N   \n",
       "...                                                        ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                       HIGH  gyrA     S91P   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1            MEDIUM  gyrA     A90V   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1              HIGH  gyrA     A90V   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                HIGH  gyrB    G594R   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                HIGH  gyrA     A90V   \n",
       "\n",
       "                                              POSITION GENE_MUTATION  \n",
       "UNIQUEID                                                              \n",
       "site.04.subj.00097.lab.633433.iso.1               94.0     gyrA_D94G  \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1      90.0     gyrA_A90V  \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1      94.0     gyrA_D94N  \n",
       "site.04.subj.00121.lab.633619.iso.1               94.0     gyrA_D94G  \n",
       "site.14.subj.2591.lab.2591.iso.1                  94.0     gyrA_D94N  \n",
       "...                                                ...           ...  \n",
       "site.04.subj.01310.lab.724468.iso.1               91.0     gyrA_S91P  \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1      90.0     gyrA_A90V  \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1      90.0     gyrA_A90V  \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1       594.0    gyrB_G594R  \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1        90.0     gyrA_A90V  \n",
       "\n",
       "[2043 rows x 9 columns]"
      ]
     },
     "execution_count": 346,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET = pd.merge(MXF_PHENOTYPES, MUTATIONS, how = 'left', on = 'UNIQUEID')\n",
    "MXF_DATASET = MXF_DATASET[['DRUG', 'MIC', 'LOG2MIC', 'BINARY_PHENOTYPE', 'PHENOTYPE_QUALITY', 'GENE', 'MUTATION', 'POSITION', 'GENE_MUTATION']]\n",
    "MXF_DATASET"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create features using sbml core"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_segid(row):\n",
    "    return row['GENE'][3:]\n",
    "\n",
    "MXF_DATASET['segid'] = MXF_DATASET.apply(make_segid, axis = 1)\n",
    "MXF_DATASET['mutation'] = MXF_DATASET['MUTATION']\n",
    "MXF_DATASET.drop('MUTATION', axis = 1, inplace = True)\n",
    "\n",
    "LEV_DATASET['segid'] = LEV_DATASET.apply(make_segid, axis = 1)\n",
    "LEV_DATASET['mutation'] = LEV_DATASET['MUTATION']\n",
    "LEV_DATASET.drop('MUTATION', axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "      <th>segid</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00097.lab.633433.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>A</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH00935117.lab.UH00935117.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>A</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00121.lab.633619.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>A</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.14.subj.2591.lab.2591.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>A</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.04.subj.01310.lab.724468.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>91.0</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "      <td>A</td>\n",
       "      <td>S91P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.10.subj.UH01302806.lab.UH01302806.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.CA-0020.lab.CO-05421-18.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>594.0</td>\n",
       "      <td>gyrB_G594R</td>\n",
       "      <td>B</td>\n",
       "      <td>G594R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.06.subj.06TB_0968.lab.06MIL1963.iso.1</th>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>A90V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2043 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DRUG   MIC  LOG2MIC  \\\n",
       "UNIQUEID                                                           \n",
       "site.04.subj.00097.lab.633433.iso.1           MXF   2.0     1.00   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1  MXF   4.0     2.00   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1  MXF    >4     3.00   \n",
       "site.04.subj.00121.lab.633619.iso.1           MXF   2.0     1.00   \n",
       "site.14.subj.2591.lab.2591.iso.1              MXF    >4     3.00   \n",
       "...                                           ...   ...      ...   \n",
       "site.04.subj.01310.lab.724468.iso.1           MXF   1.0     0.00   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1  MXF   2.0     1.00   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1  MXF   2.0     1.00   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1    MXF  0.12    -3.06   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1    MXF    >4     3.00   \n",
       "\n",
       "                                             BINARY_PHENOTYPE  \\\n",
       "UNIQUEID                                                        \n",
       "site.04.subj.00097.lab.633433.iso.1                         R   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1                R   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1                R   \n",
       "site.04.subj.00121.lab.633619.iso.1                         R   \n",
       "site.14.subj.2591.lab.2591.iso.1                            R   \n",
       "...                                                       ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                         S   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1                R   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1                R   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                  S   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                  R   \n",
       "\n",
       "                                             PHENOTYPE_QUALITY  GENE  \\\n",
       "UNIQUEID                                                               \n",
       "site.04.subj.00097.lab.633433.iso.1                       HIGH  gyrA   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1              HIGH  gyrA   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1            MEDIUM  gyrA   \n",
       "site.04.subj.00121.lab.633619.iso.1                       HIGH  gyrA   \n",
       "site.14.subj.2591.lab.2591.iso.1                          HIGH  gyrA   \n",
       "...                                                        ...   ...   \n",
       "site.04.subj.01310.lab.724468.iso.1                       HIGH  gyrA   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1            MEDIUM  gyrA   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1              HIGH  gyrA   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1                HIGH  gyrB   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1                HIGH  gyrA   \n",
       "\n",
       "                                              POSITION GENE_MUTATION segid  \\\n",
       "UNIQUEID                                                                     \n",
       "site.04.subj.00097.lab.633433.iso.1               94.0     gyrA_D94G     A   \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1      90.0     gyrA_A90V     A   \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1      94.0     gyrA_D94N     A   \n",
       "site.04.subj.00121.lab.633619.iso.1               94.0     gyrA_D94G     A   \n",
       "site.14.subj.2591.lab.2591.iso.1                  94.0     gyrA_D94N     A   \n",
       "...                                                ...           ...   ...   \n",
       "site.04.subj.01310.lab.724468.iso.1               91.0     gyrA_S91P     A   \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1      90.0     gyrA_A90V     A   \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1      90.0     gyrA_A90V     A   \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1       594.0    gyrB_G594R     B   \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1        90.0     gyrA_A90V     A   \n",
       "\n",
       "                                             mutation  \n",
       "UNIQUEID                                               \n",
       "site.04.subj.00097.lab.633433.iso.1              D94G  \n",
       "site.10.subj.UH00935117.lab.UH00935117.iso.1     A90V  \n",
       "site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1     D94N  \n",
       "site.04.subj.00121.lab.633619.iso.1              D94G  \n",
       "site.14.subj.2591.lab.2591.iso.1                 D94N  \n",
       "...                                               ...  \n",
       "site.04.subj.01310.lab.724468.iso.1              S91P  \n",
       "site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1     A90V  \n",
       "site.10.subj.UH01302806.lab.UH01302806.iso.1     A90V  \n",
       "site.05.subj.CA-0020.lab.CO-05421-18.iso.1      G594R  \n",
       "site.06.subj.06TB_0968.lab.06MIL1963.iso.1       A90V  \n",
       "\n",
       "[2043 rows x 10 columns]"
      ]
     },
     "execution_count": 348,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def update_binary_phenotype(row):\n",
    "\n",
    "#     target_string = 'gyrA_D94G'  # The string to filter rows\n",
    "    \n",
    "#     if row['GENE_MUTATION'] == target_string:\n",
    "        \n",
    "#         filtered_df = MXF_DATASET[MXF_DATASET['GENE_MUTATION'] == target_string]\n",
    "#         most_common_value = filtered_df['BINARY_PHENOTYPE'].mode().iloc[0]\n",
    "\n",
    "#         return most_common_value\n",
    "#     else:\n",
    "#         return row['BINARY_PHENOTYPE']\n",
    "    \n",
    "# MXF_DATASET['BINARY_PHENOTYPE'] = MXF_DATASET.apply(update_binary_phenotype, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 425,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_binary_phenotype(row, target_strings):\n",
    "\n",
    "    if row['GENE_MUTATION'] in target_strings:\n",
    "        \n",
    "        target_string = row['GENE_MUTATION']\n",
    "        filtered_df = MXF_DATASET[MXF_DATASET['GENE_MUTATION'] == target_string]\n",
    "        \n",
    "        if len(filtered_df) < 2:\n",
    "            \n",
    "            return row['BINARY_PHENOTYPE']\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            most_common_value = filtered_df['BINARY_PHENOTYPE'].mode().iloc[0]\n",
    "\n",
    "            return most_common_value\n",
    "    \n",
    "    else:\n",
    "        return row['BINARY_PHENOTYPE']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_strings_MXF = MXF_DATASET.GENE_MUTATION.unique()\n",
    "target_strings_LEV = LEV_DATASET.GENE_MUTATION.unique()\n",
    "target_strings = np.union1d(target_strings_MXF, target_strings_LEV)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 427,
   "metadata": {},
   "outputs": [],
   "source": [
    "MXF_DATASET['BINARY_PHENOTYPE'] = MXF_DATASET.apply(lambda row: update_binary_phenotype(row, target_strings), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "metadata": {},
   "outputs": [],
   "source": [
    "LEV_DATASET['BINARY_PHENOTYPE'] = LEV_DATASET.apply(lambda row: update_binary_phenotype(row, target_strings), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 429,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BINARY_PHENOTYPE\n",
       "R    619\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 429,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET[MXF_DATASET['GENE_MUTATION'] == 'gyrA_D94G'].BINARY_PHENOTYPE.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 430,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BINARY_PHENOTYPE\n",
       "R    619\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 430,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET[MXF_DATASET['GENE_MUTATION'] == 'gyrA_D94G'].BINARY_PHENOTYPE.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 431,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BINARY_PHENOTYPE\n",
       "R    58\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 431,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LEV_DATASET[LEV_DATASET['GENE_MUTATION'] == 'gyrA_D94Y'].BINARY_PHENOTYPE.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 432,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UNIQUEID\n",
       "site.10.subj.YA00009954.lab.YA00009954.iso.1    S\n",
       "site.20.subj.SA00329756.lab.YA00135259.iso.1    S\n",
       "site.10.subj.YA00023537.lab.YA00023537.iso.1    S\n",
       "site.10.subj.YA00106034.lab.YA00106034.iso.1    S\n",
       "site.10.subj.YA00127803.lab.YA00127803.iso.1    S\n",
       "                                               ..\n",
       "site.10.subj.TRL0024265.lab.TRL0024265.iso.1    S\n",
       "site.10.subj.YA00022119.lab.YA00022119.iso.1    S\n",
       "site.10.subj.YA00089269.lab.YA00089269.iso.1    S\n",
       "site.20.subj.SCH7908405.lab.YA00131558.iso.1    S\n",
       "site.10.subj.YA00028138.lab.YA00028138.iso.1    S\n",
       "Name: BINARY_PHENOTYPE, Length: 75, dtype: object"
      ]
     },
     "execution_count": 432,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET[MXF_DATASET['GENE_MUTATION'] == 'gyrA_Q613E'].BINARY_PHENOTYPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 433,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE_MUTATION\n",
       "gyrA_D94G     619\n",
       "gyrA_A90V     336\n",
       "gyrA_D94N     136\n",
       "gyrB_P94L     131\n",
       "gyrA_D94A      86\n",
       "gyrA_Q613E     75\n",
       "gyrA_S91P      65\n",
       "gyrA_D94Y      60\n",
       "gyrA_D94H      33\n",
       "gyrA_G88C      22\n",
       "gyrA_N193S     20\n",
       "gyrB_E501D     15\n",
       "gyrA_P472S     14\n",
       "gyrA_T267I     12\n",
       "gyrA_P8A       12\n",
       "gyrB_D461N     11\n",
       "gyrB_R446C     10\n",
       "gyrA_P154R      9\n",
       "gyrB_A130S      9\n",
       "gyrB_I84V       8\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 433,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MXF_DATASET.GENE_MUTATION.value_counts()[:20]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For now, drop rows with mutations outside of crystallized regions, might impute values later on (segid A: 15-501, segid B: 425-675)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [],
   "source": [
    "MXF_DATASET = MXF_DATASET[(MXF_DATASET.segid == 'A') & MXF_DATASET.POSITION.astype(int).isin(list(range(15,502))) \n",
    "            | (MXF_DATASET.segid == 'B') & MXF_DATASET.POSITION.astype(int).isin(list(range(425,676)))]\n",
    "\n",
    "LEV_DATASET = LEV_DATASET[(LEV_DATASET.segid == 'A') & LEV_DATASET.POSITION.astype(int).isin(list(range(15,502))) \n",
    "            | (LEV_DATASET.segid == 'B') & LEV_DATASET.POSITION.astype(int).isin(list(range(425,676)))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 435,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/5bs8.pdb'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "    \u001b[0;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "Cell \u001b[0;32mIn[435], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mMDAnalysis\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mmda\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m pdb \u001b[39m=\u001b[39m mda\u001b[39m.\u001b[39;49mUniverse(\u001b[39m'\u001b[39;49m\u001b[39mdata/5bs8.pdb\u001b[39;49m\u001b[39m'\u001b[39;49m)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/universe.py:356\u001b[0m, in \u001b[0;36mUniverse.__init__\u001b[0;34m(self, topology, all_coordinates, format, topology_format, transformations, guess_bonds, vdwradii, fudge_factor, lower_bound, in_memory, in_memory_step, *coordinates, **kwargs)\u001b[0m\n\u001b[1;32m    355\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfilename \u001b[39m=\u001b[39m _check_file_like(topology)\n\u001b[0;32m--> 356\u001b[0m     topology \u001b[39m=\u001b[39m _topology_from_file_like(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfilename,\n\u001b[1;32m    357\u001b[0m                                         topology_format\u001b[39m=\u001b[39;49mtopology_format,\n\u001b[1;32m    358\u001b[0m                                         \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    360\u001b[0m \u001b[39mif\u001b[39;00m topology \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/universe.py:118\u001b[0m, in \u001b[0;36m_topology_from_file_like\u001b[0;34m(topology_file, topology_format, **kwargs)\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[39mif\u001b[39;00m (err\u001b[39m.\u001b[39merrno \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m\n\u001b[1;32m    116\u001b[0m     errno\u001b[39m.\u001b[39merrorcode[err\u001b[39m.\u001b[39merrno] \u001b[39min\u001b[39;00m [\u001b[39m'\u001b[39m\u001b[39mENOENT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mEACCES\u001b[39m\u001b[39m'\u001b[39m]):\n\u001b[1;32m    117\u001b[0m     \u001b[39m# Runs if the error is propagated due to no permission / file not found\u001b[39;00m\n\u001b[0;32m--> 118\u001b[0m     \u001b[39mraise\u001b[39;00m sys\u001b[39m.\u001b[39mexc_info()[\u001b[39m1\u001b[39m] \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m    119\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    120\u001b[0m     \u001b[39m# Runs when the parser fails\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/universe.py:110\u001b[0m, in \u001b[0;36m_topology_from_file_like\u001b[0;34m(topology_file, topology_format, **kwargs)\u001b[0m\n\u001b[1;32m    109\u001b[0m     \u001b[39mwith\u001b[39;00m parser(topology_file) \u001b[39mas\u001b[39;00m p:\n\u001b[0;32m--> 110\u001b[0m         topology \u001b[39m=\u001b[39m p\u001b[39m.\u001b[39;49mparse(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    111\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    112\u001b[0m     \u001b[39m# There are 2 kinds of errors that might be raised here:\u001b[39;00m\n\u001b[1;32m    113\u001b[0m     \u001b[39m# one because the file isn't present\u001b[39;00m\n\u001b[1;32m    114\u001b[0m     \u001b[39m# or the permissions are bad, second when the parser fails\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/topology/PDBParser.py:210\u001b[0m, in \u001b[0;36mPDBParser.parse\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    204\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Parse atom information from PDB file\u001b[39;00m\n\u001b[1;32m    205\u001b[0m \n\u001b[1;32m    206\u001b[0m \u001b[39mReturns\u001b[39;00m\n\u001b[1;32m    207\u001b[0m \u001b[39m-------\u001b[39;00m\n\u001b[1;32m    208\u001b[0m \u001b[39mMDAnalysis Topology object\u001b[39;00m\n\u001b[1;32m    209\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m--> 210\u001b[0m top \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_parseatoms()\n\u001b[1;32m    212\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/topology/PDBParser.py:247\u001b[0m, in \u001b[0;36mPDBParser._parseatoms\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    246\u001b[0m last_wrapped_serial \u001b[39m=\u001b[39m \u001b[39m100000\u001b[39m  \u001b[39m# if serials wrap, start from here\u001b[39;00m\n\u001b[0;32m--> 247\u001b[0m \u001b[39mwith\u001b[39;00m util\u001b[39m.\u001b[39mopenany(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfilename) \u001b[39mas\u001b[39;00m f:\n\u001b[1;32m    248\u001b[0m     \u001b[39mfor\u001b[39;00m line \u001b[39min\u001b[39;00m f:\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py:135\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    134\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 135\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mnext\u001b[39;49m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgen)\n\u001b[1;32m    136\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mStopIteration\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:318\u001b[0m, in \u001b[0;36mopenany\u001b[0;34m(datasource, mode, reset)\u001b[0m\n\u001b[1;32m    276\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Context manager for :func:`anyopen`.\u001b[39;00m\n\u001b[1;32m    277\u001b[0m \n\u001b[1;32m    278\u001b[0m \u001b[39mOpen the `datasource` and close it when the context of the :keyword:`with`\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[39m:func:`anyopen`\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m--> 318\u001b[0m stream \u001b[39m=\u001b[39m anyopen(datasource, mode\u001b[39m=\u001b[39;49mmode, reset\u001b[39m=\u001b[39;49mreset)\n\u001b[1;32m    319\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:397\u001b[0m, in \u001b[0;36manyopen\u001b[0;34m(datasource, mode, reset)\u001b[0m\n\u001b[1;32m    396\u001b[0m openfunc \u001b[39m=\u001b[39m read_handlers[ext]\n\u001b[0;32m--> 397\u001b[0m stream \u001b[39m=\u001b[39m _get_stream(datasource, openfunc, mode\u001b[39m=\u001b[39;49mmode)\n\u001b[1;32m    398\u001b[0m \u001b[39mif\u001b[39;00m stream \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:439\u001b[0m, in \u001b[0;36m_get_stream\u001b[0;34m(filename, openfunction, mode)\u001b[0m\n\u001b[1;32m    438\u001b[0m \u001b[39mif\u001b[39;00m errno\u001b[39m.\u001b[39merrorcode[err\u001b[39m.\u001b[39merrno] \u001b[39min\u001b[39;00m [\u001b[39m'\u001b[39m\u001b[39mENOENT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mEACCES\u001b[39m\u001b[39m'\u001b[39m]:\n\u001b[0;32m--> 439\u001b[0m     \u001b[39mraise\u001b[39;00m sys\u001b[39m.\u001b[39mexc_info()[\u001b[39m1\u001b[39m] \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m    440\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:433\u001b[0m, in \u001b[0;36m_get_stream\u001b[0;34m(filename, openfunction, mode)\u001b[0m\n\u001b[1;32m    432\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 433\u001b[0m     stream \u001b[39m=\u001b[39m openfunction(filename, mode\u001b[39m=\u001b[39;49mmode)\n\u001b[1;32m    434\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    435\u001b[0m     \u001b[39m# An exception might be raised due to two reasons, first the openfunction is unable to open the file, in this\u001b[39;00m\n\u001b[1;32m    436\u001b[0m     \u001b[39m# case we have to ignore the error and return None. Second is when openfunction can't open the file because\u001b[39;00m\n\u001b[1;32m    437\u001b[0m     \u001b[39m# either the file isn't there or the permissions don't allow access.\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/picklable_file_io.py:478\u001b[0m, in \u001b[0;36mbz2_pickle_open\u001b[0;34m(name, mode)\u001b[0m\n\u001b[1;32m    477\u001b[0m bz_mode \u001b[39m=\u001b[39m mode\u001b[39m.\u001b[39mreplace(\u001b[39m\"\u001b[39m\u001b[39mt\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 478\u001b[0m binary_file \u001b[39m=\u001b[39m BZ2Picklable(name, bz_mode)\n\u001b[1;32m    479\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mt\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m mode:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/picklable_file_io.py:266\u001b[0m, in \u001b[0;36mBZ2Picklable.__init__\u001b[0;34m(self, name, mode)\u001b[0m\n\u001b[1;32m    265\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_bz_mode \u001b[39m=\u001b[39m mode\n\u001b[0;32m--> 266\u001b[0m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49m\u001b[39m__init__\u001b[39;49m(name, mode)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/bz2.py:81\u001b[0m, in \u001b[0;36mBZ2File.__init__\u001b[0;34m(self, filename, mode, compresslevel)\u001b[0m\n\u001b[1;32m     80\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(filename, (\u001b[39mstr\u001b[39m, \u001b[39mbytes\u001b[39m, os\u001b[39m.\u001b[39mPathLike)):\n\u001b[0;32m---> 81\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_fp \u001b[39m=\u001b[39m _builtin_open(filename, mode)\n\u001b[1;32m     82\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_closefp \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/5bs8.pdb'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "    \u001b[0;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "Cell \u001b[0;32mIn[435], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mMDAnalysis\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mmda\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m pdb \u001b[39m=\u001b[39m mda\u001b[39m.\u001b[39;49mUniverse(\u001b[39m'\u001b[39;49m\u001b[39mdata/5bs8.pdb\u001b[39;49m\u001b[39m'\u001b[39;49m)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/universe.py:356\u001b[0m, in \u001b[0;36mUniverse.__init__\u001b[0;34m(self, topology, all_coordinates, format, topology_format, transformations, guess_bonds, vdwradii, fudge_factor, lower_bound, in_memory, in_memory_step, *coordinates, **kwargs)\u001b[0m\n\u001b[1;32m    355\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfilename \u001b[39m=\u001b[39m _check_file_like(topology)\n\u001b[0;32m--> 356\u001b[0m     topology \u001b[39m=\u001b[39m _topology_from_file_like(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfilename,\n\u001b[1;32m    357\u001b[0m                                         topology_format\u001b[39m=\u001b[39;49mtopology_format,\n\u001b[1;32m    358\u001b[0m                                         \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    360\u001b[0m \u001b[39mif\u001b[39;00m topology \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/universe.py:118\u001b[0m, in \u001b[0;36m_topology_from_file_like\u001b[0;34m(topology_file, topology_format, **kwargs)\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[39mif\u001b[39;00m (err\u001b[39m.\u001b[39merrno \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m\n\u001b[1;32m    116\u001b[0m     errno\u001b[39m.\u001b[39merrorcode[err\u001b[39m.\u001b[39merrno] \u001b[39min\u001b[39;00m [\u001b[39m'\u001b[39m\u001b[39mENOENT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mEACCES\u001b[39m\u001b[39m'\u001b[39m]):\n\u001b[1;32m    117\u001b[0m     \u001b[39m# Runs if the error is propagated due to no permission / file not found\u001b[39;00m\n\u001b[0;32m--> 118\u001b[0m     \u001b[39mraise\u001b[39;00m sys\u001b[39m.\u001b[39mexc_info()[\u001b[39m1\u001b[39m] \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m    119\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    120\u001b[0m     \u001b[39m# Runs when the parser fails\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/universe.py:110\u001b[0m, in \u001b[0;36m_topology_from_file_like\u001b[0;34m(topology_file, topology_format, **kwargs)\u001b[0m\n\u001b[1;32m    109\u001b[0m     \u001b[39mwith\u001b[39;00m parser(topology_file) \u001b[39mas\u001b[39;00m p:\n\u001b[0;32m--> 110\u001b[0m         topology \u001b[39m=\u001b[39m p\u001b[39m.\u001b[39;49mparse(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    111\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    112\u001b[0m     \u001b[39m# There are 2 kinds of errors that might be raised here:\u001b[39;00m\n\u001b[1;32m    113\u001b[0m     \u001b[39m# one because the file isn't present\u001b[39;00m\n\u001b[1;32m    114\u001b[0m     \u001b[39m# or the permissions are bad, second when the parser fails\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/topology/PDBParser.py:210\u001b[0m, in \u001b[0;36mPDBParser.parse\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    204\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Parse atom information from PDB file\u001b[39;00m\n\u001b[1;32m    205\u001b[0m \n\u001b[1;32m    206\u001b[0m \u001b[39mReturns\u001b[39;00m\n\u001b[1;32m    207\u001b[0m \u001b[39m-------\u001b[39;00m\n\u001b[1;32m    208\u001b[0m \u001b[39mMDAnalysis Topology object\u001b[39;00m\n\u001b[1;32m    209\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m--> 210\u001b[0m top \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_parseatoms()\n\u001b[1;32m    212\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/topology/PDBParser.py:247\u001b[0m, in \u001b[0;36mPDBParser._parseatoms\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    246\u001b[0m last_wrapped_serial \u001b[39m=\u001b[39m \u001b[39m100000\u001b[39m  \u001b[39m# if serials wrap, start from here\u001b[39;00m\n\u001b[0;32m--> 247\u001b[0m \u001b[39mwith\u001b[39;00m util\u001b[39m.\u001b[39mopenany(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfilename) \u001b[39mas\u001b[39;00m f:\n\u001b[1;32m    248\u001b[0m     \u001b[39mfor\u001b[39;00m line \u001b[39min\u001b[39;00m f:\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py:135\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    134\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 135\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mnext\u001b[39;49m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgen)\n\u001b[1;32m    136\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mStopIteration\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:318\u001b[0m, in \u001b[0;36mopenany\u001b[0;34m(datasource, mode, reset)\u001b[0m\n\u001b[1;32m    276\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Context manager for :func:`anyopen`.\u001b[39;00m\n\u001b[1;32m    277\u001b[0m \n\u001b[1;32m    278\u001b[0m \u001b[39mOpen the `datasource` and close it when the context of the :keyword:`with`\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[39m:func:`anyopen`\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m--> 318\u001b[0m stream \u001b[39m=\u001b[39m anyopen(datasource, mode\u001b[39m=\u001b[39;49mmode, reset\u001b[39m=\u001b[39;49mreset)\n\u001b[1;32m    319\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:397\u001b[0m, in \u001b[0;36manyopen\u001b[0;34m(datasource, mode, reset)\u001b[0m\n\u001b[1;32m    396\u001b[0m openfunc \u001b[39m=\u001b[39m read_handlers[ext]\n\u001b[0;32m--> 397\u001b[0m stream \u001b[39m=\u001b[39m _get_stream(datasource, openfunc, mode\u001b[39m=\u001b[39;49mmode)\n\u001b[1;32m    398\u001b[0m \u001b[39mif\u001b[39;00m stream \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:439\u001b[0m, in \u001b[0;36m_get_stream\u001b[0;34m(filename, openfunction, mode)\u001b[0m\n\u001b[1;32m    438\u001b[0m \u001b[39mif\u001b[39;00m errno\u001b[39m.\u001b[39merrorcode[err\u001b[39m.\u001b[39merrno] \u001b[39min\u001b[39;00m [\u001b[39m'\u001b[39m\u001b[39mENOENT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mEACCES\u001b[39m\u001b[39m'\u001b[39m]:\n\u001b[0;32m--> 439\u001b[0m     \u001b[39mraise\u001b[39;00m sys\u001b[39m.\u001b[39mexc_info()[\u001b[39m1\u001b[39m] \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m    440\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:433\u001b[0m, in \u001b[0;36m_get_stream\u001b[0;34m(filename, openfunction, mode)\u001b[0m\n\u001b[1;32m    432\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 433\u001b[0m     stream \u001b[39m=\u001b[39m openfunction(filename, mode\u001b[39m=\u001b[39;49mmode)\n\u001b[1;32m    434\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    435\u001b[0m     \u001b[39m# An exception might be raised due to two reasons, first the openfunction is unable to open the file, in this\u001b[39;00m\n\u001b[1;32m    436\u001b[0m     \u001b[39m# case we have to ignore the error and return None. Second is when openfunction can't open the file because\u001b[39;00m\n\u001b[1;32m    437\u001b[0m     \u001b[39m# either the file isn't there or the permissions don't allow access.\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/picklable_file_io.py:478\u001b[0m, in \u001b[0;36mbz2_pickle_open\u001b[0;34m(name, mode)\u001b[0m\n\u001b[1;32m    477\u001b[0m bz_mode \u001b[39m=\u001b[39m mode\u001b[39m.\u001b[39mreplace(\u001b[39m\"\u001b[39m\u001b[39mt\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 478\u001b[0m binary_file \u001b[39m=\u001b[39m BZ2Picklable(name, bz_mode)\n\u001b[1;32m    479\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mt\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m mode:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/picklable_file_io.py:266\u001b[0m, in \u001b[0;36mBZ2Picklable.__init__\u001b[0;34m(self, name, mode)\u001b[0m\n\u001b[1;32m    265\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_bz_mode \u001b[39m=\u001b[39m mode\n\u001b[0;32m--> 266\u001b[0m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49m\u001b[39m__init__\u001b[39;49m(name, mode)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/bz2.py:81\u001b[0m, in \u001b[0;36mBZ2File.__init__\u001b[0;34m(self, filename, mode, compresslevel)\u001b[0m\n\u001b[1;32m     80\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(filename, (\u001b[39mstr\u001b[39m, \u001b[39mbytes\u001b[39m, os\u001b[39m.\u001b[39mPathLike)):\n\u001b[0;32m---> 81\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_fp \u001b[39m=\u001b[39m _builtin_open(filename, mode)\n\u001b[1;32m     82\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_closefp \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/5bs8.pdb'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[435], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mMDAnalysis\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mmda\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m pdb \u001b[39m=\u001b[39m mda\u001b[39m.\u001b[39;49mUniverse(\u001b[39m'\u001b[39;49m\u001b[39mdata/5bs8.pdb\u001b[39;49m\u001b[39m'\u001b[39;49m)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/universe.py:356\u001b[0m, in \u001b[0;36mUniverse.__init__\u001b[0;34m(self, topology, all_coordinates, format, topology_format, transformations, guess_bonds, vdwradii, fudge_factor, lower_bound, in_memory, in_memory_step, *coordinates, **kwargs)\u001b[0m\n\u001b[1;32m    354\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(topology, Topology) \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m topology \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    355\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfilename \u001b[39m=\u001b[39m _check_file_like(topology)\n\u001b[0;32m--> 356\u001b[0m     topology \u001b[39m=\u001b[39m _topology_from_file_like(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfilename,\n\u001b[1;32m    357\u001b[0m                                         topology_format\u001b[39m=\u001b[39;49mtopology_format,\n\u001b[1;32m    358\u001b[0m                                         \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    360\u001b[0m \u001b[39mif\u001b[39;00m topology \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    361\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_topology \u001b[39m=\u001b[39m topology\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/universe.py:118\u001b[0m, in \u001b[0;36m_topology_from_file_like\u001b[0;34m(topology_file, topology_format, **kwargs)\u001b[0m\n\u001b[1;32m    111\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    112\u001b[0m     \u001b[39m# There are 2 kinds of errors that might be raised here:\u001b[39;00m\n\u001b[1;32m    113\u001b[0m     \u001b[39m# one because the file isn't present\u001b[39;00m\n\u001b[1;32m    114\u001b[0m     \u001b[39m# or the permissions are bad, second when the parser fails\u001b[39;00m\n\u001b[1;32m    115\u001b[0m     \u001b[39mif\u001b[39;00m (err\u001b[39m.\u001b[39merrno \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m\n\u001b[1;32m    116\u001b[0m         errno\u001b[39m.\u001b[39merrorcode[err\u001b[39m.\u001b[39merrno] \u001b[39min\u001b[39;00m [\u001b[39m'\u001b[39m\u001b[39mENOENT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mEACCES\u001b[39m\u001b[39m'\u001b[39m]):\n\u001b[1;32m    117\u001b[0m         \u001b[39m# Runs if the error is propagated due to no permission / file not found\u001b[39;00m\n\u001b[0;32m--> 118\u001b[0m         \u001b[39mraise\u001b[39;00m sys\u001b[39m.\u001b[39mexc_info()[\u001b[39m1\u001b[39m] \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m    119\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    120\u001b[0m         \u001b[39m# Runs when the parser fails\u001b[39;00m\n\u001b[1;32m    121\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mIOError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mFailed to load from the topology file \u001b[39m\u001b[39m{0}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m    122\u001b[0m                         \u001b[39m\"\u001b[39m\u001b[39m with parser \u001b[39m\u001b[39m{1}\u001b[39;00m\u001b[39m.\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m    123\u001b[0m                         \u001b[39m\"\u001b[39m\u001b[39mError: \u001b[39m\u001b[39m{2}\u001b[39;00m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mformat(topology_file, parser, err))\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/core/universe.py:110\u001b[0m, in \u001b[0;36m_topology_from_file_like\u001b[0;34m(topology_file, topology_format, **kwargs)\u001b[0m\n\u001b[1;32m    108\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    109\u001b[0m     \u001b[39mwith\u001b[39;00m parser(topology_file) \u001b[39mas\u001b[39;00m p:\n\u001b[0;32m--> 110\u001b[0m         topology \u001b[39m=\u001b[39m p\u001b[39m.\u001b[39;49mparse(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    111\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    112\u001b[0m     \u001b[39m# There are 2 kinds of errors that might be raised here:\u001b[39;00m\n\u001b[1;32m    113\u001b[0m     \u001b[39m# one because the file isn't present\u001b[39;00m\n\u001b[1;32m    114\u001b[0m     \u001b[39m# or the permissions are bad, second when the parser fails\u001b[39;00m\n\u001b[1;32m    115\u001b[0m     \u001b[39mif\u001b[39;00m (err\u001b[39m.\u001b[39merrno \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m\n\u001b[1;32m    116\u001b[0m         errno\u001b[39m.\u001b[39merrorcode[err\u001b[39m.\u001b[39merrno] \u001b[39min\u001b[39;00m [\u001b[39m'\u001b[39m\u001b[39mENOENT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mEACCES\u001b[39m\u001b[39m'\u001b[39m]):\n\u001b[1;32m    117\u001b[0m         \u001b[39m# Runs if the error is propagated due to no permission / file not found\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/topology/PDBParser.py:210\u001b[0m, in \u001b[0;36mPDBParser.parse\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    203\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mparse\u001b[39m(\u001b[39mself\u001b[39m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m    204\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Parse atom information from PDB file\u001b[39;00m\n\u001b[1;32m    205\u001b[0m \n\u001b[1;32m    206\u001b[0m \u001b[39m    Returns\u001b[39;00m\n\u001b[1;32m    207\u001b[0m \u001b[39m    -------\u001b[39;00m\n\u001b[1;32m    208\u001b[0m \u001b[39m    MDAnalysis Topology object\u001b[39;00m\n\u001b[1;32m    209\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 210\u001b[0m     top \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_parseatoms()\n\u001b[1;32m    212\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    213\u001b[0m         bonds \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_parsebonds(top\u001b[39m.\u001b[39mids\u001b[39m.\u001b[39mvalues)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/topology/PDBParser.py:247\u001b[0m, in \u001b[0;36mPDBParser._parseatoms\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    245\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_wrapped_serials \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m  \u001b[39m# did serials go over 100k?\u001b[39;00m\n\u001b[1;32m    246\u001b[0m last_wrapped_serial \u001b[39m=\u001b[39m \u001b[39m100000\u001b[39m  \u001b[39m# if serials wrap, start from here\u001b[39;00m\n\u001b[0;32m--> 247\u001b[0m \u001b[39mwith\u001b[39;00m util\u001b[39m.\u001b[39mopenany(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfilename) \u001b[39mas\u001b[39;00m f:\n\u001b[1;32m    248\u001b[0m     \u001b[39mfor\u001b[39;00m line \u001b[39min\u001b[39;00m f:\n\u001b[1;32m    249\u001b[0m         line \u001b[39m=\u001b[39m line\u001b[39m.\u001b[39mstrip()  \u001b[39m# Remove extra spaces\u001b[39;00m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py:135\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    133\u001b[0m \u001b[39mdel\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39margs, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mkwds, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfunc\n\u001b[1;32m    134\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 135\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mnext\u001b[39;49m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgen)\n\u001b[1;32m    136\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mStopIteration\u001b[39;00m:\n\u001b[1;32m    137\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mgenerator didn\u001b[39m\u001b[39m'\u001b[39m\u001b[39mt yield\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39mfrom\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:318\u001b[0m, in \u001b[0;36mopenany\u001b[0;34m(datasource, mode, reset)\u001b[0m\n\u001b[1;32m    274\u001b[0m \u001b[39m@contextmanager\u001b[39m\n\u001b[1;32m    275\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mopenany\u001b[39m(datasource, mode\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mrt\u001b[39m\u001b[39m'\u001b[39m, reset\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m):\n\u001b[1;32m    276\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Context manager for :func:`anyopen`.\u001b[39;00m\n\u001b[1;32m    277\u001b[0m \n\u001b[1;32m    278\u001b[0m \u001b[39m    Open the `datasource` and close it when the context of the :keyword:`with`\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[39m    :func:`anyopen`\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 318\u001b[0m     stream \u001b[39m=\u001b[39m anyopen(datasource, mode\u001b[39m=\u001b[39;49mmode, reset\u001b[39m=\u001b[39;49mreset)\n\u001b[1;32m    319\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    320\u001b[0m         \u001b[39myield\u001b[39;00m stream\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:397\u001b[0m, in \u001b[0;36manyopen\u001b[0;34m(datasource, mode, reset)\u001b[0m\n\u001b[1;32m    395\u001b[0m \u001b[39mfor\u001b[39;00m ext \u001b[39min\u001b[39;00m (\u001b[39m'\u001b[39m\u001b[39mbz2\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mgz\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m'\u001b[39m):  \u001b[39m# file == '' should be last\u001b[39;00m\n\u001b[1;32m    396\u001b[0m     openfunc \u001b[39m=\u001b[39m read_handlers[ext]\n\u001b[0;32m--> 397\u001b[0m     stream \u001b[39m=\u001b[39m _get_stream(datasource, openfunc, mode\u001b[39m=\u001b[39;49mmode)\n\u001b[1;32m    398\u001b[0m     \u001b[39mif\u001b[39;00m stream \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    399\u001b[0m         \u001b[39mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:439\u001b[0m, in \u001b[0;36m_get_stream\u001b[0;34m(filename, openfunction, mode)\u001b[0m\n\u001b[1;32m    434\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    435\u001b[0m     \u001b[39m# An exception might be raised due to two reasons, first the openfunction is unable to open the file, in this\u001b[39;00m\n\u001b[1;32m    436\u001b[0m     \u001b[39m# case we have to ignore the error and return None. Second is when openfunction can't open the file because\u001b[39;00m\n\u001b[1;32m    437\u001b[0m     \u001b[39m# either the file isn't there or the permissions don't allow access.\u001b[39;00m\n\u001b[1;32m    438\u001b[0m     \u001b[39mif\u001b[39;00m errno\u001b[39m.\u001b[39merrorcode[err\u001b[39m.\u001b[39merrno] \u001b[39min\u001b[39;00m [\u001b[39m'\u001b[39m\u001b[39mENOENT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mEACCES\u001b[39m\u001b[39m'\u001b[39m]:\n\u001b[0;32m--> 439\u001b[0m         \u001b[39mraise\u001b[39;00m sys\u001b[39m.\u001b[39mexc_info()[\u001b[39m1\u001b[39m] \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m    440\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m    441\u001b[0m \u001b[39mif\u001b[39;00m mode\u001b[39m.\u001b[39mstartswith(\u001b[39m'\u001b[39m\u001b[39mr\u001b[39m\u001b[39m'\u001b[39m):\n\u001b[1;32m    442\u001b[0m     \u001b[39m# additional check for reading (eg can we uncompress) --- is this needed?\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/util.py:433\u001b[0m, in \u001b[0;36m_get_stream\u001b[0;34m(filename, openfunction, mode)\u001b[0m\n\u001b[1;32m    431\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Return open stream if *filename* can be opened with *openfunction* or else ``None``.\"\"\"\u001b[39;00m\n\u001b[1;32m    432\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 433\u001b[0m     stream \u001b[39m=\u001b[39m openfunction(filename, mode\u001b[39m=\u001b[39;49mmode)\n\u001b[1;32m    434\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    435\u001b[0m     \u001b[39m# An exception might be raised due to two reasons, first the openfunction is unable to open the file, in this\u001b[39;00m\n\u001b[1;32m    436\u001b[0m     \u001b[39m# case we have to ignore the error and return None. Second is when openfunction can't open the file because\u001b[39;00m\n\u001b[1;32m    437\u001b[0m     \u001b[39m# either the file isn't there or the permissions don't allow access.\u001b[39;00m\n\u001b[1;32m    438\u001b[0m     \u001b[39mif\u001b[39;00m errno\u001b[39m.\u001b[39merrorcode[err\u001b[39m.\u001b[39merrno] \u001b[39min\u001b[39;00m [\u001b[39m'\u001b[39m\u001b[39mENOENT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mEACCES\u001b[39m\u001b[39m'\u001b[39m]:\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/picklable_file_io.py:478\u001b[0m, in \u001b[0;36mbz2_pickle_open\u001b[0;34m(name, mode)\u001b[0m\n\u001b[1;32m    475\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mOnly read mode (\u001b[39m\u001b[39m'\u001b[39m\u001b[39mr\u001b[39m\u001b[39m'\u001b[39m\u001b[39m, \u001b[39m\u001b[39m'\u001b[39m\u001b[39mrt\u001b[39m\u001b[39m'\u001b[39m\u001b[39m, \u001b[39m\u001b[39m'\u001b[39m\u001b[39mrb\u001b[39m\u001b[39m'\u001b[39m\u001b[39m) \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    476\u001b[0m                      \u001b[39m\"\u001b[39m\u001b[39mfiles can be pickled.\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    477\u001b[0m bz_mode \u001b[39m=\u001b[39m mode\u001b[39m.\u001b[39mreplace(\u001b[39m\"\u001b[39m\u001b[39mt\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 478\u001b[0m binary_file \u001b[39m=\u001b[39m BZ2Picklable(name, bz_mode)\n\u001b[1;32m    479\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mt\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m mode:\n\u001b[1;32m    480\u001b[0m     \u001b[39mreturn\u001b[39;00m TextIOPicklable(binary_file)\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/MDAnalysis/lib/picklable_file_io.py:266\u001b[0m, in \u001b[0;36mBZ2Picklable.__init__\u001b[0;34m(self, name, mode)\u001b[0m\n\u001b[1;32m    264\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\u001b[39mself\u001b[39m, name, mode\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mrb\u001b[39m\u001b[39m'\u001b[39m):\n\u001b[1;32m    265\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_bz_mode \u001b[39m=\u001b[39m mode\n\u001b[0;32m--> 266\u001b[0m     \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49m\u001b[39m__init__\u001b[39;49m(name, mode)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/bz2.py:81\u001b[0m, in \u001b[0;36mBZ2File.__init__\u001b[0;34m(self, filename, mode, compresslevel)\u001b[0m\n\u001b[1;32m     78\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mInvalid mode: \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m (mode,))\n\u001b[1;32m     80\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(filename, (\u001b[39mstr\u001b[39m, \u001b[39mbytes\u001b[39m, os\u001b[39m.\u001b[39mPathLike)):\n\u001b[0;32m---> 81\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_fp \u001b[39m=\u001b[39m _builtin_open(filename, mode)\n\u001b[1;32m     82\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_closefp \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m     83\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_mode \u001b[39m=\u001b[39m mode_code\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/5bs8.pdb'"
     ]
    }
   ],
   "source": [
    "import MDAnalysis as mda\n",
    "\n",
    "pdb = mda.Universe('data/5bs8.pdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AtomGroup with 23125 atoms>"
      ]
     },
     "execution_count": 436,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein = pdb.select_atoms(\"protein\")\n",
    "protein\n",
    "#protein.write(\"data/5bs8_atoms.pdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 437,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant</th>\n",
       "      <th>Predicted Effect</th>\n",
       "      <th>Score</th>\n",
       "      <th>Expected Accuracy</th>\n",
       "      <th>segid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T1A</td>\n",
       "      <td>effect</td>\n",
       "      <td>25</td>\n",
       "      <td>63%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T1R</td>\n",
       "      <td>effect</td>\n",
       "      <td>57</td>\n",
       "      <td>75%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T1N</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-80</td>\n",
       "      <td>87%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T1D</td>\n",
       "      <td>effect</td>\n",
       "      <td>21</td>\n",
       "      <td>63%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T1C</td>\n",
       "      <td>effect</td>\n",
       "      <td>13</td>\n",
       "      <td>59%</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15115</th>\n",
       "      <td>V253S</td>\n",
       "      <td>effect</td>\n",
       "      <td>52</td>\n",
       "      <td>75%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15116</th>\n",
       "      <td>V253T</td>\n",
       "      <td>effect</td>\n",
       "      <td>30</td>\n",
       "      <td>66%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15117</th>\n",
       "      <td>V253W</td>\n",
       "      <td>effect</td>\n",
       "      <td>66</td>\n",
       "      <td>80%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15118</th>\n",
       "      <td>V253Y</td>\n",
       "      <td>effect</td>\n",
       "      <td>56</td>\n",
       "      <td>75%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15119</th>\n",
       "      <td>V253V</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-99</td>\n",
       "      <td>97%</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15120 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Variant Predicted Effect  Score Expected Accuracy segid\n",
       "0         T1A           effect     25               63%     A\n",
       "1         T1R           effect     57               75%     A\n",
       "2         T1N          neutral    -80               87%     A\n",
       "3         T1D           effect     21               63%     A\n",
       "4         T1C           effect     13               59%     A\n",
       "...       ...              ...    ...               ...   ...\n",
       "15115   V253S           effect     52               75%     B\n",
       "15116   V253T           effect     30               66%     B\n",
       "15117   V253W           effect     66               80%     B\n",
       "15118   V253Y           effect     56               75%     B\n",
       "15119   V253V          neutral    -99               97%     B\n",
       "\n",
       "[15120 rows x 5 columns]"
      ]
     },
     "execution_count": 437,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snap2 = pd.read_csv('data/5bs8_SNAP2_complete.csv')\n",
    "snap2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "metadata": {},
   "outputs": [],
   "source": [
    "def correct_position(row):\n",
    "    \n",
    "    if row.segid == 'A':\n",
    "        increase_by = 1\n",
    "    else:\n",
    "        increase_by = 422\n",
    "\n",
    "    letter1 = row.Variant[0]\n",
    "    letter2 = row.Variant[-1]  # Extract the letters (all characters except the last one)\n",
    "    number = int(row.Variant[1:-1])  # Extract the number and convert it to an integer\n",
    "    updated_number = number + increase_by  # Increase the number by the specified amount\n",
    "    updated_string = f'{letter1}{updated_number}{letter2}'  # Create the updated string\n",
    "\n",
    "    return updated_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {},
   "outputs": [],
   "source": [
    "snap2['Variant'] = snap2.apply(correct_position, axis = 1)\n",
    "snap2.to_csv('data/5bs8_SNAP2_complete_clean.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 440,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dict = {}\n",
    "names = ['MXF_df', 'LEV_df']\n",
    "i=0\n",
    "\n",
    "for dataset in [MXF_DATASET, LEV_DATASET]:\n",
    "\n",
    "    sbml_dataset = dataset[['segid', 'mutation']]\n",
    "    sbml_dataset.drop_duplicates(subset='mutation', keep='first', inplace=True)\n",
    "\n",
    "    features = sbmlcore.FeatureDataset(sbml_dataset, species='M. tuberculosis', protein = 'DNA Gyrase')\n",
    "\n",
    "    a = sbmlcore.AminoAcidVolumeChange()\n",
    "    b1 = sbmlcore.AminoAcidHydropathyChangeKyteDoolittle()\n",
    "    b2 = sbmlcore.AminoAcidHydropathyChangeWimleyWhite()\n",
    "    c = sbmlcore.AminoAcidMWChange()\n",
    "    d = sbmlcore.AminoAcidPiChange()\n",
    "    e = sbmlcore.Stride('data/pdb/5bs8_atoms.pdb')\n",
    "    f = sbmlcore.FreeSASA('data/pdb/5bs8_atoms.pdb')\n",
    "    g = sbmlcore.SNAP2(\"data/5bs8_SNAP2_complete_clean.csv\", offsets = {'A': 0, 'B': 0})\n",
    "    h1 = sbmlcore.StructuralDistances('data/pdb/5bs8_no_PTR_su_AB_MGB.pdb', distance_selection=\"resname MG\", distance_name='dist_MGB')\n",
    "    h2 = sbmlcore.StructuralDistances('data/pdb/5bs8_no_PTR_su_AB_MGE.pdb', distance_selection=\"resname MG\", distance_name='dist_MGE')\n",
    "    # h3 = sbmlcore.StructuralDistances('data/pdb/5bs8_no_PTR_su_AB_MGE.pdb', distance_selection=\"resname MG\", distance_name='dist_MGE')\n",
    "    # h4 = sbmlcore.StructuralDistances('data/pdb/5bs8_no_PTR_su_AB_MGF.pdb', distance_selection=\"resname MG\", distance_name='dist_MGF')\n",
    "    k = sbmlcore.StructuralDistances('data/pdb/5bs8_no_PTR_su_AB_one_drug.pdb', distance_selection=\"resname MFX\", distance_name='dist_MFX')\n",
    "\n",
    "    features.add_feature([a, b1, b2, c, d, e, f, g, h1, h2, k])\n",
    "    features = features.df\n",
    "\n",
    "    dataset.reset_index(inplace = True)\n",
    "\n",
    "    variable_name = names[i]\n",
    "    i = i+1\n",
    "    \n",
    "    data_dict[variable_name] = pd.merge(dataset, features, how='left', on=['segid','mutation'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['UNIQUEID', 'DRUG', 'MIC', 'LOG2MIC', 'BINARY_PHENOTYPE',\n",
       "       'PHENOTYPE_QUALITY', 'GENE', 'POSITION', 'GENE_MUTATION', 'segid',\n",
       "       'mutation', 'd_volume', 'd_hydropathy_KD', 'd_hydropathy_WW', 'd_MW',\n",
       "       'd_Pi', 'secondary_structure', 'secondary_structure_long', 'phi', 'psi',\n",
       "       'residue_sasa', 'n_hbond_acceptors', 'n_hbond_donors', 'B', 'C', 'E',\n",
       "       'G', 'H', 'T', 'SASA', 'snap2_score', 'snap2_accuracy', 'dist_MGB',\n",
       "       'dist_MGE', 'dist_MFX'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 441,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = data_dict['MXF_df']\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(152,)"
      ]
     },
     "execution_count": 442,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.GENE_MUTATION.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "File does not exist!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[443], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m g \u001b[39m=\u001b[39m sbmlcore\u001b[39m.\u001b[39;49mTempFactors(\u001b[39m'\u001b[39;49m\u001b[39mdata/5bs8.pdb\u001b[39;49m\u001b[39m'\u001b[39;49m)\n\u001b[1;32m      2\u001b[0m g\n",
      "File \u001b[0;32m~/Documents/Studium/PhD/DPhil/predict_FQ_resistance/predict_FQ_resistance/.venv_fq/lib/python3.10/site-packages/sbmlcore/TempFactors.py:29\u001b[0m, in \u001b[0;36mTempFactors.__init__\u001b[0;34m(self, pdb_file, offsets)\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\u001b[39mself\u001b[39m, pdb_file, offsets\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[1;32m     27\u001b[0m \n\u001b[1;32m     28\u001b[0m     \u001b[39m# check file exists\u001b[39;00m\n\u001b[0;32m---> 29\u001b[0m     \u001b[39massert\u001b[39;00m pathlib\u001b[39m.\u001b[39mPath(pdb_file)\u001b[39m.\u001b[39mis_file(), \u001b[39m\"\u001b[39m\u001b[39mFile does not exist!\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     31\u001b[0m     u \u001b[39m=\u001b[39m MDAnalysis\u001b[39m.\u001b[39mUniverse(pdb_file)\n\u001b[1;32m     33\u001b[0m     \u001b[39m# apply any offsets to the residue numbering\u001b[39;00m\n\u001b[1;32m     34\u001b[0m     \u001b[39m# as specified in the supplied offsets dict e.g. {'A': 3, 'B': -4}\u001b[39;00m\n\u001b[1;32m     35\u001b[0m     \u001b[39m# Chain is segid i.e. A, B, C etc.\u001b[39;00m\n",
      "\u001b[0;31mAssertionError\u001b[0m: File does not exist!"
     ]
    }
   ],
   "source": [
    "g = sbmlcore.TempFactors('data/5bs8.pdb')\n",
    "g"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Further features to be added soon (SNAP2, DeepDG, StructuralDistances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 444,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1735 entries, 0 to 1734\n",
      "Data columns (total 35 columns):\n",
      " #   Column                    Non-Null Count  Dtype   \n",
      "---  ------                    --------------  -----   \n",
      " 0   UNIQUEID                  1735 non-null   object  \n",
      " 1   DRUG                      1735 non-null   category\n",
      " 2   MIC                       1722 non-null   category\n",
      " 3   LOG2MIC                   1722 non-null   float64 \n",
      " 4   BINARY_PHENOTYPE          1734 non-null   object  \n",
      " 5   PHENOTYPE_QUALITY         1735 non-null   category\n",
      " 6   GENE                      1735 non-null   object  \n",
      " 7   POSITION                  1735 non-null   float64 \n",
      " 8   GENE_MUTATION             1735 non-null   object  \n",
      " 9   segid                     1735 non-null   object  \n",
      " 10  mutation                  1735 non-null   object  \n",
      " 11  d_volume                  1735 non-null   float64 \n",
      " 12  d_hydropathy_KD           1735 non-null   float64 \n",
      " 13  d_hydropathy_WW           1735 non-null   float64 \n",
      " 14  d_MW                      1735 non-null   float64 \n",
      " 15  d_Pi                      1735 non-null   float64 \n",
      " 16  secondary_structure       1726 non-null   object  \n",
      " 17  secondary_structure_long  1726 non-null   object  \n",
      " 18  phi                       1726 non-null   float64 \n",
      " 19  psi                       1726 non-null   float64 \n",
      " 20  residue_sasa              1726 non-null   float64 \n",
      " 21  n_hbond_acceptors         1726 non-null   float64 \n",
      " 22  n_hbond_donors            1726 non-null   float64 \n",
      " 23  B                         1726 non-null   object  \n",
      " 24  C                         1726 non-null   object  \n",
      " 25  E                         1726 non-null   object  \n",
      " 26  G                         1726 non-null   object  \n",
      " 27  H                         1726 non-null   object  \n",
      " 28  T                         1726 non-null   object  \n",
      " 29  SASA                      1735 non-null   float64 \n",
      " 30  snap2_score               1735 non-null   int64   \n",
      " 31  snap2_accuracy            1735 non-null   int64   \n",
      " 32  dist_MGB                  1726 non-null   float64 \n",
      " 33  dist_MGE                  1726 non-null   float64 \n",
      " 34  dist_MFX                  1726 non-null   float64 \n",
      "dtypes: category(3), float64(16), int64(2), object(14)\n",
      "memory usage: 441.1+ KB\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1669 entries, 0 to 1668\n",
      "Data columns (total 35 columns):\n",
      " #   Column                    Non-Null Count  Dtype   \n",
      "---  ------                    --------------  -----   \n",
      " 0   UNIQUEID                  1669 non-null   object  \n",
      " 1   DRUG                      1669 non-null   category\n",
      " 2   MIC                       1651 non-null   category\n",
      " 3   LOG2MIC                   1651 non-null   float64 \n",
      " 4   BINARY_PHENOTYPE          1667 non-null   object  \n",
      " 5   PHENOTYPE_QUALITY         1669 non-null   category\n",
      " 6   GENE                      1669 non-null   object  \n",
      " 7   POSITION                  1669 non-null   float64 \n",
      " 8   GENE_MUTATION             1669 non-null   object  \n",
      " 9   segid                     1669 non-null   object  \n",
      " 10  mutation                  1669 non-null   object  \n",
      " 11  d_volume                  1669 non-null   float64 \n",
      " 12  d_hydropathy_KD           1669 non-null   float64 \n",
      " 13  d_hydropathy_WW           1669 non-null   float64 \n",
      " 14  d_MW                      1669 non-null   float64 \n",
      " 15  d_Pi                      1669 non-null   float64 \n",
      " 16  secondary_structure       1666 non-null   object  \n",
      " 17  secondary_structure_long  1666 non-null   object  \n",
      " 18  phi                       1666 non-null   float64 \n",
      " 19  psi                       1666 non-null   float64 \n",
      " 20  residue_sasa              1666 non-null   float64 \n",
      " 21  n_hbond_acceptors         1666 non-null   float64 \n",
      " 22  n_hbond_donors            1666 non-null   float64 \n",
      " 23  B                         1666 non-null   object  \n",
      " 24  C                         1666 non-null   object  \n",
      " 25  E                         1666 non-null   object  \n",
      " 26  G                         1666 non-null   object  \n",
      " 27  H                         1666 non-null   object  \n",
      " 28  T                         1666 non-null   object  \n",
      " 29  SASA                      1669 non-null   float64 \n",
      " 30  snap2_score               1669 non-null   int64   \n",
      " 31  snap2_accuracy            1669 non-null   int64   \n",
      " 32  dist_MGB                  1666 non-null   float64 \n",
      " 33  dist_MGE                  1666 non-null   float64 \n",
      " 34  dist_MFX                  1666 non-null   float64 \n",
      "dtypes: category(3), float64(16), int64(2), object(14)\n",
      "memory usage: 424.4+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "for dataset in ['LEV_df', 'MXF_df']:\n",
    "\n",
    "    df = data_dict[dataset]\n",
    "    print(df.info())\n",
    "    \n",
    "    name = f'data/ML_DATA_{dataset[:3]}.csv'\n",
    "    df.to_csv(name, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENE_MUTATION</th>\n",
       "      <th>segid</th>\n",
       "      <th>...</th>\n",
       "      <th>E</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>T</th>\n",
       "      <th>SASA</th>\n",
       "      <th>snap2_score</th>\n",
       "      <th>snap2_accuracy</th>\n",
       "      <th>dist_MGB</th>\n",
       "      <th>dist_MGE</th>\n",
       "      <th>dist_MFX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.04.subj.00097.lab.633433.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>88.155089</td>\n",
       "      <td>63</td>\n",
       "      <td>80</td>\n",
       "      <td>18.494689</td>\n",
       "      <td>7.838828</td>\n",
       "      <td>14.503318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.10.subj.UH00935117.lab.UH00935117.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55.555816</td>\n",
       "      <td>-64</td>\n",
       "      <td>82</td>\n",
       "      <td>14.434102</td>\n",
       "      <td>5.483332</td>\n",
       "      <td>10.609526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>88.155089</td>\n",
       "      <td>56</td>\n",
       "      <td>75</td>\n",
       "      <td>18.494689</td>\n",
       "      <td>7.838828</td>\n",
       "      <td>14.503318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.04.subj.00121.lab.633619.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>88.155089</td>\n",
       "      <td>63</td>\n",
       "      <td>80</td>\n",
       "      <td>18.494689</td>\n",
       "      <td>7.838828</td>\n",
       "      <td>14.503318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>site.14.subj.2591.lab.2591.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>gyrA_D94N</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>88.155089</td>\n",
       "      <td>56</td>\n",
       "      <td>75</td>\n",
       "      <td>18.494689</td>\n",
       "      <td>7.838828</td>\n",
       "      <td>14.503318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664</th>\n",
       "      <td>site.04.subj.01310.lab.724468.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>91.0</td>\n",
       "      <td>gyrA_S91P</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>65.228155</td>\n",
       "      <td>63</td>\n",
       "      <td>80</td>\n",
       "      <td>13.712391</td>\n",
       "      <td>5.381126</td>\n",
       "      <td>11.010229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665</th>\n",
       "      <td>site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>MEDIUM</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55.555816</td>\n",
       "      <td>-64</td>\n",
       "      <td>82</td>\n",
       "      <td>14.434102</td>\n",
       "      <td>5.483332</td>\n",
       "      <td>10.609526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1666</th>\n",
       "      <td>site.10.subj.UH01302806.lab.UH01302806.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55.555816</td>\n",
       "      <td>-64</td>\n",
       "      <td>82</td>\n",
       "      <td>14.434102</td>\n",
       "      <td>5.483332</td>\n",
       "      <td>10.609526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1667</th>\n",
       "      <td>site.05.subj.CA-0020.lab.CO-05421-18.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>594.0</td>\n",
       "      <td>gyrB_G594R</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>2.551348</td>\n",
       "      <td>33</td>\n",
       "      <td>66</td>\n",
       "      <td>29.300727</td>\n",
       "      <td>43.081517</td>\n",
       "      <td>41.932227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1668</th>\n",
       "      <td>site.06.subj.06TB_0968.lab.06MIL1963.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>&gt;4</td>\n",
       "      <td>3.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrA_A90V</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>55.555816</td>\n",
       "      <td>-64</td>\n",
       "      <td>82</td>\n",
       "      <td>14.434102</td>\n",
       "      <td>5.483332</td>\n",
       "      <td>10.609526</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1669 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          UNIQUEID DRUG   MIC  LOG2MIC  \\\n",
       "0              site.04.subj.00097.lab.633433.iso.1  MXF   2.0     1.00   \n",
       "1     site.10.subj.UH00935117.lab.UH00935117.iso.1  MXF   4.0     2.00   \n",
       "2     site.11.subj.XTB-18-164.lab.XTB-18-164.iso.1  MXF    >4     3.00   \n",
       "3              site.04.subj.00121.lab.633619.iso.1  MXF   2.0     1.00   \n",
       "4                 site.14.subj.2591.lab.2591.iso.1  MXF    >4     3.00   \n",
       "...                                            ...  ...   ...      ...   \n",
       "1664           site.04.subj.01310.lab.724468.iso.1  MXF   1.0     0.00   \n",
       "1665  site.11.subj.XTB-18-186.lab.XTB-18-186.iso.1  MXF   2.0     1.00   \n",
       "1666  site.10.subj.UH01302806.lab.UH01302806.iso.1  MXF   2.0     1.00   \n",
       "1667    site.05.subj.CA-0020.lab.CO-05421-18.iso.1  MXF  0.12    -3.06   \n",
       "1668    site.06.subj.06TB_0968.lab.06MIL1963.iso.1  MXF    >4     3.00   \n",
       "\n",
       "     BINARY_PHENOTYPE PHENOTYPE_QUALITY  GENE  POSITION GENE_MUTATION segid  \\\n",
       "0                   R              HIGH  gyrA      94.0     gyrA_D94G     A   \n",
       "1                   R              HIGH  gyrA      90.0     gyrA_A90V     A   \n",
       "2                   R            MEDIUM  gyrA      94.0     gyrA_D94N     A   \n",
       "3                   R              HIGH  gyrA      94.0     gyrA_D94G     A   \n",
       "4                   R              HIGH  gyrA      94.0     gyrA_D94N     A   \n",
       "...               ...               ...   ...       ...           ...   ...   \n",
       "1664                R              HIGH  gyrA      91.0     gyrA_S91P     A   \n",
       "1665                R            MEDIUM  gyrA      90.0     gyrA_A90V     A   \n",
       "1666                R              HIGH  gyrA      90.0     gyrA_A90V     A   \n",
       "1667                S              HIGH  gyrB     594.0    gyrB_G594R     B   \n",
       "1668                R              HIGH  gyrA      90.0     gyrA_A90V     A   \n",
       "\n",
       "      ...      E      G     H      T       SASA  snap2_score snap2_accuracy  \\\n",
       "0     ...  False  False  True  False  88.155089           63             80   \n",
       "1     ...  False  False  True  False  55.555816          -64             82   \n",
       "2     ...  False  False  True  False  88.155089           56             75   \n",
       "3     ...  False  False  True  False  88.155089           63             80   \n",
       "4     ...  False  False  True  False  88.155089           56             75   \n",
       "...   ...    ...    ...   ...    ...        ...          ...            ...   \n",
       "1664  ...  False  False  True  False  65.228155           63             80   \n",
       "1665  ...  False  False  True  False  55.555816          -64             82   \n",
       "1666  ...  False  False  True  False  55.555816          -64             82   \n",
       "1667  ...  False  False  True  False   2.551348           33             66   \n",
       "1668  ...  False  False  True  False  55.555816          -64             82   \n",
       "\n",
       "       dist_MGB   dist_MGE   dist_MFX  \n",
       "0     18.494689   7.838828  14.503318  \n",
       "1     14.434102   5.483332  10.609526  \n",
       "2     18.494689   7.838828  14.503318  \n",
       "3     18.494689   7.838828  14.503318  \n",
       "4     18.494689   7.838828  14.503318  \n",
       "...         ...        ...        ...  \n",
       "1664  13.712391   5.381126  11.010229  \n",
       "1665  14.434102   5.483332  10.609526  \n",
       "1666  14.434102   5.483332  10.609526  \n",
       "1667  29.300727  43.081517  41.932227  \n",
       "1668  14.434102   5.483332  10.609526  \n",
       "\n",
       "[1669 rows x 35 columns]"
      ]
     },
     "execution_count": 445,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
